// n8n-nodes-base/dist/nodes/Aws/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.awsApiRequestSOAP = exports.awsApiRequestREST = exports.awsApiRequest = void 0;
const xml2js_1 = require("xml2js");
const n8n_workflow_1 = require("n8n-workflow");
async function awsApiRequest(service, method, path, body, headers) {
    const credentials = await this.getCredentials('aws');
    const requestOptions = {
        qs: {
            service,
            path,
        },
        method,
        body: service === 'lambda' ? body : JSON.stringify(body),
        url: '',
        headers,
        region: credentials === null || credentials === void 0 ? void 0 : credentials.region,
    };
    try {
        return await this.helpers.requestWithAuthentication.call(this, 'aws', requestOptions);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error, { parseXml: true });
    }
}
exports.awsApiRequest = awsApiRequest;
async function awsApiRequestREST(service, method, path, body, headers) {
    const response = await awsApiRequest.call(this, service, method, path, body, headers);
    try {
        return JSON.parse(response);
    }
    catch (error) {
        return response;
    }
}
exports.awsApiRequestREST = awsApiRequestREST;
async function awsApiRequestSOAP(service, method, path, body, headers) {
    const response = await awsApiRequest.call(this, service, method, path, body, headers);
    try {
        return await new Promise((resolve, reject) => {
            (0, xml2js_1.parseString)(response, { explicitArray: false }, (err, data) => {
                if (err) {
                    return reject(err);
                }
                resolve(data);
            });
        });
    }
    catch (error) {
        return response;
    }
}
exports.awsApiRequestSOAP = awsApiRequestSOAP;
