// n8n-nodes-base/dist/nodes/Linear/GenericFunctions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sort = exports.validateCredentials = exports.linearApiRequestAllItems = exports.capitalizeFirstLetter = exports.linearApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const lodash_get_1 = __importDefault(require("lodash.get"));
const Queries_1 = require("./Queries");
async function linearApiRequest(body = {}, option = {}) {
    const endpoint = 'https://api.linear.app/graphql';
    let options = {
        headers: {
            'Content-Type': 'application/json',
        },
        method: 'POST',
        body,
        uri: endpoint,
        json: true,
    };
    options = Object.assign({}, options, option);
    try {
        return await this.helpers.requestWithAuthentication.call(this, 'linearApi', options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.linearApiRequest = linearApiRequest;
function capitalizeFirstLetter(data) {
    return data.charAt(0).toUpperCase() + data.slice(1);
}
exports.capitalizeFirstLetter = capitalizeFirstLetter;
async function linearApiRequestAllItems(propertyName, body = {}) {
    const returnData = [];
    let responseData;
    body.variables.first = 50;
    body.variables.after = null;
    do {
        responseData = await linearApiRequest.call(this, body);
        returnData.push.apply(returnData, (0, lodash_get_1.default)(responseData, `${propertyName}.nodes`));
        body.variables.after = (0, lodash_get_1.default)(responseData, `${propertyName}.pageInfo.endCursor`);
    } while ((0, lodash_get_1.default)(responseData, `${propertyName}.pageInfo.hasNextPage`));
    return returnData;
}
exports.linearApiRequestAllItems = linearApiRequestAllItems;
async function validateCredentials(decryptedCredentials) {
    const credentials = decryptedCredentials;
    const options = {
        headers: {
            'Content-Type': 'application/json',
            Authorization: credentials.apiKey,
        },
        method: 'POST',
        body: {
            query: Queries_1.query.getIssues(),
            variables: {
                first: 1,
            },
        },
        uri: 'https://api.linear.app/graphql',
        json: true,
    };
    return this.helpers.request(options);
}
exports.validateCredentials = validateCredentials;
const sort = (a, b) => {
    if (a.name.toLocaleLowerCase() < b.name.toLocaleLowerCase()) {
        return -1;
    }
    if (a.name.toLocaleLowerCase() > b.name.toLocaleLowerCase()) {
        return 1;
    }
    return 0;
};
exports.sort = sort;
