// n8n-nodes-base/dist/nodes/Google/Analytics/v2/transport/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.googleApiRequestAllItems = exports.googleApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function googleApiRequest(method, endpoint, body = {}, qs = {}, uri, option = {}) {
    var _a;
    const propertyType = this.getNodeParameter('propertyType', 0, 'universal');
    const baseURL = propertyType === 'ga4'
        ? 'https://analyticsdata.googleapis.com'
        : 'https://analyticsreporting.googleapis.com';
    let options = {
        headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json',
        },
        method,
        body,
        qs,
        uri: uri !== null && uri !== void 0 ? uri : `${baseURL}${endpoint}`,
        json: true,
    };
    options = Object.assign({}, options, option);
    try {
        if (Object.keys(body).length === 0) {
            delete options.body;
        }
        if (Object.keys(qs).length === 0) {
            delete options.qs;
        }
        return await this.helpers.requestOAuth2.call(this, 'googleAnalyticsOAuth2', options);
    }
    catch (error) {
        const errorData = (error.message || '').split(' - ')[1];
        if (errorData) {
            const parsedError = JSON.parse(errorData.trim());
            if ((_a = parsedError.error) === null || _a === void 0 ? void 0 : _a.message) {
                const [message, ...rest] = parsedError.error.message.split('\n');
                const description = rest.join('\n');
                const httpCode = parsedError.error.code;
                throw new n8n_workflow_1.NodeApiError(this.getNode(), error, {
                    message,
                    description,
                    httpCode,
                });
            }
        }
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error, { message: error.message });
    }
}
exports.googleApiRequest = googleApiRequest;
async function googleApiRequestAllItems(propertyName, method, endpoint, body = {}, query = {}, uri) {
    var _a;
    const propertyType = this.getNodeParameter('propertyType', 0, 'universal');
    const returnData = [];
    let responseData;
    if (propertyType === 'ga4') {
        let rows = [];
        query.limit = 100000;
        query.offset = 0;
        responseData = await googleApiRequest.call(this, method, endpoint, body, query, uri);
        rows = rows.concat(responseData.rows);
        query.offset = rows.length;
        while (responseData.rowCount > rows.length) {
            responseData = await googleApiRequest.call(this, method, endpoint, body, query, uri);
            rows = rows.concat(responseData.rows);
            query.offset = rows.length;
        }
        responseData.rows = rows;
        returnData.push(responseData);
    }
    else {
        do {
            responseData = await googleApiRequest.call(this, method, endpoint, body, query, uri);
            if (body.reportRequests && Array.isArray(body.reportRequests)) {
                body.reportRequests[0].pageToken =
                    responseData[propertyName][0].nextPageToken;
            }
            else {
                body.pageToken = responseData.nextPageToken;
            }
            returnData.push.apply(returnData, responseData[propertyName]);
        } while ((responseData.nextPageToken !== undefined && responseData.nextPageToken !== '') ||
            ((_a = responseData[propertyName]) === null || _a === void 0 ? void 0 : _a[0].nextPageToken) !== undefined);
    }
    return returnData;
}
exports.googleApiRequestAllItems = googleApiRequestAllItems;
