// n8n-nodes-base/dist/nodes/Jira/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.filterSortSearchListItems = exports.allEvents = exports.simplifyIssueOutput = exports.getId = exports.eventExists = exports.validateJSON = exports.jiraSoftwareCloudApiRequestAllItems = exports.jiraSoftwareCloudApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function jiraSoftwareCloudApiRequest(endpoint, method, body = {}, query, uri, option = {}) {
    var _a;
    const jiraVersion = this.getNodeParameter('jiraVersion', 0);
    let domain = '';
    let credentialType;
    if (jiraVersion === 'server') {
        domain = (await this.getCredentials('jiraSoftwareServerApi')).domain;
        credentialType = 'jiraSoftwareServerApi';
    }
    else {
        domain = (await this.getCredentials('jiraSoftwareCloudApi')).domain;
        credentialType = 'jiraSoftwareCloudApi';
    }
    const options = {
        headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json',
            'X-Atlassian-Token': 'no-check',
        },
        method,
        qs: query,
        uri: uri || `${domain}/rest${endpoint}`,
        body,
        json: true,
    };
    if (Object.keys(option).length !== 0) {
        Object.assign(options, option);
    }
    if (Object.keys(body).length === 0) {
        delete options.body;
    }
    if (Object.keys(query || {}).length === 0) {
        delete options.qs;
    }
    try {
        return await this.helpers.requestWithAuthentication.call(this, credentialType, options);
    }
    catch (error) {
        if (((_a = error.description) === null || _a === void 0 ? void 0 : _a.includes) &&
            error.description.includes("Field 'priority' cannot be set")) {
            throw new n8n_workflow_1.NodeApiError(this.getNode(), error, {
                message: "Field 'priority' cannot be set. You need to add the Priority field to your Jira Project's Issue Types.",
            });
        }
        throw error;
    }
}
exports.jiraSoftwareCloudApiRequest = jiraSoftwareCloudApiRequest;
async function jiraSoftwareCloudApiRequestAllItems(propertyName, endpoint, method, body = {}, query = {}) {
    const returnData = [];
    let responseData;
    query.startAt = 0;
    body.startAt = 0;
    query.maxResults = 100;
    body.maxResults = 100;
    do {
        responseData = await jiraSoftwareCloudApiRequest.call(this, endpoint, method, body, query);
        returnData.push.apply(returnData, responseData[propertyName]);
        query.startAt = responseData.startAt + responseData.maxResults;
        body.startAt = responseData.startAt + responseData.maxResults;
    } while (responseData.startAt + responseData.maxResults <
        responseData.total);
    return returnData;
}
exports.jiraSoftwareCloudApiRequestAllItems = jiraSoftwareCloudApiRequestAllItems;
function validateJSON(json) {
    let result;
    try {
        result = JSON.parse(json);
    }
    catch (exception) {
        result = '';
    }
    return result;
}
exports.validateJSON = validateJSON;
function eventExists(currentEvents, webhookEvents) {
    for (const currentEvent of currentEvents) {
        if (!webhookEvents.includes(currentEvent)) {
            return false;
        }
    }
    return true;
}
exports.eventExists = eventExists;
function getId(url) {
    return url.split('/').pop();
}
exports.getId = getId;
function simplifyIssueOutput(responseData) {
    const mappedFields = {
        id: responseData.id,
        key: responseData.key,
        self: responseData.self,
    };
    const customField = /^customfield_\d+$/;
    const sortedFields = Object.keys(responseData.fields).sort((a, b) => {
        if (customField.test(a) && customField.test(b)) {
            return a > b ? 1 : -1;
        }
        if (customField.test(a)) {
            return 1;
        }
        if (customField.test(b)) {
            return -1;
        }
        return a > b ? 1 : -1;
    });
    for (const field of sortedFields) {
        if (responseData.names[field] in mappedFields) {
            let newField = responseData.names[field];
            let counter = 0;
            while (newField in mappedFields) {
                counter++;
                newField = `${responseData.names[field]}_${counter}`;
            }
            mappedFields[newField] = responseData.fields[field];
        }
        else {
            mappedFields[responseData.names[field] || field] = responseData.fields[field];
        }
    }
    return mappedFields;
}
exports.simplifyIssueOutput = simplifyIssueOutput;
exports.allEvents = [
    'board_created',
    'board_updated',
    'board_deleted',
    'board_configuration_changed',
    'comment_created',
    'comment_updated',
    'comment_deleted',
    'jira:issue_created',
    'jira:issue_updated',
    'jira:issue_deleted',
    'option_voting_changed',
    'option_watching_changed',
    'option_unassigned_issues_changed',
    'option_subtasks_changed',
    'option_attachments_changed',
    'option_issuelinks_changed',
    'option_timetracking_changed',
    'project_created',
    'project_updated',
    'project_deleted',
    'sprint_created',
    'sprint_deleted',
    'sprint_updated',
    'sprint_started',
    'sprint_closed',
    'user_created',
    'user_updated',
    'user_deleted',
    'jira:version_released',
    'jira:version_unreleased',
    'jira:version_created',
    'jira:version_moved',
    'jira:version_updated',
    'jira:version_deleted',
    'issuelink_created',
    'issuelink_deleted',
    'worklog_created',
    'worklog_updated',
    'worklog_deleted',
];
function filterSortSearchListItems(items, filter) {
    return items
        .filter((item) => !filter ||
        item.name.toLowerCase().includes(filter.toLowerCase()) ||
        item.value.toString().toLowerCase().includes(filter.toLowerCase()))
        .sort((a, b) => {
        if (a.name.toLocaleLowerCase() < b.name.toLocaleLowerCase()) {
            return -1;
        }
        if (a.name.toLocaleLowerCase() > b.name.toLocaleLowerCase()) {
            return 1;
        }
        return 0;
    });
}
exports.filterSortSearchListItems = filterSortSearchListItems;
