// n8n-nodes-base/dist/nodes/Jira/Jira.node.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Jira = void 0;
const lodash_mergewith_1 = __importDefault(require("lodash.mergewith"));
const n8n_workflow_1 = require("n8n-workflow");
const GenericFunctions_1 = require("./GenericFunctions");
const IssueAttachmentDescription_1 = require("./IssueAttachmentDescription");
const IssueCommentDescription_1 = require("./IssueCommentDescription");
const IssueDescription_1 = require("./IssueDescription");
const UserDescription_1 = require("./UserDescription");
class Jira {
    constructor() {
        this.description = {
            displayName: 'Jira Software',
            name: 'jira',
            icon: 'file:jira.svg',
            group: ['output'],
            version: 1,
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            description: 'Consume Jira Software API',
            defaults: {
                name: 'Jira Software',
            },
            inputs: ['main'],
            outputs: ['main'],
            credentials: [
                {
                    name: 'jiraSoftwareCloudApi',
                    required: true,
                    displayOptions: {
                        show: {
                            jiraVersion: ['cloud'],
                        },
                    },
                },
                {
                    name: 'jiraSoftwareServerApi',
                    required: true,
                    displayOptions: {
                        show: {
                            jiraVersion: ['server'],
                        },
                    },
                },
            ],
            properties: [
                {
                    displayName: 'Jira Version',
                    name: 'jiraVersion',
                    type: 'options',
                    options: [
                        {
                            name: 'Cloud',
                            value: 'cloud',
                        },
                        {
                            name: 'Server (Self Hosted)',
                            value: 'server',
                        },
                    ],
                    default: 'cloud',
                },
                {
                    displayName: 'Resource',
                    name: 'resource',
                    type: 'options',
                    noDataExpression: true,
                    options: [
                        {
                            name: 'Issue',
                            value: 'issue',
                            description: 'Creates an issue or, where the option to create subtasks is enabled in Jira, a subtask',
                        },
                        {
                            name: 'Issue Attachment',
                            value: 'issueAttachment',
                            description: 'Add, remove, and get an attachment from an issue',
                        },
                        {
                            name: 'Issue Comment',
                            value: 'issueComment',
                            description: 'Get, create, update, and delete a comment from an issue',
                        },
                        {
                            name: 'User',
                            value: 'user',
                            description: 'Get, create and delete a user',
                        },
                    ],
                    default: 'issue',
                },
                ...IssueDescription_1.issueOperations,
                ...IssueDescription_1.issueFields,
                ...IssueAttachmentDescription_1.issueAttachmentOperations,
                ...IssueAttachmentDescription_1.issueAttachmentFields,
                ...IssueCommentDescription_1.issueCommentOperations,
                ...IssueCommentDescription_1.issueCommentFields,
                ...UserDescription_1.userOperations,
                ...UserDescription_1.userFields,
            ],
        };
        this.methods = {
            listSearch: {
                async getProjects(filter) {
                    const returnData = [];
                    const jiraVersion = this.getCurrentNodeParameter('jiraVersion');
                    let endpoint = '';
                    let projects;
                    if (jiraVersion === 'server') {
                        endpoint = '/api/2/project';
                        projects = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, endpoint, 'GET');
                    }
                    else {
                        endpoint = '/api/2/project/search';
                        projects = await GenericFunctions_1.jiraSoftwareCloudApiRequestAllItems.call(this, 'values', endpoint, 'GET');
                    }
                    if (projects.values && Array.isArray(projects.values)) {
                        projects = projects.values;
                    }
                    for (const project of projects) {
                        const projectName = project.name;
                        const projectId = project.id;
                        returnData.push({
                            name: projectName,
                            value: projectId,
                        });
                    }
                    return { results: (0, GenericFunctions_1.filterSortSearchListItems)(returnData, filter) };
                },
                async getIssueTypes() {
                    const projectId = this.getCurrentNodeParameter('project', { extractValue: true });
                    const returnData = [];
                    const { issueTypes } = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/2/project/${projectId}`, 'GET');
                    for (const issueType of issueTypes) {
                        const issueTypeName = issueType.name;
                        const issueTypeId = issueType.id;
                        returnData.push({
                            name: issueTypeName,
                            value: issueTypeId,
                        });
                    }
                    returnData.sort((a, b) => {
                        if (a.name < b.name) {
                            return -1;
                        }
                        if (a.name > b.name) {
                            return 1;
                        }
                        return 0;
                    });
                    return { results: returnData };
                },
                async getUsers(filter) {
                    const jiraVersion = this.getCurrentNodeParameter('jiraVersion');
                    const query = {};
                    let endpoint = '/api/2/users/search';
                    if (jiraVersion === 'server') {
                        endpoint = '/api/2/user/search';
                        query.username = "'";
                    }
                    const users = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, endpoint, 'GET', {}, query);
                    const returnData = users.reduce((activeUsers, user) => {
                        var _a;
                        if (user.active) {
                            activeUsers.push({
                                name: user.displayName,
                                value: ((_a = user.accountId) !== null && _a !== void 0 ? _a : user.name),
                            });
                        }
                        return activeUsers;
                    }, []);
                    return { results: (0, GenericFunctions_1.filterSortSearchListItems)(returnData, filter) };
                },
                async getPriorities() {
                    const returnData = [];
                    const priorities = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, '/api/2/priority', 'GET');
                    for (const priority of priorities) {
                        const priorityName = priority.name;
                        const priorityId = priority.id;
                        returnData.push({
                            name: priorityName,
                            value: priorityId,
                        });
                    }
                    returnData.sort((a, b) => {
                        if (a.name < b.name) {
                            return -1;
                        }
                        if (a.name > b.name) {
                            return 1;
                        }
                        return 0;
                    });
                    return { results: returnData };
                },
                async getTransitions() {
                    const returnData = [];
                    const issueKey = this.getCurrentNodeParameter('issueKey');
                    const transitions = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/2/issue/${issueKey}/transitions`, 'GET');
                    for (const transition of transitions.transitions) {
                        returnData.push({
                            name: transition.name,
                            value: transition.id,
                        });
                    }
                    returnData.sort((a, b) => {
                        if (a.name < b.name) {
                            return -1;
                        }
                        if (a.name > b.name) {
                            return 1;
                        }
                        return 0;
                    });
                    return { results: returnData };
                },
                async getCustomFields() {
                    const returnData = [];
                    const operation = this.getCurrentNodeParameter('operation');
                    let projectId;
                    let issueTypeId;
                    if (operation === 'create') {
                        projectId = this.getCurrentNodeParameter('project', { extractValue: true });
                        issueTypeId = this.getCurrentNodeParameter('issueType', { extractValue: true });
                    }
                    else {
                        const issueKey = this.getCurrentNodeParameter('issueKey');
                        const res = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/2/issue/${issueKey}`, 'GET', {}, {});
                        projectId = res.fields.project.id;
                        issueTypeId = res.fields.issuetype.id;
                    }
                    const res = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/2/issue/createmeta?projectIds=${projectId}&issueTypeIds=${issueTypeId}&expand=projects.issuetypes.fields`, 'GET');
                    const fields = res.projects
                        .find((o) => o.id === projectId)
                        .issuetypes.find((o) => o.id === issueTypeId).fields;
                    for (const key of Object.keys(fields)) {
                        const field = fields[key];
                        if (field.schema && Object.keys(field.schema).includes('customId')) {
                            returnData.push({
                                name: field.name,
                                value: field.key || field.fieldId,
                            });
                        }
                    }
                    return { results: returnData };
                },
            },
            loadOptions: {
                async getLabels() {
                    const returnData = [];
                    const labels = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, '/api/2/label', 'GET');
                    for (const label of labels.values) {
                        const labelName = label;
                        const labelId = label;
                        returnData.push({
                            name: labelName,
                            value: labelId,
                        });
                    }
                    returnData.sort((a, b) => {
                        if (a.name < b.name) {
                            return -1;
                        }
                        if (a.name > b.name) {
                            return 1;
                        }
                        return 0;
                    });
                    return returnData;
                },
                async getUsers() {
                    const jiraVersion = this.getCurrentNodeParameter('jiraVersion');
                    const query = {};
                    let endpoint = '/api/2/users/search';
                    if (jiraVersion === 'server') {
                        endpoint = '/api/2/user/search';
                        query.username = "'";
                    }
                    const users = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, endpoint, 'GET', {}, query);
                    return users
                        .reduce((activeUsers, user) => {
                        if (user.active) {
                            activeUsers.push({
                                name: user.displayName,
                                value: (user.accountId || user.name),
                            });
                        }
                        return activeUsers;
                    }, [])
                        .sort((a, b) => {
                        return a.name.toLowerCase() > b.name.toLowerCase() ? 1 : -1;
                    });
                },
                async getGroups() {
                    const returnData = [];
                    const groups = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, '/api/2/groups/picker', 'GET');
                    for (const group of groups.groups) {
                        const groupName = group.name;
                        const groupId = group.name;
                        returnData.push({
                            name: groupName,
                            value: groupId,
                        });
                    }
                    returnData.sort((a, b) => {
                        if (a.name < b.name) {
                            return -1;
                        }
                        if (a.name > b.name) {
                            return 1;
                        }
                        return 0;
                    });
                    return returnData;
                },
                async getProjectComponents() {
                    const returnData = [];
                    const project = this.getCurrentNodeParameter('project', { extractValue: true });
                    const { values: components } = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/2/project/${project}/component`, 'GET');
                    for (const component of components) {
                        returnData.push({
                            name: component.name,
                            value: component.id,
                        });
                    }
                    returnData.sort((a, b) => {
                        if (a.name < b.name) {
                            return -1;
                        }
                        if (a.name > b.name) {
                            return 1;
                        }
                        return 0;
                    });
                    return returnData;
                },
            },
        };
    }
    async execute() {
        const items = this.getInputData();
        const returnData = [];
        const length = items.length;
        let responseData;
        const qs = {};
        const resource = this.getNodeParameter('resource', 0);
        const operation = this.getNodeParameter('operation', 0);
        const jiraVersion = this.getNodeParameter('jiraVersion', 0);
        if (resource === 'issue') {
            if (operation === 'create') {
                for (let i = 0; i < length; i++) {
                    const summary = this.getNodeParameter('summary', i);
                    const projectId = this.getNodeParameter('project', i, '', {
                        extractValue: true,
                    });
                    const issueTypeId = this.getNodeParameter('issueType', i, '', {
                        extractValue: true,
                    });
                    const additionalFields = this.getNodeParameter('additionalFields', i);
                    const assignee = this.getNodeParameter('additionalFields.assignee', i, '', {
                        extractValue: true,
                    });
                    if (assignee)
                        additionalFields.assignee = assignee;
                    const reporter = this.getNodeParameter('additionalFields.reporter', i, '', {
                        extractValue: true,
                    });
                    if (reporter)
                        additionalFields.reporter = reporter;
                    const priority = this.getNodeParameter('additionalFields.priority', i, '', {
                        extractValue: true,
                    });
                    if (priority)
                        additionalFields.priority = priority;
                    const body = {};
                    const fields = {
                        summary,
                        project: {
                            id: projectId,
                        },
                        issuetype: {
                            id: issueTypeId,
                        },
                    };
                    if (additionalFields.labels) {
                        fields.labels = additionalFields.labels;
                    }
                    if (additionalFields.serverLabels) {
                        fields.labels = additionalFields.serverLabels;
                    }
                    if (additionalFields.priority) {
                        fields.priority = {
                            id: additionalFields.priority,
                        };
                    }
                    if (additionalFields.assignee) {
                        if (jiraVersion === 'server') {
                            fields.assignee = {
                                name: additionalFields.assignee,
                            };
                        }
                        else {
                            fields.assignee = {
                                id: additionalFields.assignee,
                            };
                        }
                    }
                    if (additionalFields.reporter) {
                        if (jiraVersion === 'server') {
                            fields.reporter = {
                                name: additionalFields.reporter,
                            };
                        }
                        else {
                            fields.reporter = {
                                id: additionalFields.reporter,
                            };
                        }
                    }
                    if (additionalFields.description) {
                        fields.description = additionalFields.description;
                    }
                    if (additionalFields.updateHistory) {
                        qs.updateHistory = additionalFields.updateHistory;
                    }
                    if (additionalFields.componentIds) {
                        fields.components = additionalFields.componentIds.map((id) => ({ id }));
                    }
                    if (additionalFields.customFieldsUi) {
                        const customFields = additionalFields.customFieldsUi
                            .customFieldsValues;
                        if (customFields) {
                            customFields.forEach((cf) => {
                                if (typeof cf.fieldId !== 'string') {
                                    cf.fieldId = cf.fieldId.value.trim();
                                }
                            });
                            const data = customFields.reduce((obj, value) => Object.assign(obj, { [`${value.fieldId}`]: value.fieldValue }), {});
                            Object.assign(fields, data);
                        }
                    }
                    const issueTypes = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, '/api/2/issuetype', 'GET', body, qs);
                    const subtaskIssues = [];
                    for (const issueType of issueTypes) {
                        if (issueType.subtask) {
                            subtaskIssues.push(issueType.id);
                        }
                    }
                    if (!additionalFields.parentIssueKey && subtaskIssues.includes(issueTypeId)) {
                        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'You must define a Parent Issue Key when Issue type is sub-task', { itemIndex: i });
                    }
                    else if (additionalFields.parentIssueKey && subtaskIssues.includes(issueTypeId)) {
                        fields.parent = {
                            key: additionalFields.parentIssueKey.toUpperCase(),
                        };
                    }
                    body.fields = fields;
                    responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, '/api/2/issue', 'POST', body);
                    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                    returnData.push(...executionData);
                }
            }
            if (operation === 'update') {
                for (let i = 0; i < length; i++) {
                    const issueKey = this.getNodeParameter('issueKey', i);
                    const updateFields = this.getNodeParameter('updateFields', i);
                    const assignee = this.getNodeParameter('updateFields.assignee', i, '', {
                        extractValue: true,
                    });
                    if (assignee)
                        updateFields.assignee = assignee;
                    const reporter = this.getNodeParameter('updateFields.reporter', i, '', {
                        extractValue: true,
                    });
                    if (reporter)
                        updateFields.reporter = reporter;
                    const priority = this.getNodeParameter('updateFields.priority', i, '', {
                        extractValue: true,
                    });
                    if (priority)
                        updateFields.priority = priority;
                    const statusId = this.getNodeParameter('updateFields.statusId', i, '', {
                        extractValue: true,
                    });
                    if (statusId)
                        updateFields.statusId = statusId;
                    const body = {};
                    const fields = {};
                    if (updateFields.summary) {
                        fields.summary = updateFields.summary;
                    }
                    if (updateFields.issueType) {
                        fields.issuetype = {
                            id: updateFields.issueType,
                        };
                    }
                    if (updateFields.labels) {
                        fields.labels = updateFields.labels;
                    }
                    if (updateFields.serverLabels) {
                        fields.labels = updateFields.serverLabels;
                    }
                    if (updateFields.priority) {
                        fields.priority = {
                            id: updateFields.priority,
                        };
                    }
                    if (updateFields.assignee) {
                        if (jiraVersion === 'server') {
                            fields.assignee = {
                                name: updateFields.assignee,
                            };
                        }
                        else {
                            fields.assignee = {
                                id: updateFields.assignee,
                            };
                        }
                    }
                    if (updateFields.reporter) {
                        if (jiraVersion === 'server') {
                            fields.reporter = {
                                name: updateFields.reporter,
                            };
                        }
                        else {
                            fields.reporter = {
                                id: updateFields.reporter,
                            };
                        }
                    }
                    if (updateFields.description) {
                        fields.description = updateFields.description;
                    }
                    if (updateFields.customFieldsUi) {
                        const customFields = updateFields.customFieldsUi
                            .customFieldsValues;
                        if (customFields) {
                            customFields.forEach((cf) => {
                                if (typeof cf.fieldId !== 'string') {
                                    cf.fieldId = cf.fieldId.value.trim();
                                }
                            });
                            const data = customFields.reduce((obj, value) => Object.assign(obj, { [`${value.fieldId}`]: value.fieldValue }), {});
                            Object.assign(fields, data);
                        }
                    }
                    const issueTypes = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, '/api/2/issuetype', 'GET', body);
                    const subtaskIssues = [];
                    for (const issueType of issueTypes) {
                        if (issueType.subtask) {
                            subtaskIssues.push(issueType.id);
                        }
                    }
                    if (!updateFields.parentIssueKey && subtaskIssues.includes(updateFields.issueType)) {
                        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'You must define a Parent Issue Key when Issue type is sub-task', { itemIndex: i });
                    }
                    else if (updateFields.parentIssueKey &&
                        subtaskIssues.includes(updateFields.issueType)) {
                        fields.parent = {
                            key: updateFields.parentIssueKey.toUpperCase(),
                        };
                    }
                    body.fields = fields;
                    if (updateFields.statusId) {
                        responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/2/issue/${issueKey}/transitions`, 'POST', { transition: { id: updateFields.statusId } });
                    }
                    responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/2/issue/${issueKey}`, 'PUT', body);
                    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray({ success: true }), { itemData: { item: i } });
                    returnData.push(...executionData);
                }
            }
            if (operation === 'get') {
                for (let i = 0; i < length; i++) {
                    const issueKey = this.getNodeParameter('issueKey', i);
                    const simplifyOutput = this.getNodeParameter('simplifyOutput', i);
                    const additionalFields = this.getNodeParameter('additionalFields', i);
                    if (additionalFields.fields) {
                        qs.fields = additionalFields.fields;
                    }
                    if (additionalFields.fieldsByKey) {
                        qs.fieldsByKey = additionalFields.fieldsByKey;
                    }
                    if (additionalFields.expand) {
                        qs.expand = additionalFields.expand;
                    }
                    if (simplifyOutput) {
                        qs.expand = `${qs.expand || ''},names`;
                    }
                    if (additionalFields.properties) {
                        qs.properties = additionalFields.properties;
                    }
                    if (additionalFields.updateHistory) {
                        qs.updateHistory = additionalFields.updateHistory;
                    }
                    responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/2/issue/${issueKey}`, 'GET', {}, qs);
                    if (simplifyOutput) {
                        qs.expand = qs.expand || '';
                        if (qs.expand.toLowerCase().indexOf('renderedfields') !== -1 &&
                            responseData.renderedFields &&
                            Object.keys(responseData.renderedFields).length) {
                            responseData.fields = (0, lodash_mergewith_1.default)(responseData.fields, responseData.renderedFields, (a, b) => (b === null ? a : b));
                        }
                        const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray((0, GenericFunctions_1.simplifyIssueOutput)(responseData)), { itemData: { item: i } });
                        returnData.push(...executionData);
                    }
                    else {
                        const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                        returnData.push(...executionData);
                    }
                }
            }
            if (operation === 'getAll') {
                for (let i = 0; i < length; i++) {
                    const returnAll = this.getNodeParameter('returnAll', i);
                    const options = this.getNodeParameter('options', i);
                    const body = {};
                    if (options.fields) {
                        body.fields = options.fields.split(',');
                    }
                    if (options.jql) {
                        body.jql = options.jql;
                    }
                    if (options.expand) {
                        if (typeof options.expand === 'string') {
                            body.expand = options.expand.split(',');
                        }
                        else {
                            body.expand = options.expand;
                        }
                    }
                    if (returnAll) {
                        responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequestAllItems.call(this, 'issues', '/api/2/search', 'POST', body);
                    }
                    else {
                        const limit = this.getNodeParameter('limit', i);
                        body.maxResults = limit;
                        responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, '/api/2/search', 'POST', body);
                        responseData = responseData.issues;
                    }
                    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                    returnData.push(...executionData);
                }
            }
            if (operation === 'changelog') {
                for (let i = 0; i < length; i++) {
                    const issueKey = this.getNodeParameter('issueKey', i);
                    const returnAll = this.getNodeParameter('returnAll', i);
                    if (returnAll) {
                        responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequestAllItems.call(this, 'values', `/api/2/issue/${issueKey}/changelog`, 'GET');
                    }
                    else {
                        qs.maxResults = this.getNodeParameter('limit', i);
                        responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/2/issue/${issueKey}/changelog`, 'GET', {}, qs);
                        responseData = responseData.values;
                    }
                    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                    returnData.push(...executionData);
                }
            }
            if (operation === 'notify') {
                for (let i = 0; i < length; i++) {
                    const issueKey = this.getNodeParameter('issueKey', i);
                    const additionalFields = this.getNodeParameter('additionalFields', i);
                    const jsonActive = this.getNodeParameter('jsonParameters', 0);
                    const body = {};
                    if (additionalFields.textBody) {
                        body.textBody = additionalFields.textBody;
                    }
                    if (additionalFields.htmlBody) {
                        body.htmlBody = additionalFields.htmlBody;
                    }
                    if (!jsonActive) {
                        const notificationRecipientsValues = this.getNodeParameter('notificationRecipientsUi', i).notificationRecipientsValues;
                        const notificationRecipients = {};
                        if (notificationRecipientsValues) {
                            if (notificationRecipientsValues.reporter) {
                                notificationRecipients.reporter = notificationRecipientsValues.reporter;
                            }
                            if (notificationRecipientsValues.assignee) {
                                notificationRecipients.assignee = notificationRecipientsValues.assignee;
                            }
                            if (notificationRecipientsValues.assignee) {
                                notificationRecipients.watchers = notificationRecipientsValues.watchers;
                            }
                            if (notificationRecipientsValues.voters) {
                                notificationRecipients.watchers = notificationRecipientsValues.voters;
                            }
                            if ((notificationRecipientsValues.users || []).length > 0) {
                                notificationRecipients.users = notificationRecipientsValues.users.map((user) => {
                                    return {
                                        accountId: user,
                                    };
                                });
                            }
                            if ((notificationRecipientsValues.groups || []).length > 0) {
                                notificationRecipients.groups = notificationRecipientsValues.groups.map((group) => {
                                    return {
                                        name: group,
                                    };
                                });
                            }
                        }
                        body.to = notificationRecipients;
                        const notificationRecipientsRestrictionsValues = this.getNodeParameter('notificationRecipientsRestrictionsUi', i).notificationRecipientsRestrictionsValues;
                        const notificationRecipientsRestrictions = {};
                        if (notificationRecipientsRestrictionsValues) {
                            if ((notificationRecipientsRestrictionsValues.groups || []).length >
                                0) {
                                notificationRecipientsRestrictions.groups = notificationRecipientsRestrictionsValues.groups.map((group) => {
                                    return {
                                        name: group,
                                    };
                                });
                            }
                        }
                        body.restrict = notificationRecipientsRestrictions;
                    }
                    else {
                        const notificationRecipientsJson = (0, GenericFunctions_1.validateJSON)(this.getNodeParameter('notificationRecipientsJson', i));
                        if (notificationRecipientsJson) {
                            body.to = notificationRecipientsJson;
                        }
                        const notificationRecipientsRestrictionsJson = (0, GenericFunctions_1.validateJSON)(this.getNodeParameter('notificationRecipientsRestrictionsJson', i));
                        if (notificationRecipientsRestrictionsJson) {
                            body.restrict = notificationRecipientsRestrictionsJson;
                        }
                    }
                    responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/2/issue/${issueKey}/notify`, 'POST', body, qs);
                    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                    returnData.push(...executionData);
                }
            }
            if (operation === 'transitions') {
                for (let i = 0; i < length; i++) {
                    const issueKey = this.getNodeParameter('issueKey', i);
                    const additionalFields = this.getNodeParameter('additionalFields', i);
                    if (additionalFields.transitionId) {
                        qs.transitionId = additionalFields.transitionId;
                    }
                    if (additionalFields.expand) {
                        qs.expand = additionalFields.expand;
                    }
                    if (additionalFields.skipRemoteOnlyCondition) {
                        qs.skipRemoteOnlyCondition = additionalFields.skipRemoteOnlyCondition;
                    }
                    responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/2/issue/${issueKey}/transitions`, 'GET', {}, qs);
                    responseData = responseData.transitions;
                    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                    returnData.push(...executionData);
                }
            }
            if (operation === 'delete') {
                for (let i = 0; i < length; i++) {
                    const issueKey = this.getNodeParameter('issueKey', i);
                    const deleteSubtasks = this.getNodeParameter('deleteSubtasks', i);
                    qs.deleteSubtasks = deleteSubtasks;
                    responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/2/issue/${issueKey}`, 'DELETE', {}, qs);
                    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray({ success: true }), { itemData: { item: i } });
                    returnData.push(...executionData);
                }
            }
        }
        if (resource === 'issueAttachment') {
            const apiVersion = jiraVersion === 'server' ? '2' : '3';
            if (operation === 'add') {
                for (let i = 0; i < length; i++) {
                    const binaryPropertyName = this.getNodeParameter('binaryPropertyName', i);
                    const issueKey = this.getNodeParameter('issueKey', i);
                    const binaryData = this.helpers.assertBinaryData(i, binaryPropertyName);
                    let uploadData;
                    if (binaryData.id) {
                        uploadData = this.helpers.getBinaryStream(binaryData.id);
                    }
                    else {
                        uploadData = Buffer.from(binaryData.data, n8n_workflow_1.BINARY_ENCODING);
                    }
                    responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/${apiVersion}/issue/${issueKey}/attachments`, 'POST', {}, {}, undefined, {
                        formData: {
                            file: {
                                value: uploadData,
                                options: {
                                    filename: binaryData.fileName,
                                },
                            },
                        },
                    });
                    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                    returnData.push(...executionData);
                }
            }
            if (operation === 'remove') {
                for (let i = 0; i < length; i++) {
                    const attachmentId = this.getNodeParameter('attachmentId', i);
                    responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/${apiVersion}/attachment/${attachmentId}`, 'DELETE', {}, qs);
                    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray({ success: true }), { itemData: { item: i } });
                    returnData.push(...executionData);
                }
            }
            if (operation === 'get') {
                const download = this.getNodeParameter('download', 0);
                for (let i = 0; i < length; i++) {
                    const attachmentId = this.getNodeParameter('attachmentId', i);
                    responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/${apiVersion}/attachment/${attachmentId}`, 'GET', {}, qs);
                    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                    returnData.push(...executionData);
                }
                if (download) {
                    const binaryPropertyName = this.getNodeParameter('binaryProperty', 0);
                    for (const [index, attachment] of returnData.entries()) {
                        returnData[index].binary = {};
                        const buffer = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, '', 'GET', {}, {}, attachment === null || attachment === void 0 ? void 0 : attachment.json.content, { json: false, encoding: null, useStream: true });
                        returnData[index].binary[binaryPropertyName] =
                            await this.helpers.prepareBinaryData(buffer, attachment.json.filename, attachment.json.mimeType);
                    }
                }
            }
            if (operation === 'getAll') {
                const download = this.getNodeParameter('download', 0);
                for (let i = 0; i < length; i++) {
                    const issueKey = this.getNodeParameter('issueKey', i);
                    const returnAll = this.getNodeParameter('returnAll', i);
                    const { fields: { attachment }, } = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/2/issue/${issueKey}`, 'GET', {}, qs);
                    responseData = attachment;
                    if (!returnAll) {
                        const limit = this.getNodeParameter('limit', i);
                        responseData = responseData.slice(0, limit);
                    }
                    responseData = responseData.map((data) => ({ json: data }));
                    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                    returnData.push(...executionData);
                }
                if (download) {
                    const binaryPropertyName = this.getNodeParameter('binaryProperty', 0);
                    for (const [index, attachment] of returnData.entries()) {
                        returnData[index].binary = {};
                        const buffer = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, '', 'GET', {}, {}, attachment.json.content, { json: false, encoding: null, useStream: true });
                        returnData[index].binary[binaryPropertyName] =
                            await this.helpers.prepareBinaryData(buffer, attachment.json.filename, attachment.json.mimeType);
                    }
                }
            }
        }
        if (resource === 'issueComment') {
            const apiVersion = jiraVersion === 'server' ? '2' : '3';
            if (operation === 'add') {
                for (let i = 0; i < length; i++) {
                    const jsonParameters = this.getNodeParameter('jsonParameters', 0);
                    const issueKey = this.getNodeParameter('issueKey', i);
                    const options = this.getNodeParameter('options', i);
                    const body = {};
                    if (options.expand) {
                        qs.expand = options.expand;
                        delete options.expand;
                    }
                    Object.assign(body, options);
                    if (!jsonParameters) {
                        const comment = this.getNodeParameter('comment', i);
                        if (jiraVersion === 'server') {
                            Object.assign(body, { body: comment });
                        }
                        else {
                            Object.assign(body, {
                                body: {
                                    type: 'doc',
                                    version: 1,
                                    content: [
                                        {
                                            type: 'paragraph',
                                            content: [
                                                {
                                                    type: 'text',
                                                    text: comment,
                                                },
                                            ],
                                        },
                                    ],
                                },
                            });
                        }
                    }
                    else {
                        const commentJson = this.getNodeParameter('commentJson', i);
                        const json = (0, GenericFunctions_1.validateJSON)(commentJson);
                        if (json === '') {
                            throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Document Format must be a valid JSON', {
                                itemIndex: i,
                            });
                        }
                        Object.assign(body, { body: json });
                    }
                    responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/${apiVersion}/issue/${issueKey}/comment`, 'POST', body, qs);
                    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                    returnData.push(...executionData);
                }
            }
            if (operation === 'get') {
                for (let i = 0; i < length; i++) {
                    const issueKey = this.getNodeParameter('issueKey', i);
                    const commentId = this.getNodeParameter('commentId', i);
                    const options = this.getNodeParameter('options', i);
                    Object.assign(qs, options);
                    responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/${apiVersion}/issue/${issueKey}/comment/${commentId}`, 'GET', {}, qs);
                    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                    returnData.push(...executionData);
                }
            }
            if (operation === 'getAll') {
                for (let i = 0; i < length; i++) {
                    const issueKey = this.getNodeParameter('issueKey', i);
                    const returnAll = this.getNodeParameter('returnAll', i);
                    const options = this.getNodeParameter('options', i);
                    const body = {};
                    Object.assign(qs, options);
                    if (returnAll) {
                        responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequestAllItems.call(this, 'comments', `/api/${apiVersion}/issue/${issueKey}/comment`, 'GET', body, qs);
                    }
                    else {
                        const limit = this.getNodeParameter('limit', i);
                        body.maxResults = limit;
                        responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/${apiVersion}/issue/${issueKey}/comment`, 'GET', body, qs);
                        responseData = responseData.comments;
                    }
                    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                    returnData.push(...executionData);
                }
            }
            if (operation === 'remove') {
                for (let i = 0; i < length; i++) {
                    const issueKey = this.getNodeParameter('issueKey', i);
                    const commentId = this.getNodeParameter('commentId', i);
                    responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/${apiVersion}/issue/${issueKey}/comment/${commentId}`, 'DELETE', {}, qs);
                    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray({ success: true }), { itemData: { item: i } });
                    returnData.push(...executionData);
                }
            }
            if (operation === 'update') {
                for (let i = 0; i < length; i++) {
                    const issueKey = this.getNodeParameter('issueKey', i);
                    const commentId = this.getNodeParameter('commentId', i);
                    const options = this.getNodeParameter('options', i);
                    const jsonParameters = this.getNodeParameter('jsonParameters', 0);
                    const body = {};
                    if (options.expand) {
                        qs.expand = options.expand;
                        delete options.expand;
                    }
                    Object.assign(qs, options);
                    if (!jsonParameters) {
                        const comment = this.getNodeParameter('comment', i);
                        if (jiraVersion === 'server') {
                            Object.assign(body, { body: comment });
                        }
                        else {
                            Object.assign(body, {
                                body: {
                                    type: 'doc',
                                    version: 1,
                                    content: [
                                        {
                                            type: 'paragraph',
                                            content: [
                                                {
                                                    type: 'text',
                                                    text: comment,
                                                },
                                            ],
                                        },
                                    ],
                                },
                            });
                        }
                    }
                    else {
                        const commentJson = this.getNodeParameter('commentJson', i);
                        const json = (0, GenericFunctions_1.validateJSON)(commentJson);
                        if (json === '') {
                            throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Document Format must be a valid JSON', {
                                itemIndex: i,
                            });
                        }
                        Object.assign(body, { body: json });
                    }
                    responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/${apiVersion}/issue/${issueKey}/comment/${commentId}`, 'PUT', body, qs);
                    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                    returnData.push(...executionData);
                }
            }
        }
        if (resource === 'user') {
            const apiVersion = jiraVersion === 'server' ? '2' : '3';
            if (operation === 'create') {
                for (let i = 0; i < length; i++) {
                    const body = {
                        name: this.getNodeParameter('username', i),
                        emailAddress: this.getNodeParameter('emailAddress', i),
                        displayName: this.getNodeParameter('displayName', i),
                    };
                    const additionalFields = this.getNodeParameter('additionalFields', i);
                    Object.assign(body, additionalFields);
                    responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/${apiVersion}/user`, 'POST', body, {});
                    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                    returnData.push(...executionData);
                }
            }
            else if (operation === 'delete') {
                for (let i = 0; i < length; i++) {
                    qs.accountId = this.getNodeParameter('accountId', i);
                    responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/${apiVersion}/user`, 'DELETE', {}, qs);
                    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray({ success: true }), { itemData: { item: i } });
                    returnData.push(...executionData);
                }
            }
            else if (operation === 'get') {
                for (let i = 0; i < length; i++) {
                    qs.accountId = this.getNodeParameter('accountId', i);
                    const { expand } = this.getNodeParameter('additionalFields', i);
                    if (expand) {
                        qs.expand = expand.join(',');
                    }
                    responseData = await GenericFunctions_1.jiraSoftwareCloudApiRequest.call(this, `/api/${apiVersion}/user`, 'GET', {}, qs);
                    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                    returnData.push(...executionData);
                }
            }
        }
        return this.prepareOutputData(returnData);
    }
}
exports.Jira = Jira;
