// n8n-nodes-base/dist/nodes/SyncroMSP/v1/SyncroMspV1.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SyncroMspV1 = void 0;
const versionDescription_1 = require("./actions/versionDescription");
const methods_1 = require("./methods");
const router_1 = require("./actions/router");
const transport_1 = require("./transport");
class SyncroMspV1 {
    constructor(baseDescription) {
        this.methods = {
            loadOptions: methods_1.loadOptions,
            credentialTest: {
                async syncroMspApiCredentialTest(credential) {
                    try {
                        await transport_1.validateCredentials.call(this, credential.data);
                    }
                    catch (error) {
                        if (error.statusCode === 401) {
                            return {
                                status: 'Error',
                                message: 'The API Key included in the request is invalid',
                            };
                        }
                    }
                    return {
                        status: 'OK',
                        message: 'Connection successful!',
                    };
                },
            },
        };
        this.description = {
            ...baseDescription,
            ...versionDescription_1.versionDescription,
        };
    }
    async execute() {
        return router_1.router.call(this);
    }
}
exports.SyncroMspV1 = SyncroMspV1;
