// n8n-nodes-base/dist/nodes/Google/Sheet/v2/actions/sheet/read.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const GoogleSheets_utils_1 = require("../../helpers/GoogleSheets.utils");
const commonDescription_1 = require("./commonDescription");
exports.description = [
    {
        displayName: 'Filters',
        name: 'filtersUI',
        placeholder: 'Add Filter',
        type: 'fixedCollection',
        typeOptions: {
            multipleValueButtonText: 'Add Filter',
            multipleValues: true,
        },
        default: {},
        options: [
            {
                displayName: 'Filter',
                name: 'values',
                values: [
                    {
                        displayName: 'Column',
                        name: 'lookupColumn',
                        type: 'options',
                        typeOptions: {
                            loadOptionsDependsOn: ['sheetName.value'],
                            loadOptionsMethod: 'getSheetHeaderRowWithGeneratedColumnNames',
                        },
                        default: '',
                        description: 'Choose from the list, or specify an ID using an <a href="https://docs.n8n.io/code-examples/expressions/">expression</a>',
                    },
                    {
                        displayName: 'Value',
                        name: 'lookupValue',
                        type: 'string',
                        default: '',
                        hint: 'The column must have this value to be matched',
                    },
                ],
            },
        ],
        displayOptions: {
            show: {
                resource: ['sheet'],
                operation: ['read'],
            },
            hide: {
                ...GoogleSheets_utils_1.untilSheetSelected,
            },
        },
    },
    {
        displayName: 'Options',
        name: 'options',
        type: 'collection',
        placeholder: 'Add Option',
        default: {},
        displayOptions: {
            show: {
                resource: ['sheet'],
                operation: ['read'],
            },
            hide: {
                ...GoogleSheets_utils_1.untilSheetSelected,
            },
        },
        options: [
            ...commonDescription_1.dataLocationOnSheet,
            ...commonDescription_1.outputFormatting,
            {
                displayName: 'When Filter Has Multiple Matches',
                name: 'returnAllMatches',
                type: 'options',
                default: 'returnFirstMatch',
                options: [
                    {
                        name: 'Return First Match',
                        value: 'returnFirstMatch',
                        description: 'Return only the first match',
                    },
                    {
                        name: 'Return All Matches',
                        value: 'returnAllMatches',
                        description: 'Return all values that match',
                    },
                ],
                description: 'By default only the first result gets returned, Set to "Return All Matches" to get multiple matches',
            },
        ],
    },
];
async function execute(sheet, sheetName) {
    var _a, _b;
    const options = this.getNodeParameter('options', 0, {});
    const outputFormattingOption = ((_a = options.outputFormatting) === null || _a === void 0 ? void 0 : _a.values) || {};
    const dataLocationOnSheetOptions = ((_b = options.dataLocationOnSheet) === null || _b === void 0 ? void 0 : _b.values) || {};
    if (dataLocationOnSheetOptions.rangeDefinition === undefined) {
        dataLocationOnSheetOptions.rangeDefinition = 'detectAutomatically';
    }
    const range = (0, GoogleSheets_utils_1.getRangeString)(sheetName, dataLocationOnSheetOptions);
    const valueRenderMode = (outputFormattingOption.general ||
        'UNFORMATTED_VALUE');
    const dateTimeRenderOption = (outputFormattingOption.date || 'FORMATTED_STRING');
    const sheetData = (await sheet.getData(range, valueRenderMode, dateTimeRenderOption));
    if (sheetData === undefined || sheetData.length === 0) {
        return [];
    }
    const { data, headerRow, firstDataRow } = (0, GoogleSheets_utils_1.prepareSheetData)(sheetData, dataLocationOnSheetOptions);
    let returnData = [];
    const lookupValues = this.getNodeParameter('filtersUI.values', 0, []);
    if (lookupValues.length) {
        const returnAllMatches = options.returnAllMatches === 'returnAllMatches' ? true : false;
        const items = this.getInputData();
        for (let i = 1; i < items.length; i++) {
            const itemLookupValues = this.getNodeParameter('filtersUI.values', i, []);
            if (itemLookupValues.length) {
                lookupValues.push(...itemLookupValues);
            }
        }
        returnData = await sheet.lookupValues(data, headerRow, firstDataRow, lookupValues, returnAllMatches);
    }
    else {
        returnData = sheet.structureArrayDataByColumn(data, headerRow, firstDataRow);
    }
    return this.helpers.returnJsonArray(returnData);
}
exports.execute = execute;
