// n8n-nodes-base/dist/nodes/Google/Sheet/v2/actions/spreadsheet/delete.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const transport_1 = require("../../transport");
exports.description = [
    {
        displayName: 'Document',
        name: 'documentId',
        type: 'resourceLocator',
        default: { mode: 'list', value: '' },
        required: true,
        modes: [
            {
                displayName: 'From List',
                name: 'list',
                type: 'list',
                typeOptions: {
                    searchListMethod: 'spreadSheetsSearch',
                    searchable: true,
                },
            },
            {
                displayName: 'By URL',
                name: 'url',
                type: 'string',
                extractValue: {
                    type: 'regex',
                    regex: 'https:\\/\\/(?:drive|docs)\\.google\\.com\\/\\w+\\/d\\/([0-9a-zA-Z\\-_]+)(?:\\/.*|)',
                },
                validation: [
                    {
                        type: 'regex',
                        properties: {
                            regex: 'https:\\/\\/(?:drive|docs)\\.google.com\\/\\w+\\/d\\/([0-9a-zA-Z\\-_]+)(?:\\/.*|)',
                            errorMessage: 'Not a valid Google Drive File URL',
                        },
                    },
                ],
            },
            {
                displayName: 'By ID',
                name: 'id',
                type: 'string',
                validation: [
                    {
                        type: 'regex',
                        properties: {
                            regex: '[a-zA-Z0-9\\-_]{2,}',
                            errorMessage: 'Not a valid Google Drive File ID',
                        },
                    },
                ],
                url: '=https://docs.google.com/spreadsheets/d/{{$value}}/edit',
            },
        ],
        displayOptions: {
            show: {
                resource: ['spreadsheet'],
                operation: ['deleteSpreadsheet'],
            },
        },
    },
];
async function execute() {
    const items = this.getInputData();
    const returnData = [];
    for (let i = 0; i < items.length; i++) {
        const documentId = this.getNodeParameter('documentId', i, undefined, {
            extractValue: true,
        });
        await transport_1.apiRequest.call(this, 'DELETE', '', {}, {}, `https://www.googleapis.com/drive/v3/files/${documentId}`);
        returnData.push({ success: true });
    }
    return this.helpers.returnJsonArray(returnData);
}
exports.execute = execute;
