// n8n-nodes-base/dist/nodes/Postgres/v2/actions/database/executeQuery.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const utilities_1 = require("../../../../../utils/utilities");
const utils_1 = require("../../helpers/utils");
const common_descriptions_1 = require("../common.descriptions");
const properties = [
    {
        displayName: 'Query',
        name: 'query',
        type: 'string',
        default: '',
        placeholder: 'e.g. SELECT id, name FROM product WHERE quantity > $1 AND price <= $2',
        required: true,
        description: "The SQL query to execute. You can use n8n expressions and $1, $2, $3, etc to refer to the 'Query Parameters' set in options below.",
        typeOptions: {
            editor: 'sqlEditor',
            sqlDialect: 'postgres',
        },
        hint: 'Prefer using query parameters over n8n expressions to avoid SQL injection attacks',
    },
    {
        displayName: `
		To use query parameters in your SQL query, reference them as $1, $2, $3, etc in the corresponding order. <a href="https://docs.n8n.io/integrations/builtin/app-nodes/n8n-nodes-base.postgres/#use-query-parameters">More info</a>.
		`,
        name: 'notice',
        type: 'notice',
        default: '',
    },
    common_descriptions_1.optionsCollection,
];
const displayOptions = {
    show: {
        resource: ['database'],
        operation: ['executeQuery'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(runQueries, items, nodeOptions, _db) {
    items = (0, utils_1.replaceEmptyStringsByNulls)(items, nodeOptions.replaceEmptyStrings);
    const queries = [];
    for (let i = 0; i < items.length; i++) {
        const query = this.getNodeParameter('query', i);
        let values = [];
        let queryReplacement = this.getNodeParameter('options.queryReplacement', i, '');
        if (typeof queryReplacement === 'string') {
            queryReplacement = queryReplacement.split(',').map((entry) => entry.trim());
        }
        if (Array.isArray(queryReplacement)) {
            values = queryReplacement;
        }
        else {
            throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Query Replacement must be a string of comma-separated values, or an array of values', { itemIndex: i });
        }
        queries.push({ query, values });
    }
    return runQueries(queries, items, nodeOptions);
}
exports.execute = execute;
