// n8n/dist/PublicApi/v1/handlers/workflows/workflows.service.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseTagNames = exports.getStartNode = exports.hasStartNode = exports.updateWorkflow = exports.getWorkflowsAndCount = exports.deleteWorkflow = exports.setWorkflowAsInactive = exports.setWorkflowAsActive = exports.createWorkflow = exports.getWorkflowIdsViaTags = exports.getWorkflowById = exports.getSharedWorkflows = exports.getSharedWorkflow = exports.getSharedWorkflowIds = void 0;
const typeorm_1 = require("typeorm");
const lodash_intersection_1 = __importDefault(require("lodash.intersection"));
const uuid_1 = require("uuid");
const Db = __importStar(require("../../../../Db"));
const WorkflowEntity_1 = require("../../../../databases/entities/WorkflowEntity");
const SharedWorkflow_1 = require("../../../../databases/entities/SharedWorkflow");
const users_service_1 = require("../users/users.service");
const config_1 = __importDefault(require("../../../../config"));
const constants_1 = require("../../../../constants");
function insertIf(condition, elements) {
    return condition ? elements : [];
}
async function getSharedWorkflowIds(user) {
    const where = user.globalRole.name === 'owner' ? {} : { userId: user.id };
    const sharedWorkflows = await Db.collections.SharedWorkflow.find({
        where,
        select: ['workflowId'],
    });
    return sharedWorkflows.map(({ workflowId }) => workflowId);
    return sharedWorkflows.map(({ workflowId }) => workflowId);
}
exports.getSharedWorkflowIds = getSharedWorkflowIds;
async function getSharedWorkflow(user, workflowId) {
    return Db.collections.SharedWorkflow.findOne({
        where: {
            ...(!(0, users_service_1.isInstanceOwner)(user) && { userId: user.id }),
            ...(workflowId && { workflowId }),
        },
        relations: [...insertIf(!config_1.default.getEnv('workflowTagsDisabled'), ['workflow.tags']), 'workflow'],
    });
}
exports.getSharedWorkflow = getSharedWorkflow;
async function getSharedWorkflows(user, options) {
    return Db.collections.SharedWorkflow.find({
        where: {
            ...(!(0, users_service_1.isInstanceOwner)(user) && { userId: user.id }),
            ...(options.workflowIds && { workflowId: (0, typeorm_1.In)(options.workflowIds) }),
        },
        ...(options.relations && { relations: options.relations }),
    });
}
exports.getSharedWorkflows = getSharedWorkflows;
async function getWorkflowById(id) {
    return Db.collections.Workflow.findOne({
        where: { id },
    });
}
exports.getWorkflowById = getWorkflowById;
async function getWorkflowIdsViaTags(tags) {
    const dbTags = await Db.collections.Tag.find({
        where: { name: (0, typeorm_1.In)(tags) },
        relations: ['workflows'],
    });
    const workflowIdsPerTag = dbTags.map((tag) => tag.workflows.map((workflow) => workflow.id));
    return (0, lodash_intersection_1.default)(...workflowIdsPerTag);
}
exports.getWorkflowIdsViaTags = getWorkflowIdsViaTags;
async function createWorkflow(workflow, user, role) {
    return Db.transaction(async (transactionManager) => {
        const newWorkflow = new WorkflowEntity_1.WorkflowEntity();
        Object.assign(newWorkflow, workflow);
        const savedWorkflow = await transactionManager.save(newWorkflow);
        const newSharedWorkflow = new SharedWorkflow_1.SharedWorkflow();
        Object.assign(newSharedWorkflow, {
            role,
            user,
            workflow: savedWorkflow,
        });
        await transactionManager.save(newSharedWorkflow);
        return savedWorkflow;
    });
}
exports.createWorkflow = createWorkflow;
async function setWorkflowAsActive(workflow) {
    return Db.collections.Workflow.update(workflow.id, { active: true, updatedAt: new Date() });
}
exports.setWorkflowAsActive = setWorkflowAsActive;
async function setWorkflowAsInactive(workflow) {
    return Db.collections.Workflow.update(workflow.id, { active: false, updatedAt: new Date() });
}
exports.setWorkflowAsInactive = setWorkflowAsInactive;
async function deleteWorkflow(workflow) {
    return Db.collections.Workflow.remove(workflow);
}
exports.deleteWorkflow = deleteWorkflow;
async function getWorkflowsAndCount(options) {
    return Db.collections.Workflow.findAndCount(options);
}
exports.getWorkflowsAndCount = getWorkflowsAndCount;
async function updateWorkflow(workflowId, updateData) {
    return Db.collections.Workflow.update(workflowId, updateData);
}
exports.updateWorkflow = updateWorkflow;
function hasStartNode(workflow) {
    if (!workflow.nodes.length)
        return false;
    const found = workflow.nodes.find((node) => constants_1.START_NODES.includes(node.type));
    return Boolean(found);
}
exports.hasStartNode = hasStartNode;
function getStartNode() {
    return {
        id: (0, uuid_1.v4)(),
        parameters: {},
        name: 'Start',
        type: 'n8n-nodes-base.start',
        typeVersion: 1,
        position: [240, 300],
    };
}
exports.getStartNode = getStartNode;
function parseTagNames(tags) {
    return tags.split(',').map((tag) => tag.trim());
}
exports.parseTagNames = parseTagNames;
