// n8n-nodes-base/dist/credentials/ActiveCampaignApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActiveCampaignApi = void 0;
class ActiveCampaignApi {
    constructor() {
        this.name = 'activeCampaignApi';
        this.displayName = 'ActiveCampaign API';
        this.documentationUrl = 'activeCampaign';
        this.properties = [
            {
                displayName: 'API URL',
                name: 'apiUrl',
                type: 'string',
                default: '',
            },
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'Api-Token': '={{$credentials.apiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{$credentials.apiUrl}}',
                url: '/api/3/fields',
            },
        };
    }
}
exports.ActiveCampaignApi = ActiveCampaignApi;
