// n8n-nodes-base/dist/credentials/AirtableApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AirtableApi = void 0;
class AirtableApi {
    constructor() {
        this.name = 'airtableApi';
        this.displayName = 'Airtable API';
        this.documentationUrl = 'airtable';
        this.properties = [
            {
                displayName: 'API Keys will be deprecated by the end of January 2024, see <a href="https://support.airtable.com/docs/airtable-api-key-deprecation-notice" target="_blank">this article</a> for more details. We recommend to use Personal Access Token instead.',
                name: 'deprecated',
                type: 'notice',
                default: '',
            },
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                qs: {
                    api_key: '={{$credentials.apiKey}}',
                },
            },
        };
    }
}
exports.AirtableApi = AirtableApi;
