// n8n-nodes-base/dist/credentials/BeeminderApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BeeminderApi = void 0;
class BeeminderApi {
    constructor() {
        this.name = 'beeminderApi';
        this.displayName = 'Beeminder API';
        this.documentationUrl = 'beeminder';
        this.properties = [
            {
                displayName: 'User',
                name: 'user',
                type: 'string',
                default: '',
            },
            {
                displayName: 'Auth Token',
                name: 'authToken',
                type: 'string',
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                body: {
                    auth_token: '={{$credentials.authToken}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://www.beeminder.com/api/v1',
                url: '=/users/{{$credentials.user}}.json',
            },
        };
    }
}
exports.BeeminderApi = BeeminderApi;
