// n8n-nodes-base/dist/credentials/CitrixAdcApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CitrixAdcApi = void 0;
class CitrixAdcApi {
    constructor() {
        this.name = 'citrixAdcApi';
        this.displayName = 'Citrix ADC API';
        this.documentationUrl = 'citrixadc';
        this.properties = [
            {
                displayName: 'URL',
                name: 'url',
                type: 'string',
                default: '',
                required: true,
            },
            {
                displayName: 'Username',
                name: 'username',
                type: 'string',
                default: '',
                required: true,
            },
            {
                displayName: 'Password',
                name: 'password',
                type: 'string',
                default: '',
                required: true,
                typeOptions: {
                    password: true,
                },
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'X-NITRO-USER': '={{$credentials.username}}',
                    'X-NITRO-PASS': '={{$credentials.password}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{$credentials.url}}',
                url: '/nitro/v1/config/nspartition?view=summary',
            },
        };
    }
}
exports.CitrixAdcApi = CitrixAdcApi;
