// n8n-nodes-base/dist/credentials/DeepLApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DeepLApi = void 0;
class DeepLApi {
    constructor() {
        this.name = 'deepLApi';
        this.displayName = 'DeepL API';
        this.documentationUrl = 'deepL';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
            {
                displayName: 'API Plan',
                name: 'apiPlan',
                type: 'options',
                options: [
                    {
                        name: 'Pro Plan',
                        value: 'pro',
                    },
                    {
                        name: 'Free Plan',
                        value: 'free',
                    },
                ],
                default: 'pro',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                qs: {
                    auth_key: '={{$credentials.apiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{$credentials.apiPlan === "pro" ? "https://api.deepl.com/v2" : "https://api-free.deepl.com/v2" }}',
                url: '/usage',
            },
        };
    }
}
exports.DeepLApi = DeepLApi;
