// n8n-nodes-base/dist/credentials/DiscourseApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiscourseApi = void 0;
class DiscourseApi {
    constructor() {
        this.name = 'discourseApi';
        this.displayName = 'Discourse API';
        this.documentationUrl = 'discourse';
        this.properties = [
            {
                displayName: 'URL',
                name: 'url',
                required: true,
                type: 'string',
                default: '',
            },
            {
                displayName: 'API Key',
                name: 'apiKey',
                required: true,
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
            {
                displayName: 'Username',
                name: 'username',
                required: true,
                type: 'string',
                default: '',
            },
        ];
        this.test = {
            request: {
                baseURL: '={{$credentials.url}}',
                url: '/groups.json',
                method: 'GET',
            },
        };
    }
    async authenticate(credentials, requestOptions) {
        requestOptions.headers = {
            'Api-Key': credentials.apiKey,
            'Api-Username': credentials.username,
        };
        if (requestOptions.method === 'GET') {
            delete requestOptions.body;
        }
        return requestOptions;
    }
}
exports.DiscourseApi = DiscourseApi;
