// n8n-nodes-base/dist/credentials/DropcontactApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DropcontactApi = void 0;
class DropcontactApi {
    constructor() {
        this.name = 'dropcontactApi';
        this.displayName = 'Dropcontact API';
        this.documentationUrl = 'dropcontact';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'user-agent': 'n8n',
                    'X-Access-Token': '={{$credentials.apiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://api.dropcontact.io',
                url: '/batch',
                method: 'POST',
                body: {
                    data: [{ email: '' }],
                },
            },
        };
    }
}
exports.DropcontactApi = DropcontactApi;
