// n8n-nodes-base/dist/credentials/GhostContentApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GhostContentApi = void 0;
class GhostContentApi {
    constructor() {
        this.name = 'ghostContentApi';
        this.displayName = 'Ghost Content API';
        this.documentationUrl = 'ghost';
        this.properties = [
            {
                displayName: 'URL',
                name: 'url',
                type: 'string',
                default: '',
                placeholder: 'http://localhost:3001',
            },
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.test = {
            request: {
                baseURL: '={{$credentials.url}}',
                url: '/ghost/api/v3/content/settings/',
                method: 'GET',
            },
        };
    }
    async authenticate(credentials, requestOptions) {
        requestOptions.qs = {
            ...requestOptions.qs,
            key: credentials.apiKey,
        };
        return requestOptions;
    }
}
exports.GhostContentApi = GhostContentApi;
