// n8n-nodes-base/dist/credentials/GoogleCloudStorageOAuth2Api.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoogleCloudStorageOAuth2Api = void 0;
const scopes = [
    'https://www.googleapis.com/auth/cloud-platform',
    'https://www.googleapis.com/auth/cloud-platform.read-only',
    'https://www.googleapis.com/auth/devstorage.full_control',
    'https://www.googleapis.com/auth/devstorage.read_only',
    'https://www.googleapis.com/auth/devstorage.read_write',
];
class GoogleCloudStorageOAuth2Api {
    constructor() {
        this.name = 'googleCloudStorageOAuth2Api';
        this.extends = ['googleOAuth2Api'];
        this.displayName = 'Google Cloud Storage OAuth2 API';
        this.documentationUrl = 'google/oauth-single-service';
        this.properties = [
            {
                displayName: 'Scope',
                name: 'scope',
                type: 'hidden',
                default: scopes.join(' '),
            },
        ];
    }
}
exports.GoogleCloudStorageOAuth2Api = GoogleCloudStorageOAuth2Api;
