// n8n-nodes-base/dist/credentials/HubspotApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HubspotApi = void 0;
class HubspotApi {
    constructor() {
        this.name = 'hubspotApi';
        this.displayName = 'HubSpot API';
        this.documentationUrl = 'hubspot';
        this.properties = [
            {
                displayName: 'On 30 November, 2022 Hubspot will remove API key support. You will have to connect to HubSpot using private app or Oauth2 auth method. <a target="_blank" href="https://developers.hubspot.com/changelog/upcoming-api-key-sunset">More details (HubSpot.com)</a>',
                name: 'notice',
                type: 'notice',
                default: '',
            },
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                qs: {
                    hapikey: '={{$credentials.apiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://api.hubapi.com',
                url: '/account-info/v3/details',
            },
        };
    }
}
exports.HubspotApi = HubspotApi;
