// n8n-nodes-base/dist/credentials/LemlistApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LemlistApi = void 0;
class LemlistApi {
    constructor() {
        this.name = 'lemlistApi';
        this.displayName = 'Lemlist API';
        this.documentationUrl = 'lemlist';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.test = {
            request: {
                baseURL: 'https://api.lemlist.com/api',
                url: '/campaigns',
            },
        };
    }
    async authenticate(credentials, requestOptions) {
        const encodedApiKey = Buffer.from(':' + credentials.apiKey).toString('base64');
        requestOptions.headers.Authorization = `Basic ${encodedApiKey}`;
        requestOptions.headers['user-agent'] = 'n8n';
        return requestOptions;
    }
}
exports.LemlistApi = LemlistApi;
