// n8n-nodes-base/dist/credentials/MicrosoftDynamicsOAuth2Api.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MicrosoftDynamicsOAuth2Api = void 0;
class MicrosoftDynamicsOAuth2Api {
    constructor() {
        this.name = 'microsoftDynamicsOAuth2Api';
        this.extends = ['microsoftOAuth2Api'];
        this.displayName = 'Microsoft Dynamics OAuth2 API';
        this.documentationUrl = 'microsoft';
        this.properties = [
            {
                displayName: 'Subdomain',
                name: 'subdomain',
                type: 'string',
                required: true,
                placeholder: 'organization',
                default: '',
            },
            {
                displayName: 'Region',
                name: 'region',
                type: 'options',
                default: 'crm.dynamics.com',
                options: [
                    {
                        name: 'Asia Pacific (APAC/ APJ)',
                        value: 'crm5.dynamics.com',
                    },
                    {
                        name: 'Australia (OCE)',
                        value: 'crm6.dynamics.com',
                    },
                    {
                        name: 'Canada (CAN)',
                        value: 'crm3.dynamics.com',
                    },
                    {
                        name: 'China (CHN)',
                        value: 'crm.dynamics.cn',
                    },
                    {
                        name: 'Europe, Middle East, Africa (EMEA/ EUR)',
                        value: 'crm4.dynamics.com',
                    },
                    {
                        name: 'France (FRA)',
                        value: 'crm12.dynamics.com',
                    },
                    {
                        name: 'Germany (GER)',
                        value: 'crm16.dynamics.com',
                    },
                    {
                        name: 'India (IND)',
                        value: 'crm8.dynamics.com',
                    },
                    {
                        name: 'Japan (JPN)',
                        value: 'crm7.dynamics.com',
                    },
                    {
                        name: 'Microsoft Cloud Germany (DEU)',
                        value: 'crm.microsoftdynamics.de',
                    },
                    {
                        name: 'North America (NAM)',
                        value: 'crm.dynamics.com',
                    },
                    {
                        name: 'North America 2 (US Gov GCC)',
                        value: 'crm9.dynamics.com',
                    },
                    {
                        name: 'South Africa (ZAF)',
                        value: 'crm14.dynamics.com',
                    },
                    {
                        name: 'South America (LATAM/ SAM)',
                        value: 'crm2.dynamics.com',
                    },
                    {
                        name: 'Switzerland (CHE)',
                        value: 'crm17.dynamics.com',
                    },
                    {
                        name: 'United Arab Emirates (UAE)',
                        value: 'crm15.dynamics.com',
                    },
                    {
                        name: 'United Kingdom (UK/ GBR)',
                        value: 'crm11.dynamics.com',
                    },
                    {
                        name: 'United States Government Community Cloud (GCC High)',
                        value: 'crm.microsoftdynamics.us',
                    },
                ],
            },
            {
                displayName: 'Scope',
                name: 'scope',
                type: 'hidden',
                default: '=openid offline_access https://{{$self.subdomain}}.{{$self.region}}/.default',
            },
        ];
    }
}
exports.MicrosoftDynamicsOAuth2Api = MicrosoftDynamicsOAuth2Api;
