// n8n-nodes-base/dist/credentials/MindeeReceiptApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MindeeReceiptApi = void 0;
class MindeeReceiptApi {
    constructor() {
        this.name = 'mindeeReceiptApi';
        this.displayName = 'Mindee Receipt API';
        this.documentationUrl = 'mindee';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
    }
    async authenticate(credentials, requestOptions) {
        const url = requestOptions.url ? requestOptions.url : requestOptions.uri;
        if (url.includes('https://api.mindee.net/v1/')) {
            requestOptions.headers.Authorization = `Token ${credentials.apiKey}`;
        }
        else {
            requestOptions.headers['X-Inferuser-Token'] = `${credentials.apiKey}`;
        }
        return requestOptions;
    }
}
exports.MindeeReceiptApi = MindeeReceiptApi;
