// n8n-nodes-base/dist/credentials/MindeeInvoiceApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MindeeInvoiceApi = void 0;
class MindeeInvoiceApi {
    constructor() {
        this.name = 'mindeeInvoiceApi';
        this.displayName = 'Mindee Invoice API';
        this.documentationUrl = 'mindee';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
    }
    async authenticate(credentials, requestOptions) {
        const url = requestOptions.url ? requestOptions.url : requestOptions.uri;
        if (url.includes('https://api.mindee.net/v1/')) {
            requestOptions.headers.Authorization = `Token ${credentials.apiKey}`;
        }
        else {
            requestOptions.headers['X-Inferuser-Token'] = `${credentials.apiKey}`;
        }
        return requestOptions;
    }
}
exports.MindeeInvoiceApi = MindeeInvoiceApi;
