// n8n-nodes-base/dist/credentials/N8nApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.N8nApi = void 0;
class N8nApi {
    constructor() {
        this.name = 'n8nApi';
        this.displayName = 'n8n API';
        this.documentationUrl = 'https://docs.n8n.io/api/authentication/';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
                description: 'The API key for the n8n instance',
            },
            {
                displayName: 'Base URL',
                name: 'baseUrl',
                type: 'string',
                default: '',
                placeholder: 'https://<name>.app.n8n.cloud/api/v1',
                description: 'The API URL of the n8n instance',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'X-N8N-API-KEY': '={{ $credentials.apiKey }}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{ $credentials.baseUrl }}',
                url: '/workflows?limit=5',
            },
        };
    }
}
exports.N8nApi = N8nApi;
