// n8n-nodes-base/dist/credentials/NextCloudApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NextCloudApi = void 0;
class NextCloudApi {
    constructor() {
        this.name = 'nextCloudApi';
        this.displayName = 'NextCloud API';
        this.documentationUrl = 'nextCloud';
        this.properties = [
            {
                displayName: 'Web DAV URL',
                name: 'webDavUrl',
                type: 'string',
                placeholder: 'https://nextcloud.example.com/remote.php/webdav',
                default: '',
            },
            {
                displayName: 'User',
                name: 'user',
                type: 'string',
                default: '',
            },
            {
                displayName: 'Password',
                name: 'password',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.test = {
            request: {
                baseURL: "={{$credentials.webDavUrl.replace('/remote.php/webdav', '')}}",
                url: '/ocs/v1.php/cloud/capabilities',
                headers: { 'OCS-APIRequest': true },
            },
        };
    }
    async authenticate(credentials, requestOptions) {
        requestOptions.auth = {
            username: credentials.user,
            password: credentials.password,
        };
        return requestOptions;
    }
}
exports.NextCloudApi = NextCloudApi;
