// n8n-nodes-base/dist/credentials/NocoDbApiToken.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NocoDbApiToken = void 0;
class NocoDbApiToken {
    constructor() {
        this.name = 'nocoDbApiToken';
        this.displayName = 'NocoDB API Token';
        this.documentationUrl = 'nocoDb';
        this.properties = [
            {
                displayName: 'API Token',
                name: 'apiToken',
                type: 'string',
                default: '',
            },
            {
                displayName: 'Host',
                name: 'host',
                type: 'string',
                default: '',
                placeholder: 'http(s)://localhost:8080',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'xc-token': '={{$credentials.apiToken}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{ $credentials.host }}',
                url: '/api/v1/auth/user/me',
            },
        };
    }
}
exports.NocoDbApiToken = NocoDbApiToken;
