// n8n-nodes-base/dist/credentials/NotionApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotionApi = void 0;
class NotionApi {
    constructor() {
        this.name = 'notionApi';
        this.displayName = 'Notion API';
        this.documentationUrl = 'notion';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.test = {
            request: {
                baseURL: 'https://api.notion.com/v1',
                url: '/users/me',
            },
        };
    }
    async authenticate(credentials, requestOptions) {
        requestOptions.headers = {
            ...requestOptions.headers,
            Authorization: `Bearer ${credentials.apiKey} `,
        };
        if (!requestOptions.headers['Notion-Version']) {
            requestOptions.headers = {
                ...requestOptions.headers,
                'Notion-Version': '2022-02-22',
            };
        }
        return requestOptions;
    }
}
exports.NotionApi = NotionApi;
