// n8n-nodes-base/dist/credentials/NpmApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NpmApi = void 0;
class NpmApi {
    constructor() {
        this.name = 'npmApi';
        this.displayName = 'Npm API';
        this.documentationUrl = 'npm';
        this.properties = [
            {
                displayName: 'Access Token',
                name: 'accessToken',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
            {
                displayName: 'Registry Url',
                name: 'registryUrl',
                type: 'string',
                default: 'https://registry.npmjs.org',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.accessToken}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{$credentials.registryUrl}}',
                url: '/-/whoami',
            },
        };
    }
}
exports.NpmApi = NpmApi;
