// n8n-nodes-base/dist/credentials/OpenAiApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenAiApi = void 0;
class OpenAiApi {
    constructor() {
        this.name = 'openAiApi';
        this.displayName = 'OpenAi';
        this.documentationUrl = 'openAi';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                required: true,
                default: '',
            },
            {
                displayName: 'Organization ID',
                name: 'organizationId',
                type: 'string',
                default: '',
                description: "For users who belong to multiple organizations, you can set which organization is used for an API request. Usage from these API requests will count against the specified organization's subscription quota.",
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.apiKey}}',
                    'OpenAI-Organization': '={{$credentials.organizationId}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://api.openai.com',
                url: '/v1/models',
            },
        };
    }
}
exports.OpenAiApi = OpenAiApi;
