// n8n-nodes-base/dist/credentials/PhantombusterApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PhantombusterApi = void 0;
class PhantombusterApi {
    constructor() {
        this.name = 'phantombusterApi';
        this.displayName = 'Phantombuster API';
        this.documentationUrl = 'phantombuster';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'X-Phantombuster-Key': '={{$credentials.apiKey}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://api.phantombuster.com/api/v2',
                url: '/agents/fetch-all',
            },
        };
    }
}
exports.PhantombusterApi = PhantombusterApi;
