// n8n-nodes-base/dist/credentials/PushoverApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PushoverApi = void 0;
class PushoverApi {
    constructor() {
        this.name = 'pushoverApi';
        this.displayName = 'Pushover API';
        this.documentationUrl = 'pushover';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.test = {
            request: {
                baseURL: 'https://api.pushover.net/1',
                url: '=/licenses.json?token={{$credentials?.apiKey}}',
                method: 'GET',
            },
        };
    }
    async authenticate(credentials, requestOptions) {
        if (requestOptions.method === 'GET' && requestOptions.qs) {
            Object.assign(requestOptions.qs, { token: credentials.apiKey });
        }
        else if (requestOptions.body) {
            Object.assign(requestOptions.body, { token: credentials.apiKey });
        }
        return requestOptions;
    }
}
exports.PushoverApi = PushoverApi;
