// n8n-nodes-base/dist/credentials/RabbitMQ.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RabbitMQ = void 0;
class RabbitMQ {
    constructor() {
        this.name = 'rabbitmq';
        this.displayName = 'RabbitMQ';
        this.documentationUrl = 'rabbitmq';
        this.properties = [
            {
                displayName: 'Hostname',
                name: 'hostname',
                type: 'string',
                default: '',
                placeholder: 'localhost',
            },
            {
                displayName: 'Port',
                name: 'port',
                type: 'number',
                default: 5672,
            },
            {
                displayName: 'User',
                name: 'username',
                type: 'string',
                default: '',
                placeholder: 'guest',
            },
            {
                displayName: 'Password',
                name: 'password',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                default: '',
                placeholder: 'guest',
            },
            {
                displayName: 'Vhost',
                name: 'vhost',
                type: 'string',
                default: '/',
            },
            {
                displayName: 'SSL',
                name: 'ssl',
                type: 'boolean',
                default: false,
            },
            {
                displayName: 'Passwordless',
                name: 'passwordless',
                type: 'boolean',
                displayOptions: {
                    show: {
                        ssl: [true],
                    },
                },
                default: true,
                description: 'Whether to use passwordless connection with certificates (SASL mechanism EXTERNAL)',
            },
            {
                displayName: 'CA Certificates',
                name: 'ca',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                displayOptions: {
                    show: {
                        ssl: [true],
                    },
                },
                default: '',
                description: 'SSL CA Certificates to use',
            },
            {
                displayName: 'Client Certificate',
                name: 'cert',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                displayOptions: {
                    show: {
                        ssl: [true],
                        passwordless: [true],
                    },
                },
                default: '',
                description: 'SSL Client Certificate to use',
            },
            {
                displayName: 'Client Key',
                name: 'key',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                displayOptions: {
                    show: {
                        ssl: [true],
                        passwordless: [true],
                    },
                },
                default: '',
                description: 'SSL Client Key to use',
            },
            {
                displayName: 'Passphrase',
                name: 'passphrase',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                displayOptions: {
                    show: {
                        ssl: [true],
                        passwordless: [true],
                    },
                },
                default: '',
                description: 'SSL passphrase to use',
            },
        ];
    }
}
exports.RabbitMQ = RabbitMQ;
