// n8n-nodes-base/dist/credentials/RedditOAuth2Api.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RedditOAuth2Api = void 0;
const scopes = ['identity', 'edit', 'history', 'mysubreddits', 'read', 'save', 'submit'];
class RedditOAuth2Api {
    constructor() {
        this.name = 'redditOAuth2Api';
        this.extends = ['oAuth2Api'];
        this.displayName = 'Reddit OAuth2 API';
        this.documentationUrl = 'reddit';
        this.properties = [
            {
                displayName: 'Grant Type',
                name: 'grantType',
                type: 'hidden',
                default: 'authorizationCode',
            },
            {
                displayName: 'Auth URI Query Parameters',
                name: 'authQueryParameters',
                type: 'hidden',
                default: 'response_type=code',
            },
            {
                displayName: 'Auth URI Query Parameters',
                name: 'authQueryParameters',
                type: 'hidden',
                default: 'grant_type=authorization_code',
            },
            {
                displayName: 'Auth URI Query Parameters',
                name: 'authQueryParameters',
                type: 'hidden',
                default: 'duration=permanent',
            },
            {
                displayName: 'Authorization URL',
                name: 'authUrl',
                type: 'hidden',
                default: 'https://www.reddit.com/api/v1/authorize',
            },
            {
                displayName: 'Access Token URL',
                name: 'accessTokenUrl',
                type: 'hidden',
                default: 'https://www.reddit.com/api/v1/access_token',
            },
            {
                displayName: 'Scope',
                name: 'scope',
                type: 'hidden',
                default: scopes.join(' '),
            },
            {
                displayName: 'Authentication',
                name: 'authentication',
                type: 'hidden',
                default: 'header',
            },
        ];
    }
}
exports.RedditOAuth2Api = RedditOAuth2Api;
