// n8n-nodes-base/dist/credentials/ShopifyAccessTokenApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShopifyAccessTokenApi = void 0;
class ShopifyAccessTokenApi {
    constructor() {
        this.name = 'shopifyAccessTokenApi';
        this.displayName = 'Shopify Access Token API';
        this.documentationUrl = 'shopify';
        this.properties = [
            {
                displayName: 'Shop Subdomain',
                name: 'shopSubdomain',
                required: true,
                type: 'string',
                default: '',
                description: 'Only the subdomain without .myshopify.com',
            },
            {
                displayName: 'Access Token',
                name: 'accessToken',
                required: true,
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
            {
                displayName: 'APP Secret Key',
                name: 'appSecretKey',
                required: true,
                type: 'string',
                default: '',
                description: 'Secret key needed to verify the webhook when using Shopify Trigger node',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'X-Shopify-Access-Token': '={{$credentials?.accessToken}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '=https://{{$credentials?.shopSubdomain}}.myshopify.com/admin/api/2019-10',
                url: '/products.json',
            },
        };
    }
}
exports.ShopifyAccessTokenApi = ShopifyAccessTokenApi;
