// n8n-nodes-base/dist/credentials/TotpApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TotpApi = void 0;
class TotpApi {
    constructor() {
        this.name = 'totpApi';
        this.displayName = 'TOTP API';
        this.documentationUrl = 'totp';
        this.properties = [
            {
                displayName: 'Secret',
                name: 'secret',
                type: 'string',
                typeOptions: { password: true },
                default: '',
                placeholder: 'e.g. BVDRSBXQB2ZEL5HE',
                required: true,
                description: 'Secret key encoded in the QR code during setup. <a href="https://github.com/google/google-authenticator/wiki/Key-Uri-Format#secret">Learn more</a>.',
            },
            {
                displayName: 'Label',
                name: 'label',
                type: 'string',
                default: '',
                required: true,
                placeholder: 'e.g. GitHub:john-doe',
                description: 'Identifier for the TOTP account, in the <code>issuer:username</code> format. <a href="https://github.com/google/google-authenticator/wiki/Key-Uri-Format#label">Learn more</a>.',
            },
        ];
    }
}
exports.TotpApi = TotpApi;
