// n8n-nodes-base/dist/credentials/TrelloApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TrelloApi = void 0;
class TrelloApi {
    constructor() {
        this.name = 'trelloApi';
        this.displayName = 'Trello API';
        this.documentationUrl = 'trello';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                required: true,
                default: '',
            },
            {
                displayName: 'API Token',
                name: 'apiToken',
                type: 'string',
                required: true,
                default: '',
            },
            {
                displayName: 'OAuth Secret',
                name: 'oauthSecret',
                type: 'hidden',
                default: '',
            },
        ];
        this.test = {
            request: {
                baseURL: 'https://api.trello.com',
                url: '=/1/tokens/{{$credentials.apiToken}}/member',
            },
        };
    }
    async authenticate(credentials, requestOptions) {
        requestOptions.qs = {
            ...requestOptions.qs,
            key: credentials.apiKey,
            token: credentials.apiToken,
        };
        return requestOptions;
    }
}
exports.TrelloApi = TrelloApi;
