// n8n-nodes-base/dist/credentials/UProcApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UProcApi = void 0;
class UProcApi {
    constructor() {
        this.name = 'uprocApi';
        this.displayName = 'uProc API';
        this.documentationUrl = 'uProc';
        this.properties = [
            {
                displayName: 'Email',
                name: 'email',
                type: 'string',
                placeholder: 'name@email.com',
                default: '',
            },
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.test = {
            request: {
                baseURL: 'https://api.uproc.io/api/v2',
                url: '/profile',
                method: 'GET',
            },
        };
    }
    async authenticate(credentials, requestOptions) {
        const token = Buffer.from(`${credentials.email}:${credentials.apiKey}`).toString('base64');
        requestOptions.headers = {
            ...requestOptions.headers,
            Authorization: `Basic ${token}`,
        };
        return requestOptions;
    }
}
exports.UProcApi = UProcApi;
