// n8n-nodes-base/dist/credentials/VenafiTlsProtectDatacenterApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VenafiTlsProtectDatacenterApi = void 0;
class VenafiTlsProtectDatacenterApi {
    constructor() {
        this.name = 'venafiTlsProtectDatacenterApi';
        this.displayName = 'Venafi TLS Protect Datacenter API';
        this.documentationUrl = 'venafitlsprotectdatacenter';
        this.properties = [
            {
                displayName: 'Domain',
                name: 'domain',
                type: 'string',
                default: '',
                placeholder: 'https://example.com',
            },
            {
                displayName: 'Client ID',
                name: 'clientId',
                type: 'string',
                default: '',
            },
            {
                displayName: 'Username',
                name: 'username',
                type: 'string',
                default: '',
            },
            {
                displayName: 'Password',
                name: 'password',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                default: '',
            },
            {
                displayName: 'Allow Self-Signed Certificates',
                name: 'allowUnauthorizedCerts',
                type: 'boolean',
                default: true,
            },
            {
                displayName: 'Access Token',
                name: 'token',
                type: 'hidden',
                typeOptions: {
                    expirable: true,
                },
                default: '',
            },
            {
                displayName: 'Scope',
                name: 'scope',
                type: 'hidden',
                default: 'certificate:manage',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.token}}',
                },
            },
        };
    }
    async preAuthentication(credentials) {
        const url = `${credentials.domain}/vedauth/authorize/oauth`;
        const requestOptions = {
            url,
            method: 'POST',
            json: true,
            skipSslCertificateValidation: credentials.allowUnauthorizedCerts,
            body: {
                client_id: credentials.clientId,
                username: credentials.username,
                password: credentials.password,
                scope: credentials.scope,
            },
        };
        const { access_token } = (await this.helpers.httpRequest(requestOptions));
        return { token: access_token };
    }
}
exports.VenafiTlsProtectDatacenterApi = VenafiTlsProtectDatacenterApi;
