// n8n-nodes-base/dist/credentials/WekanApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WekanApi = void 0;
class WekanApi {
    constructor() {
        this.name = 'wekanApi';
        this.displayName = 'Wekan API';
        this.documentationUrl = 'wekan';
        this.properties = [
            {
                displayName: 'Username',
                name: 'username',
                type: 'string',
                default: '',
            },
            {
                displayName: 'Password',
                name: 'password',
                type: 'string',
                typeOptions: {
                    password: true,
                },
                default: '',
            },
            {
                displayName: 'URL',
                name: 'url',
                type: 'string',
                default: '',
                placeholder: 'https://wekan.yourdomain.com',
            },
            {
                displayName: 'Session Token',
                name: 'token',
                type: 'hidden',
                typeOptions: {
                    expirable: true,
                },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.token}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '={{$credentials.url.replace(new RegExp("/$"), "")}}',
                url: '/api/user',
            },
        };
    }
    async preAuthentication(credentials) {
        const url = credentials.url;
        const { token } = (await this.helpers.httpRequest({
            method: 'POST',
            url: `${url.endsWith('/') ? url.slice(0, -1) : url}/users/login`,
            body: {
                username: credentials.username,
                password: credentials.password,
            },
        }));
        return { token };
    }
}
exports.WekanApi = WekanApi;
