// n8n-nodes-base/dist/credentials/WhatsAppApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WhatsAppApi = void 0;
class WhatsAppApi {
    constructor() {
        this.name = 'whatsAppApi';
        this.displayName = 'WhatsApp API';
        this.documentationUrl = 'whatsApp';
        this.properties = [
            {
                displayName: 'Access Token',
                type: 'string',
                typeOptions: { password: true },
                name: 'accessToken',
                default: '',
                required: true,
            },
            {
                displayName: 'Bussiness Account ID',
                type: 'string',
                name: 'businessAccountId',
                default: '',
                required: true,
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.accessToken}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://graph.facebook.com/v13.0',
                url: '/',
                ignoreHttpStatusErrors: true,
            },
            rules: [
                {
                    type: 'responseSuccessBody',
                    properties: {
                        key: 'error.type',
                        value: 'OAuthException',
                        message: 'Invalid access token',
                    },
                },
            ],
        };
    }
}
exports.WhatsAppApi = WhatsAppApi;
