// n8n-nodes-base/dist/credentials/WufooApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WufooApi = void 0;
class WufooApi {
    constructor() {
        this.name = 'wufooApi';
        this.displayName = 'Wufoo API';
        this.documentationUrl = 'wufoo';
        this.properties = [
            {
                displayName: 'API Key',
                name: 'apiKey',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
            {
                displayName: 'Subdomain',
                name: 'subdomain',
                type: 'string',
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                auth: {
                    username: '={{$credentials.apiKey}}',
                    password: 'not-needed',
                },
            },
        };
        this.test = {
            request: {
                baseURL: '=https://{{$credentials.subdomain}}.wufoo.com',
                url: '/api/v3/forms.json',
            },
        };
    }
}
exports.WufooApi = WufooApi;
