// n8n-nodes-base/dist/credentials/ZendeskApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZendeskApi = void 0;
class ZendeskApi {
    constructor() {
        this.name = 'zendeskApi';
        this.displayName = 'Zendesk API';
        this.documentationUrl = 'zendesk';
        this.properties = [
            {
                displayName: 'Subdomain',
                name: 'subdomain',
                type: 'string',
                description: 'The subdomain of your Zendesk work environment',
                placeholder: 'company',
                default: '',
            },
            {
                displayName: 'Email',
                name: 'email',
                type: 'string',
                placeholder: 'name@email.com',
                default: '',
            },
            {
                displayName: 'API Token',
                name: 'apiToken',
                type: 'string',
                default: '',
            },
        ];
        this.test = {
            request: {
                baseURL: '=https://{{$credentials.subdomain}}.zendesk.com/api/v2',
                url: '/ticket_fields.json',
            },
        };
    }
    async authenticate(credentials, requestOptions) {
        requestOptions.auth = {
            username: `${credentials.email}/token`,
            password: credentials.apiToken,
        };
        return requestOptions;
    }
}
exports.ZendeskApi = ZendeskApi;
