// n8n-nodes-base/dist/credentials/ZoomApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZoomApi = void 0;
class ZoomApi {
    constructor() {
        this.name = 'zoomApi';
        this.displayName = 'Zoom API';
        this.documentationUrl = 'zoom';
        this.properties = [
            {
                displayName: 'On 1 June, 2023 Zoom will remove JWT App support. You will have to connect to Zoom using the Oauth2 auth method. <a target="_blank" href="https://marketplace.zoom.us/docs/guides/build/jwt-app/jwt-faq/">More details (zoom.us)</a>',
                name: 'notice',
                type: 'notice',
                default: '',
            },
            {
                displayName: 'JWT Token',
                name: 'accessToken',
                type: 'string',
                typeOptions: { password: true },
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    Authorization: '=Bearer {{$credentials.accessToken}}',
                },
            },
        };
        this.test = {
            request: {
                baseURL: 'https://api.zoom.us/v2',
                url: '/users/me',
            },
        };
    }
}
exports.ZoomApi = ZoomApi;
