// n8n-nodes-base/dist/nodes/Vero/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateJSON = exports.veroApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function veroApiRequest(method, resource, body = {}, qs = {}, uri, option = {}) {
    const credentials = await this.getCredentials('veroApi');
    let options = {
        method,
        qs,
        body,
        form: {
            auth_token: credentials.authToken,
            ...body,
        },
        uri: uri || `https://api.getvero.com/api/v2${resource}`,
        json: true,
    };
    options = Object.assign({}, options, option);
    if (Object.keys(options.body).length === 0) {
        delete options.body;
    }
    try {
        return await this.helpers.request(options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.veroApiRequest = veroApiRequest;
function validateJSON(json) {
    let result;
    try {
        result = JSON.parse(json);
    }
    catch (exception) {
        result = undefined;
    }
    return result;
}
exports.validateJSON = validateJSON;
