// n8n-nodes-base/dist/nodes/Microsoft/Excel/v2/actions/table/getColumns.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const utilities_1 = require("../../../../../../utils/utilities");
const transport_1 = require("../../transport");
const common_descriptions_1 = require("../common.descriptions");
const properties = [
    common_descriptions_1.workbookRLC,
    common_descriptions_1.worksheetRLC,
    common_descriptions_1.tableRLC,
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        default: false,
        description: 'Whether to return all results or only up to a given limit',
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        displayOptions: {
            show: {
                returnAll: [false],
            },
        },
        typeOptions: {
            minValue: 1,
            maxValue: 500,
        },
        default: 100,
        description: 'Max number of results to return',
    },
    {
        displayName: 'RAW Data',
        name: 'rawData',
        type: 'boolean',
        default: false,
        description: 'Whether the data should be returned RAW instead of parsed into keys according to their header',
    },
    {
        displayName: 'Data Property',
        name: 'dataProperty',
        type: 'string',
        default: 'data',
        displayOptions: {
            show: {
                rawData: [true],
            },
        },
        description: 'The name of the property into which to write the RAW data',
    },
    {
        displayName: 'Filters',
        name: 'filters',
        type: 'collection',
        placeholder: 'Add Filter',
        default: {},
        displayOptions: {
            show: {
                rawData: [true],
            },
        },
        options: [
            {
                displayName: 'Fields',
                name: 'fields',
                type: 'string',
                default: '',
                description: 'A comma-separated list of the fields to include in the response',
            },
        ],
    },
];
const displayOptions = {
    show: {
        resource: ['table'],
        operation: ['getColumns'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(items) {
    const returnData = [];
    for (let i = 0; i < items.length; i++) {
        try {
            const qs = {};
            const workbookId = this.getNodeParameter('workbook', i, undefined, {
                extractValue: true,
            });
            const worksheetId = this.getNodeParameter('worksheet', i, undefined, {
                extractValue: true,
            });
            const tableId = this.getNodeParameter('table', i, undefined, {
                extractValue: true,
            });
            const returnAll = this.getNodeParameter('returnAll', i);
            const rawData = this.getNodeParameter('rawData', i);
            if (rawData) {
                const filters = this.getNodeParameter('filters', i);
                if (filters.fields) {
                    qs.$select = filters.fields;
                }
            }
            let responseData;
            if (returnAll) {
                responseData = await transport_1.microsoftApiRequestAllItemsSkip.call(this, 'value', 'GET', `/drive/items/${workbookId}/workbook/worksheets/${worksheetId}/tables/${tableId}/columns`, {}, qs);
            }
            else {
                qs.$top = this.getNodeParameter('limit', i);
                responseData = await transport_1.microsoftApiRequest.call(this, 'GET', `/drive/items/${workbookId}/workbook/worksheets/${worksheetId}/tables/${tableId}/columns`, {}, qs);
                responseData = responseData.value;
            }
            if (!rawData) {
                responseData = responseData.map((column) => ({ name: column.name }));
            }
            else {
                const dataProperty = this.getNodeParameter('dataProperty', i);
                responseData = { [dataProperty]: responseData };
            }
            const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
            returnData.push(...executionData);
        }
        catch (error) {
            if (this.continueOnFail()) {
                const executionErrorData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray({ error: error.message }), { itemData: { item: i } });
                returnData.push(...executionErrorData);
                continue;
            }
            throw error;
        }
    }
    return returnData;
}
exports.execute = execute;
