// n8n-nodes-base/dist/nodes/BambooHr/v1/actions/employeeDocument/download/description.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.employeeDocumentDownloadDescription = void 0;
exports.employeeDocumentDownloadDescription = [
    {
        displayName: 'Employee ID',
        name: 'employeeId',
        type: 'string',
        required: true,
        displayOptions: {
            show: {
                operation: ['download'],
                resource: ['employeeDocument'],
            },
        },
        default: '',
        description: 'ID of the employee',
    },
    {
        displayName: 'File ID',
        name: 'fileId',
        type: 'string',
        required: true,
        displayOptions: {
            show: {
                operation: ['download'],
                resource: ['employeeDocument'],
            },
        },
        default: '',
        description: 'ID of the employee file',
    },
    {
        displayName: 'Put Output In Field',
        name: 'output',
        type: 'string',
        default: 'data',
        required: true,
        description: 'The name of the output field to put the binary file data in',
        displayOptions: {
            show: {
                operation: ['download'],
                resource: ['employeeDocument'],
            },
        },
    },
];
