// n8n-nodes-base/dist/nodes/MySql/v2/actions/database/Database.resource.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.upsert = exports.update = exports.select = exports.insert = exports.executeQuery = exports.deleteTable = void 0;
const deleteTable = __importStar(require("./deleteTable.operation"));
exports.deleteTable = deleteTable;
const executeQuery = __importStar(require("./executeQuery.operation"));
exports.executeQuery = executeQuery;
const insert = __importStar(require("./insert.operation"));
exports.insert = insert;
const select = __importStar(require("./select.operation"));
exports.select = select;
const update = __importStar(require("./update.operation"));
exports.update = update;
const upsert = __importStar(require("./upsert.operation"));
exports.upsert = upsert;
const common_descriptions_1 = require("../common.descriptions");
exports.description = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        options: [
            {
                name: 'Delete',
                value: 'deleteTable',
                description: 'Delete an entire table or rows in a table',
                action: 'Delete table or rows',
            },
            {
                name: 'Execute SQL',
                value: 'executeQuery',
                description: 'Execute an SQL query',
                action: 'Execute a SQL query',
            },
            {
                name: 'Insert',
                value: 'insert',
                description: 'Insert rows in a table',
                action: 'Insert rows in a table',
            },
            {
                name: 'Insert or Update',
                value: 'upsert',
                description: 'Insert or update rows in a table',
                action: 'Insert or update rows in a table',
            },
            {
                name: 'Select',
                value: 'select',
                description: 'Select rows from a table',
                action: 'Select rows from a table',
            },
            {
                name: 'Update',
                value: 'update',
                description: 'Update rows in a table',
                action: 'Update rows in a table',
            },
        ],
        displayOptions: {
            show: {
                resource: ['database'],
            },
        },
        default: 'insert',
    },
    {
        ...common_descriptions_1.tableRLC,
        displayOptions: { hide: { operation: ['executeQuery'] } },
    },
    ...deleteTable.description,
    ...executeQuery.description,
    ...insert.description,
    ...select.description,
    ...update.description,
    ...upsert.description,
];
