// n8n-nodes-base/dist/nodes/MySql/v2/actions/database/executeQuery.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const utilities_1 = require("../../../../../utils/utilities");
const utils_1 = require("../../helpers/utils");
const common_descriptions_1 = require("../common.descriptions");
const properties = [
    {
        displayName: 'Query',
        name: 'query',
        type: 'string',
        default: '',
        placeholder: 'e.g. SELECT id, name FROM product WHERE id < 40',
        required: true,
        description: "The SQL query to execute. You can use n8n expressions and $1, $2, $3, etc to refer to the 'Query Parameters' set in options below.",
        typeOptions: {
            editor: 'sqlEditor',
            sqlDialect: 'mysql',
        },
        hint: 'Prefer using query parameters over n8n expressions to avoid SQL injection attacks',
    },
    {
        displayName: `
		To use query parameters in your SQL query, reference them as $1, $2, $3, etc in the corresponding order. <a target="_blank" href="https://docs.n8n.io/integrations/builtin/app-nodes/n8n-nodes-base.mysql/">More info</a>.
		`,
        name: 'notice',
        type: 'notice',
        default: '',
    },
    common_descriptions_1.optionsCollection,
];
const displayOptions = {
    show: {
        resource: ['database'],
        operation: ['executeQuery'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(inputItems, runQueries, nodeOptions) {
    let returnData = [];
    const items = (0, utils_1.replaceEmptyStringsByNulls)(inputItems, nodeOptions.replaceEmptyStrings);
    const queries = [];
    for (let i = 0; i < items.length; i++) {
        const rawQuery = this.getNodeParameter('query', i);
        const options = this.getNodeParameter('options', i, {});
        let values;
        let queryReplacement = options.queryReplacement || [];
        if (typeof queryReplacement === 'string') {
            queryReplacement = queryReplacement.split(',').map((entry) => entry.trim());
        }
        if (Array.isArray(queryReplacement)) {
            values = queryReplacement;
        }
        else {
            throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Query Replacement must be a string of comma-separated values, or an array of values', { itemIndex: i });
        }
        const preparedQuery = (0, utils_1.prepareQueryAndReplacements)(rawQuery, values);
        queries.push(preparedQuery);
    }
    returnData = await runQueries(queries);
    return returnData;
}
exports.execute = execute;
