// n8n-nodes-base/dist/nodes/Line/Line.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Line = void 0;
const GenericFunctions_1 = require("./GenericFunctions");
const NotificationDescription_1 = require("./NotificationDescription");
class Line {
    constructor() {
        this.description = {
            displayName: 'Line',
            name: 'line',
            icon: 'file:line.png',
            group: ['input'],
            version: 1,
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            description: 'Consume Line API',
            defaults: {
                name: 'Line',
            },
            inputs: ['main'],
            outputs: ['main'],
            credentials: [
                {
                    name: 'lineNotifyOAuth2Api',
                    required: true,
                    displayOptions: {
                        show: {
                            resource: ['notification'],
                        },
                    },
                },
            ],
            properties: [
                {
                    displayName: 'Resource',
                    name: 'resource',
                    type: 'options',
                    noDataExpression: true,
                    options: [
                        {
                            name: 'Notification',
                            value: 'notification',
                        },
                    ],
                    default: 'notification',
                },
                ...NotificationDescription_1.notificationOperations,
                ...NotificationDescription_1.notificationFields,
            ],
        };
    }
    async execute() {
        const items = this.getInputData();
        const returnData = [];
        const length = items.length;
        let responseData;
        const resource = this.getNodeParameter('resource', 0);
        const operation = this.getNodeParameter('operation', 0);
        for (let i = 0; i < length; i++) {
            try {
                if (resource === 'notification') {
                    if (operation === 'send') {
                        const message = this.getNodeParameter('message', i);
                        const additionalFields = this.getNodeParameter('additionalFields', i);
                        const body = {
                            message,
                        };
                        Object.assign(body, additionalFields);
                        if (body.hasOwnProperty('notificationDisabled')) {
                            body.notificationDisabled = body.notificationDisabled ? 'true' : 'false';
                        }
                        if (body.stickerUi) {
                            const sticker = body.stickerUi.stickerValue;
                            if (sticker) {
                                body.stickerId = sticker.stickerId;
                                body.stickerPackageId = sticker.stickerPackageId;
                            }
                            delete body.stickerUi;
                        }
                        if (body.imageUi) {
                            const image = body.imageUi.imageValue;
                            if (image && image.binaryData === true) {
                                const binaryProperty = image.binaryProperty;
                                const binaryData = this.helpers.assertBinaryData(i, binaryProperty);
                                const binaryDataBuffer = await this.helpers.getBinaryDataBuffer(i, binaryProperty);
                                body.imageFile = {
                                    value: binaryDataBuffer,
                                    options: {
                                        filename: binaryData.fileName,
                                    },
                                };
                            }
                            else {
                                body.imageFullsize = image.imageFullsize;
                                body.imageThumbnail = image.imageThumbnail;
                            }
                            delete body.imageUi;
                        }
                        responseData = await GenericFunctions_1.lineApiRequest.call(this, 'POST', '', {}, {}, 'https://notify-api.line.me/api/notify', { formData: body });
                    }
                }
                const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                returnData.push(...executionData);
            }
            catch (error) {
                if (this.continueOnFail()) {
                    const executionErrorData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray({ error: error.message }), { itemData: { item: i } });
                    returnData.push(...executionErrorData);
                    continue;
                }
                throw error;
            }
        }
        return this.prepareOutputData(returnData);
    }
}
exports.Line = Line;
