// n8n-nodes-base/dist/nodes/Emelia/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.emeliaApiTest = exports.loadResource = exports.emeliaGraphqlRequest = exports.emeliaApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function emeliaApiRequest(method, endpoint, body = {}, qs = {}) {
    const { apiKey } = (await this.getCredentials('emeliaApi'));
    const options = {
        headers: {
            Authorization: apiKey,
        },
        method,
        body,
        qs,
        uri: `https://graphql.emelia.io${endpoint}`,
        json: true,
    };
    try {
        return await this.helpers.request.call(this, options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.emeliaApiRequest = emeliaApiRequest;
async function emeliaGraphqlRequest(body = {}) {
    const response = await emeliaApiRequest.call(this, 'POST', '/graphql', body);
    if (response.errors) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), response);
    }
    return response;
}
exports.emeliaGraphqlRequest = emeliaGraphqlRequest;
async function loadResource(resource) {
    const mapping = {
        campaign: {
            query: `
				query GetCampaigns {
					campaigns {
						_id
						name
					}
				}`,
            key: 'campaigns',
        },
        contactList: {
            query: `
			query GetContactLists {
				contact_lists {
					_id
					name
				}
			}`,
            key: 'contact_lists',
        },
    };
    const responseData = await emeliaGraphqlRequest.call(this, { query: mapping[resource].query });
    return responseData.data[mapping[resource].key].map((campaign) => ({
        name: campaign.name,
        value: campaign._id,
    }));
}
exports.loadResource = loadResource;
async function emeliaApiTest(credential) {
    const credentials = credential.data;
    const body = {
        query: `
				query all_campaigns {
					all_campaigns {
						_id
						name
						status
						createdAt
						stats {
							mailsSent
						}
					}
				}`,
        operationName: 'all_campaigns',
    };
    const options = {
        headers: {
            Authorization: credentials === null || credentials === void 0 ? void 0 : credentials.apiKey,
        },
        method: 'POST',
        body,
        uri: 'https://graphql.emelia.io/graphql',
        json: true,
    };
    try {
        await this.helpers.request(options);
    }
    catch (error) {
        return {
            status: 'Error',
            message: `Connection details not valid: ${error.message}`,
        };
    }
    return {
        status: 'OK',
        message: 'Authentication successful!',
    };
}
exports.emeliaApiTest = emeliaApiTest;
