// n8n-core/dist/Constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CUSTOM_NODES_CATEGORY = exports.RESPONSE_ERROR_MESSAGES = exports.TUNNEL_SUBDOMAIN_ENV = exports.PLACEHOLDER_EMPTY_WORKFLOW_ID = exports.PLACEHOLDER_EMPTY_EXECUTION_ID = exports.USER_SETTINGS_SUBFOLDER = exports.USER_SETTINGS_FILE_NAME = exports.USER_FOLDER_ENV_OVERWRITE = exports.EXTENSIONS_SUBDIRECTORY = exports.ENCRYPTION_KEY_ENV_OVERWRITE = exports.DOWNLOADED_NODES_SUBDIRECTORY = exports.CUSTOM_EXTENSION_ENV = void 0;
exports.CUSTOM_EXTENSION_ENV = 'N8N_CUSTOM_EXTENSIONS';
exports.DOWNLOADED_NODES_SUBDIRECTORY = 'nodes';
exports.ENCRYPTION_KEY_ENV_OVERWRITE = 'N8N_ENCRYPTION_KEY';
exports.EXTENSIONS_SUBDIRECTORY = 'custom';
exports.USER_FOLDER_ENV_OVERWRITE = 'N8N_USER_FOLDER';
exports.USER_SETTINGS_FILE_NAME = 'config';
exports.USER_SETTINGS_SUBFOLDER = '.n8n';
exports.PLACEHOLDER_EMPTY_EXECUTION_ID = '__UNKNOWN__';
exports.PLACEHOLDER_EMPTY_WORKFLOW_ID = '__EMPTY__';
exports.TUNNEL_SUBDOMAIN_ENV = 'N8N_TUNNEL_SUBDOMAIN';
exports.RESPONSE_ERROR_MESSAGES = {
    NO_ENCRYPTION_KEY: 'Encryption key is missing or was not set',
};
exports.CUSTOM_NODES_CATEGORY = 'Custom Nodes';
