// n8n-core/dist/WorkflowExecutionMetadata.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWorkflowExecutionMetadata = exports.getAllWorkflowExecutionMetadata = exports.setAllWorkflowExecutionMetadata = exports.setWorkflowExecutionMetadata = exports.ExecutionMetadataValidationError = exports.KV_LIMIT = void 0;
const n8n_workflow_1 = require("n8n-workflow");
exports.KV_LIMIT = 10;
class ExecutionMetadataValidationError extends Error {
    constructor(type, key, message, options) {
        super(message !== null && message !== void 0 ? message : `Custom data ${type}s must be a string (key "${key}")`, options);
        this.type = type;
    }
}
exports.ExecutionMetadataValidationError = ExecutionMetadataValidationError;
function setWorkflowExecutionMetadata(executionData, key, value) {
    if (!executionData.resultData.metadata) {
        executionData.resultData.metadata = {};
    }
    if (!(key in executionData.resultData.metadata) &&
        Object.keys(executionData.resultData.metadata).length >= exports.KV_LIMIT) {
        return;
    }
    if (typeof key !== 'string') {
        throw new ExecutionMetadataValidationError('key', key);
    }
    if (key.replace(/[A-Za-z0-9_]/g, '').length !== 0) {
        throw new ExecutionMetadataValidationError('key', key, `Custom date key can only contain characters "A-Za-z0-9_" (key "${key}")`);
    }
    if (typeof value !== 'string' && typeof value !== 'number' && typeof value !== 'bigint') {
        throw new ExecutionMetadataValidationError('value', key);
    }
    const val = String(value);
    if (key.length > 50) {
        n8n_workflow_1.LoggerProxy.error('Custom data key over 50 characters long. Truncating to 50 characters.');
    }
    if (val.length > 255) {
        n8n_workflow_1.LoggerProxy.error('Custom data value over 255 characters long. Truncating to 255 characters.');
    }
    executionData.resultData.metadata[key.slice(0, 50)] = val.slice(0, 255);
}
exports.setWorkflowExecutionMetadata = setWorkflowExecutionMetadata;
function setAllWorkflowExecutionMetadata(executionData, obj) {
    const errors = [];
    Object.entries(obj).forEach(([key, value]) => {
        try {
            setWorkflowExecutionMetadata(executionData, key, value);
        }
        catch (e) {
            errors.push(e);
        }
    });
    if (errors.length) {
        throw errors[0];
    }
}
exports.setAllWorkflowExecutionMetadata = setAllWorkflowExecutionMetadata;
function getAllWorkflowExecutionMetadata(executionData) {
    var _a;
    return (_a = { ...executionData.resultData.metadata }) !== null && _a !== void 0 ? _a : {};
}
exports.getAllWorkflowExecutionMetadata = getAllWorkflowExecutionMetadata;
function getWorkflowExecutionMetadata(executionData, key) {
    return getAllWorkflowExecutionMetadata(executionData)[String(key).slice(0, 50)];
}
exports.getWorkflowExecutionMetadata = getWorkflowExecutionMetadata;
