// n8n/dist/PublicApi/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadPublicApiVersions = void 0;
const express_1 = __importDefault(require("express"));
const promises_1 = __importDefault(require("fs/promises"));
const path_1 = __importDefault(require("path"));
const validator_1 = __importDefault(require("validator"));
const express_openapi_validator_1 = require("express-openapi-validator");
const yamljs_1 = __importDefault(require("yamljs"));
const config_1 = __importDefault(require("../config"));
const Db = __importStar(require("../Db"));
const UserManagementHelper_1 = require("../UserManagement/UserManagementHelper");
const typedi_1 = require("typedi");
const InternalHooks_1 = require("../InternalHooks");
async function createApiRouter(version, openApiSpecPath, handlersDirectory, publicApiEndpoint) {
    const n8nPath = config_1.default.getEnv('path');
    const swaggerDocument = yamljs_1.default.load(openApiSpecPath);
    swaggerDocument.server = [
        {
            url: `${(0, UserManagementHelper_1.getInstanceBaseUrl)()}/${publicApiEndpoint}/${version}}`,
        },
    ];
    const apiController = express_1.default.Router();
    if (!config_1.default.getEnv('publicApi.swaggerUi.disabled')) {
        const { serveFiles, setup } = await Promise.resolve().then(() => __importStar(require('swagger-ui-express')));
        const swaggerThemePath = path_1.default.join(__dirname, 'swaggerTheme.css');
        const swaggerThemeCss = await promises_1.default.readFile(swaggerThemePath, { encoding: 'utf-8' });
        apiController.use(`/${publicApiEndpoint}/${version}/docs`, serveFiles(swaggerDocument), setup(swaggerDocument, {
            customCss: swaggerThemeCss,
            customSiteTitle: 'n8n Public API UI',
            customfavIcon: `${n8nPath}favicon.ico`,
        }));
    }
    apiController.get(`/${publicApiEndpoint}/${version}/openapi.yml`, (req, res) => {
        res.sendFile(openApiSpecPath);
    });
    apiController.use(`/${publicApiEndpoint}/${version}`, express_1.default.json(), (0, express_openapi_validator_1.middleware)({
        apiSpec: openApiSpecPath,
        operationHandlers: handlersDirectory,
        validateRequests: true,
        validateApiSpec: true,
        formats: [
            {
                name: 'email',
                type: 'string',
                validate: (email) => validator_1.default.isEmail(email),
            },
            {
                name: 'identifier',
                type: 'string',
                validate: (identifier) => validator_1.default.isUUID(identifier) || validator_1.default.isEmail(identifier),
            },
            {
                name: 'jsonString',
                validate: (data) => {
                    try {
                        JSON.parse(data);
                        return true;
                    }
                    catch (e) {
                        return false;
                    }
                },
            },
        ],
        validateSecurity: {
            handlers: {
                ApiKeyAuth: async (req, _scopes, schema) => {
                    const apiKey = req.headers[schema.name.toLowerCase()];
                    const user = await Db.collections.User.findOne({
                        where: { apiKey },
                        relations: ['globalRole'],
                    });
                    if (!user)
                        return false;
                    void typedi_1.Container.get(InternalHooks_1.InternalHooks).onUserInvokedApi({
                        user_id: user.id,
                        path: req.path,
                        method: req.method,
                        api_version: version,
                    });
                    req.user = user;
                    return true;
                },
            },
        },
    }));
    apiController.use((error, _req, res, _next) => {
        return res.status(error.status || 400).json({
            message: error.message,
        });
    });
    return apiController;
}
const loadPublicApiVersions = async (publicApiEndpoint) => {
    var _a, _b;
    const folders = await promises_1.default.readdir(__dirname);
    const versions = folders.filter((folderName) => folderName.startsWith('v'));
    const apiRouters = await Promise.all(versions.map(async (version) => {
        const openApiPath = path_1.default.join(__dirname, version, 'openapi.yml');
        return createApiRouter(version, openApiPath, __dirname, publicApiEndpoint);
    }));
    return {
        apiRouters,
        apiLatestVersion: (_b = Number((_a = versions.pop()) === null || _a === void 0 ? void 0 : _a.charAt(1))) !== null && _b !== void 0 ? _b : 1,
    };
};
exports.loadPublicApiVersions = loadPublicApiVersions;
