// n8n-nodes-base/dist/nodes/Microsoft/Excel/v2/actions/workbook/getAll.operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.execute = exports.description = void 0;
const utilities_1 = require("../../../../../../utils/utilities");
const transport_1 = require("../../transport");
const properties = [
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        default: false,
        description: 'Whether to return all results or only up to a given limit',
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        displayOptions: {
            show: {
                returnAll: [false],
            },
        },
        typeOptions: {
            minValue: 1,
            maxValue: 500,
        },
        default: 100,
        description: 'Max number of results to return',
    },
    {
        displayName: 'Filters',
        name: 'filters',
        type: 'collection',
        placeholder: 'Add Filter',
        default: {},
        options: [
            {
                displayName: 'Fields',
                name: 'fields',
                type: 'string',
                default: '',
                description: 'A comma-separated list of the fields to include in the response',
            },
        ],
    },
];
const displayOptions = {
    show: {
        resource: ['workbook'],
        operation: ['getAll'],
    },
};
exports.description = (0, utilities_1.updateDisplayOptions)(displayOptions, properties);
async function execute(items) {
    const returnData = [];
    for (let i = 0; i < items.length; i++) {
        try {
            const returnAll = this.getNodeParameter('returnAll', i);
            const filters = this.getNodeParameter('filters', i);
            const qs = {};
            if (filters.fields) {
                qs.$select = filters.fields;
            }
            let responseData;
            if (returnAll) {
                responseData = await transport_1.microsoftApiRequestAllItems.call(this, 'value', 'GET', "/drive/root/search(q='.xlsx')", {}, qs);
            }
            else {
                qs.$top = this.getNodeParameter('limit', i);
                responseData = await transport_1.microsoftApiRequest.call(this, 'GET', "/drive/root/search(q='.xlsx')", {}, qs);
                responseData = responseData.value;
            }
            if (Array.isArray(responseData)) {
                const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                returnData.push(...executionData);
            }
            else if (responseData !== undefined) {
                const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                returnData.push(...executionData);
            }
        }
        catch (error) {
            if (this.continueOnFail()) {
                const executionErrorData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray({ error: error.message }), { itemData: { item: i } });
                returnData.push(...executionErrorData);
                continue;
            }
            throw error;
        }
    }
    return returnData;
}
exports.execute = execute;
