// n8n-nodes-base/dist/nodes/Microsoft/Excel/v2/actions/workbook/Workbook.resource.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.getAll = exports.deleteWorkbook = exports.addWorksheet = void 0;
const addWorksheet = __importStar(require("./addWorksheet.operation"));
exports.addWorksheet = addWorksheet;
const deleteWorkbook = __importStar(require("./deleteWorkbook.operation"));
exports.deleteWorkbook = deleteWorkbook;
const getAll = __importStar(require("./getAll.operation"));
exports.getAll = getAll;
exports.description = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                resource: ['workbook'],
            },
        },
        options: [
            {
                name: 'Add Sheet',
                value: 'addWorksheet',
                description: 'Add a new sheet to the workbook',
                action: 'Add a sheet to a workbook',
            },
            {
                name: 'Delete',
                value: 'deleteWorkbook',
                description: 'Delete workbook',
                action: 'Delete workbook',
            },
            {
                name: 'Get Many',
                value: 'getAll',
                description: 'Get workbooks',
                action: 'Get workbooks',
            },
        ],
        default: 'getAll',
    },
    ...addWorksheet.description,
    ...deleteWorkbook.description,
    ...getAll.description,
];
