// n8n-nodes-base/dist/nodes/Venafi/Datacenter/GenericFunctions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.venafiApiRequestAllItems = exports.venafiApiRequest = void 0;
const lodash_get_1 = __importDefault(require("lodash.get"));
async function venafiApiRequest(method, resource, body = {}, qs = {}, uri, headers = {}) {
    var _a, _b;
    const credentials = (await this.getCredentials('venafiTlsProtectDatacenterApi'));
    const options = {
        headers: {
            'Content-Type': 'application/json',
        },
        method,
        body,
        qs,
        rejectUnauthorized: !credentials.allowUnauthorizedCerts,
        uri: uri || `${credentials.domain}${resource}`,
        json: true,
    };
    try {
        if (Object.keys(headers).length !== 0) {
            options.headers = Object.assign({}, options.headers, headers);
        }
        if (Object.keys(body).length === 0) {
            delete options.body;
        }
        return await this.helpers.requestWithAuthentication.call(this, 'venafiTlsProtectDatacenterApi', options);
    }
    catch (error) {
        if ((_b = (_a = error.response) === null || _a === void 0 ? void 0 : _a.body) === null || _b === void 0 ? void 0 : _b.error) {
            let errors = error.response.body.error.errors;
            errors = errors.map((e) => e.message);
            throw new Error(`Venafi error response [${error.statusCode}]: ${errors.join('|')}`);
        }
        throw error;
    }
}
exports.venafiApiRequest = venafiApiRequest;
async function venafiApiRequestAllItems(propertyName, method, endpoint, body = {}, query = {}) {
    var _a;
    const returnData = [];
    let responseData;
    do {
        responseData = await venafiApiRequest.call(this, method, endpoint, body, query);
        endpoint = (0, lodash_get_1.default)(responseData, '_links[0].Next');
        returnData.push.apply(returnData, responseData[propertyName]);
    } while ((_a = responseData._links) === null || _a === void 0 ? void 0 : _a[0].Next);
    return returnData;
}
exports.venafiApiRequestAllItems = venafiApiRequestAllItems;
