// n8n-nodes-base/dist/nodes/Google/BigQuery/v2/actions/database/Database.resource.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.insert = exports.executeQuery = void 0;
const RLC_description_1 = require("../commonDescriptions/RLC.description");
const insert = __importStar(require("./insert.operation"));
exports.insert = insert;
const executeQuery = __importStar(require("./executeQuery.operation"));
exports.executeQuery = executeQuery;
exports.description = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                resource: ['database'],
            },
        },
        options: [
            {
                name: 'Execute Query',
                value: 'executeQuery',
                description: 'Execute a SQL query',
                action: 'Execute a SQL query',
            },
            {
                name: 'Insert',
                value: 'insert',
                description: 'Insert rows in a table',
                action: 'Insert rows in a table',
            },
        ],
        default: 'executeQuery',
    },
    {
        ...RLC_description_1.projectRLC,
        displayOptions: {
            show: {
                resource: ['database'],
                operation: ['executeQuery', 'insert'],
            },
        },
    },
    {
        ...RLC_description_1.datasetRLC,
        displayOptions: {
            show: {
                resource: ['database'],
                operation: ['insert'],
            },
        },
    },
    {
        ...RLC_description_1.tableRLC,
        displayOptions: {
            show: {
                resource: ['database'],
                operation: ['insert'],
            },
        },
    },
    ...executeQuery.description,
    ...insert.description,
];
