// n8n-nodes-base/dist/nodes/ConvertKit/ConvertKit.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConvertKit = void 0;
const GenericFunctions_1 = require("./GenericFunctions");
const CustomFieldDescription_1 = require("./CustomFieldDescription");
const FormDescription_1 = require("./FormDescription");
const SequenceDescription_1 = require("./SequenceDescription");
const TagDescription_1 = require("./TagDescription");
const TagSubscriberDescription_1 = require("./TagSubscriberDescription");
class ConvertKit {
    constructor() {
        this.description = {
            displayName: 'ConvertKit',
            name: 'convertKit',
            icon: 'file:convertKit.svg',
            group: ['input'],
            version: 1,
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            description: 'Consume ConvertKit API',
            defaults: {
                name: 'ConvertKit',
            },
            inputs: ['main'],
            outputs: ['main'],
            credentials: [
                {
                    name: 'convertKitApi',
                    required: true,
                },
            ],
            properties: [
                {
                    displayName: 'Resource',
                    name: 'resource',
                    type: 'options',
                    noDataExpression: true,
                    options: [
                        {
                            name: 'Custom Field',
                            value: 'customField',
                        },
                        {
                            name: 'Form',
                            value: 'form',
                        },
                        {
                            name: 'Sequence',
                            value: 'sequence',
                        },
                        {
                            name: 'Tag',
                            value: 'tag',
                        },
                        {
                            name: 'Tag Subscriber',
                            value: 'tagSubscriber',
                        },
                    ],
                    default: 'form',
                },
                ...CustomFieldDescription_1.customFieldOperations,
                ...CustomFieldDescription_1.customFieldFields,
                ...FormDescription_1.formOperations,
                ...FormDescription_1.formFields,
                ...SequenceDescription_1.sequenceOperations,
                ...SequenceDescription_1.sequenceFields,
                ...TagDescription_1.tagOperations,
                ...TagDescription_1.tagFields,
                ...TagSubscriberDescription_1.tagSubscriberOperations,
                ...TagSubscriberDescription_1.tagSubscriberFields,
            ],
        };
        this.methods = {
            loadOptions: {
                async getTags() {
                    const returnData = [];
                    const { tags } = await GenericFunctions_1.convertKitApiRequest.call(this, 'GET', '/tags');
                    for (const tag of tags) {
                        const tagName = tag.name;
                        const tagId = tag.id;
                        returnData.push({
                            name: tagName,
                            value: tagId,
                        });
                    }
                    return returnData;
                },
                async getForms() {
                    const returnData = [];
                    const { forms } = await GenericFunctions_1.convertKitApiRequest.call(this, 'GET', '/forms');
                    for (const form of forms) {
                        const formName = form.name;
                        const formId = form.id;
                        returnData.push({
                            name: formName,
                            value: formId,
                        });
                    }
                    return returnData;
                },
                async getSequences() {
                    const returnData = [];
                    const { courses } = await GenericFunctions_1.convertKitApiRequest.call(this, 'GET', '/sequences');
                    for (const course of courses) {
                        const courseName = course.name;
                        const courseId = course.id;
                        returnData.push({
                            name: courseName,
                            value: courseId,
                        });
                    }
                    return returnData;
                },
            },
        };
    }
    async execute() {
        const items = this.getInputData();
        const returnData = [];
        const qs = {};
        let responseData;
        const resource = this.getNodeParameter('resource', 0);
        const operation = this.getNodeParameter('operation', 0);
        for (let i = 0; i < items.length; i++) {
            try {
                if (resource === 'customField') {
                    if (operation === 'create') {
                        const label = this.getNodeParameter('label', i);
                        responseData = await GenericFunctions_1.convertKitApiRequest.call(this, 'POST', '/custom_fields', { label }, qs);
                    }
                    if (operation === 'delete') {
                        const id = this.getNodeParameter('id', i);
                        responseData = await GenericFunctions_1.convertKitApiRequest.call(this, 'DELETE', `/custom_fields/${id}`);
                    }
                    if (operation === 'get') {
                        const id = this.getNodeParameter('id', i);
                        responseData = await GenericFunctions_1.convertKitApiRequest.call(this, 'GET', `/custom_fields/${id}`);
                    }
                    if (operation === 'getAll') {
                        const returnAll = this.getNodeParameter('returnAll', i);
                        responseData = await GenericFunctions_1.convertKitApiRequest.call(this, 'GET', '/custom_fields');
                        responseData = responseData.custom_fields;
                        if (!returnAll) {
                            const limit = this.getNodeParameter('limit', i);
                            responseData = responseData.slice(0, limit);
                        }
                    }
                    if (operation === 'update') {
                        const id = this.getNodeParameter('id', i);
                        const label = this.getNodeParameter('label', i);
                        responseData = await GenericFunctions_1.convertKitApiRequest.call(this, 'PUT', `/custom_fields/${id}`, {
                            label,
                        });
                        responseData = { success: true };
                    }
                }
                if (resource === 'form') {
                    if (operation === 'addSubscriber') {
                        const email = this.getNodeParameter('email', i);
                        const formId = this.getNodeParameter('id', i);
                        const additionalFields = this.getNodeParameter('additionalFields', i);
                        const body = {
                            email,
                        };
                        if (additionalFields.firstName) {
                            body.first_name = additionalFields.firstName;
                        }
                        if (additionalFields.tags) {
                            body.tags = additionalFields.tags;
                        }
                        if (additionalFields.fieldsUi) {
                            const fieldValues = additionalFields.fieldsUi
                                .fieldsValues;
                            if (fieldValues) {
                                body.fields = {};
                                for (const fieldValue of fieldValues) {
                                    body.fields[fieldValue.key] = fieldValue.value;
                                }
                            }
                        }
                        const { subscription } = await GenericFunctions_1.convertKitApiRequest.call(this, 'POST', `/forms/${formId}/subscribe`, body);
                        responseData = subscription;
                    }
                    if (operation === 'getAll') {
                        const returnAll = this.getNodeParameter('returnAll', i);
                        responseData = await GenericFunctions_1.convertKitApiRequest.call(this, 'GET', '/forms');
                        responseData = responseData.forms;
                        if (!returnAll) {
                            const limit = this.getNodeParameter('limit', i);
                            responseData = responseData.slice(0, limit);
                        }
                    }
                    if (operation === 'getSubscriptions') {
                        const formId = this.getNodeParameter('id', i);
                        const returnAll = this.getNodeParameter('returnAll', i);
                        const additionalFields = this.getNodeParameter('additionalFields', i);
                        if (additionalFields.subscriberState) {
                            qs.subscriber_state = additionalFields.subscriberState;
                        }
                        responseData = await GenericFunctions_1.convertKitApiRequest.call(this, 'GET', `/forms/${formId}/subscriptions`, {}, qs);
                        responseData = responseData.subscriptions;
                        if (!returnAll) {
                            const limit = this.getNodeParameter('limit', i);
                            responseData = responseData.slice(0, limit);
                        }
                    }
                }
                if (resource === 'sequence') {
                    if (operation === 'addSubscriber') {
                        const email = this.getNodeParameter('email', i);
                        const sequenceId = this.getNodeParameter('id', i);
                        const additionalFields = this.getNodeParameter('additionalFields', i);
                        const body = {
                            email,
                        };
                        if (additionalFields.firstName) {
                            body.first_name = additionalFields.firstName;
                        }
                        if (additionalFields.tags) {
                            body.tags = additionalFields.tags;
                        }
                        if (additionalFields.fieldsUi) {
                            const fieldValues = additionalFields.fieldsUi
                                .fieldsValues;
                            if (fieldValues) {
                                body.fields = {};
                                for (const fieldValue of fieldValues) {
                                    body.fields[fieldValue.key] = fieldValue.value;
                                }
                            }
                        }
                        const { subscription } = await GenericFunctions_1.convertKitApiRequest.call(this, 'POST', `/sequences/${sequenceId}/subscribe`, body);
                        responseData = subscription;
                    }
                    if (operation === 'getAll') {
                        const returnAll = this.getNodeParameter('returnAll', i);
                        responseData = await GenericFunctions_1.convertKitApiRequest.call(this, 'GET', '/sequences');
                        responseData = responseData.courses;
                        if (!returnAll) {
                            const limit = this.getNodeParameter('limit', i);
                            responseData = responseData.slice(0, limit);
                        }
                    }
                    if (operation === 'getSubscriptions') {
                        const sequenceId = this.getNodeParameter('id', i);
                        const returnAll = this.getNodeParameter('returnAll', i);
                        const additionalFields = this.getNodeParameter('additionalFields', i);
                        if (additionalFields.subscriberState) {
                            qs.subscriber_state = additionalFields.subscriberState;
                        }
                        responseData = await GenericFunctions_1.convertKitApiRequest.call(this, 'GET', `/sequences/${sequenceId}/subscriptions`, {}, qs);
                        responseData = responseData.subscriptions;
                        if (!returnAll) {
                            const limit = this.getNodeParameter('limit', i);
                            responseData = responseData.slice(0, limit);
                        }
                    }
                }
                if (resource === 'tag') {
                    if (operation === 'create') {
                        const names = this.getNodeParameter('name', i)
                            .split(',')
                            .map((e) => ({ name: e }));
                        const body = {
                            tag: names,
                        };
                        responseData = await GenericFunctions_1.convertKitApiRequest.call(this, 'POST', '/tags', body);
                    }
                    if (operation === 'getAll') {
                        const returnAll = this.getNodeParameter('returnAll', i);
                        responseData = await GenericFunctions_1.convertKitApiRequest.call(this, 'GET', '/tags');
                        responseData = responseData.tags;
                        if (!returnAll) {
                            const limit = this.getNodeParameter('limit', i);
                            responseData = responseData.slice(0, limit);
                        }
                    }
                }
                if (resource === 'tagSubscriber') {
                    if (operation === 'add') {
                        const tagId = this.getNodeParameter('tagId', i);
                        const email = this.getNodeParameter('email', i);
                        const additionalFields = this.getNodeParameter('additionalFields', i);
                        const body = {
                            email,
                        };
                        if (additionalFields.firstName) {
                            body.first_name = additionalFields.firstName;
                        }
                        if (additionalFields.fieldsUi) {
                            const fieldValues = additionalFields.fieldsUi
                                .fieldsValues;
                            if (fieldValues) {
                                body.fields = {};
                                for (const fieldValue of fieldValues) {
                                    body.fields[fieldValue.key] = fieldValue.value;
                                }
                            }
                        }
                        const { subscription } = await GenericFunctions_1.convertKitApiRequest.call(this, 'POST', `/tags/${tagId}/subscribe`, body);
                        responseData = subscription;
                    }
                    if (operation === 'getAll') {
                        const tagId = this.getNodeParameter('tagId', i);
                        const returnAll = this.getNodeParameter('returnAll', i);
                        responseData = await GenericFunctions_1.convertKitApiRequest.call(this, 'GET', `/tags/${tagId}/subscriptions`);
                        responseData = responseData.subscriptions;
                        if (!returnAll) {
                            const limit = this.getNodeParameter('limit', i);
                            responseData = responseData.slice(0, limit);
                        }
                    }
                    if (operation === 'delete') {
                        const tagId = this.getNodeParameter('tagId', i);
                        const email = this.getNodeParameter('email', i);
                        responseData = await GenericFunctions_1.convertKitApiRequest.call(this, 'POST', `/tags/${tagId}>/unsubscribe`, { email });
                    }
                }
                const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                returnData.push(...executionData);
            }
            catch (error) {
                if (this.continueOnFail()) {
                    returnData.push({ error: error.message, json: {} });
                    continue;
                }
                throw error;
            }
        }
        return this.prepareOutputData(returnData);
    }
}
exports.ConvertKit = ConvertKit;
