// n8n-nodes-base/dist/nodes/ConvertKit/TagDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.tagFields = exports.tagOperations = void 0;
exports.tagOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                resource: ['tag'],
            },
        },
        options: [
            {
                name: 'Create',
                value: 'create',
                description: 'Create a tag',
                action: 'Create a tag',
            },
            {
                name: 'Get Many',
                value: 'getAll',
                description: 'Get many tags',
                action: 'Get many tags',
            },
        ],
        default: 'create',
    },
];
exports.tagFields = [
    {
        displayName: 'Name',
        name: 'name',
        type: 'string',
        required: true,
        displayOptions: {
            show: {
                resource: ['tag'],
                operation: ['create'],
            },
        },
        default: '',
        description: 'Tag name, multiple can be added separated by comma',
    },
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        displayOptions: {
            show: {
                operation: ['getAll'],
                resource: ['tag'],
            },
        },
        default: false,
        description: 'Whether to return all results or only up to a given limit',
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        displayOptions: {
            show: {
                operation: ['getAll'],
                resource: ['tag'],
                returnAll: [false],
            },
        },
        typeOptions: {
            minValue: 1,
            maxValue: 500,
        },
        default: 100,
        description: 'Max number of results to return',
    },
];
