// n8n-nodes-base/dist/nodes/ItemLists/ItemLists.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ItemLists = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const ItemListsV1_node_1 = require("./V1/ItemListsV1.node");
const ItemListsV2_node_1 = require("./V2/ItemListsV2.node");
class ItemLists extends n8n_workflow_1.VersionedNodeType {
    constructor() {
        const baseDescription = {
            displayName: 'Item Lists',
            name: 'itemLists',
            icon: 'file:itemLists.svg',
            group: ['input'],
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            description: 'Helper for working with lists of items and transforming arrays',
            defaultVersion: 2.1,
        };
        const nodeVersions = {
            1: new ItemListsV1_node_1.ItemListsV1(baseDescription),
            2: new ItemListsV2_node_1.ItemListsV2(baseDescription),
            2.1: new ItemListsV2_node_1.ItemListsV2(baseDescription),
        };
        super(nodeVersions, baseDescription);
    }
}
exports.ItemLists = ItemLists;
