// n8n/dist/databases/entities/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.entities = void 0;
const AuthIdentity_1 = require("./AuthIdentity");
const AuthProviderSyncHistory_1 = require("./AuthProviderSyncHistory");
const CredentialsEntity_1 = require("./CredentialsEntity");
const EventDestinations_1 = require("./EventDestinations");
const ExecutionEntity_1 = require("./ExecutionEntity");
const InstalledNodes_1 = require("./InstalledNodes");
const InstalledPackages_1 = require("./InstalledPackages");
const Role_1 = require("./Role");
const Settings_1 = require("./Settings");
const SharedCredentials_1 = require("./SharedCredentials");
const SharedWorkflow_1 = require("./SharedWorkflow");
const TagEntity_1 = require("./TagEntity");
const User_1 = require("./User");
const Variables_1 = require("./Variables");
const WebhookEntity_1 = require("./WebhookEntity");
const WorkflowEntity_1 = require("./WorkflowEntity");
const WorkflowTagMapping_1 = require("./WorkflowTagMapping");
const WorkflowStatistics_1 = require("./WorkflowStatistics");
const ExecutionMetadata_1 = require("./ExecutionMetadata");
exports.entities = {
    AuthIdentity: AuthIdentity_1.AuthIdentity,
    AuthProviderSyncHistory: AuthProviderSyncHistory_1.AuthProviderSyncHistory,
    CredentialsEntity: CredentialsEntity_1.CredentialsEntity,
    EventDestinations: EventDestinations_1.EventDestinations,
    ExecutionEntity: ExecutionEntity_1.ExecutionEntity,
    InstalledNodes: InstalledNodes_1.InstalledNodes,
    InstalledPackages: InstalledPackages_1.InstalledPackages,
    Role: Role_1.Role,
    Settings: Settings_1.Settings,
    SharedCredentials: SharedCredentials_1.SharedCredentials,
    SharedWorkflow: SharedWorkflow_1.SharedWorkflow,
    TagEntity: TagEntity_1.TagEntity,
    User: User_1.User,
    Variables: Variables_1.Variables,
    WebhookEntity: WebhookEntity_1.WebhookEntity,
    WorkflowEntity: WorkflowEntity_1.WorkflowEntity,
    WorkflowTagMapping: WorkflowTagMapping_1.WorkflowTagMapping,
    WorkflowStatistics: WorkflowStatistics_1.WorkflowStatistics,
    ExecutionMetadata: ExecutionMetadata_1.ExecutionMetadata,
};
