// n8n/dist/databases/entities/WorkflowStatistics.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkflowStatistics = exports.StatisticsNames = void 0;
const typeorm_1 = require("typeorm");
const transformers_1 = require("../utils/transformers");
const AbstractEntity_1 = require("./AbstractEntity");
const WorkflowEntity_1 = require("./WorkflowEntity");
var StatisticsNames;
(function (StatisticsNames) {
    StatisticsNames["productionSuccess"] = "production_success";
    StatisticsNames["productionError"] = "production_error";
    StatisticsNames["manualSuccess"] = "manual_success";
    StatisticsNames["manualError"] = "manual_error";
    StatisticsNames["dataLoaded"] = "data_loaded";
})(StatisticsNames = exports.StatisticsNames || (exports.StatisticsNames = {}));
let WorkflowStatistics = class WorkflowStatistics {
};
__decorate([
    (0, typeorm_1.Column)(),
    __metadata("design:type", Number)
], WorkflowStatistics.prototype, "count", void 0);
__decorate([
    (0, typeorm_1.Column)(AbstractEntity_1.datetimeColumnType),
    __metadata("design:type", Date)
], WorkflowStatistics.prototype, "latestEvent", void 0);
__decorate([
    (0, typeorm_1.PrimaryColumn)({ length: 128 }),
    __metadata("design:type", String)
], WorkflowStatistics.prototype, "name", void 0);
__decorate([
    (0, typeorm_1.ManyToOne)('WorkflowEntity', 'shared'),
    __metadata("design:type", WorkflowEntity_1.WorkflowEntity)
], WorkflowStatistics.prototype, "workflow", void 0);
__decorate([
    (0, typeorm_1.PrimaryColumn)({ transformer: transformers_1.idStringifier }),
    __metadata("design:type", String)
], WorkflowStatistics.prototype, "workflowId", void 0);
WorkflowStatistics = __decorate([
    (0, typeorm_1.Entity)()
], WorkflowStatistics);
exports.WorkflowStatistics = WorkflowStatistics;
