// n8n-nodes-base/dist/nodes/ExecuteWorkflowTrigger/ExecuteWorkflowTrigger.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExecuteWorkflowTrigger = void 0;
class ExecuteWorkflowTrigger {
    constructor() {
        this.description = {
            displayName: 'Execute Workflow Trigger',
            name: 'executeWorkflowTrigger',
            icon: 'fa:sign-out-alt',
            group: ['trigger'],
            version: 1,
            description: 'Helpers for calling other n8n workflows. Used for designing modular, microservice-like workflows.',
            eventTriggerDescription: '',
            maxNodes: 1,
            defaults: {
                name: 'Execute Workflow Trigger',
                color: '#ff6d5a',
            },
            inputs: [],
            outputs: ['main'],
            properties: [
                {
                    displayName: "When an ‘execute workflow’ node calls this workflow, the execution starts here. Any data passed into the 'execute workflow' node will be output by this node.",
                    name: 'notice',
                    type: 'notice',
                    default: '',
                },
                {
                    displayName: 'Events',
                    name: 'events',
                    type: 'hidden',
                    noDataExpression: true,
                    options: [
                        {
                            name: 'Workflow Call',
                            value: 'worklfow_call',
                            description: 'When called by another workflow using Execute Workflow Trigger',
                            action: 'When Called by Another Workflow',
                        },
                    ],
                    default: 'worklfow_call',
                },
            ],
        };
    }
    async execute() {
        const items = this.getInputData();
        return this.prepareOutputData(items);
    }
}
exports.ExecuteWorkflowTrigger = ExecuteWorkflowTrigger;
