// n8n-nodes-base/dist/nodes/Freshdesk/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.capitalize = exports.validateJSON = exports.freshdeskApiRequestAllItems = exports.freshdeskApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function freshdeskApiRequest(method, resource, body = {}, query = {}, uri, option = {}) {
    const credentials = await this.getCredentials('freshdeskApi');
    const apiKey = `${credentials.apiKey}:X`;
    const endpoint = 'freshdesk.com/api/v2';
    let options = {
        headers: {
            'Content-Type': 'application/json',
            Authorization: `${Buffer.from(apiKey).toString(n8n_workflow_1.BINARY_ENCODING)}`,
        },
        method,
        body,
        qs: query,
        uri: uri || `https://${credentials.domain}.${endpoint}${resource}`,
        json: true,
    };
    if (!Object.keys(body).length) {
        delete options.body;
    }
    if (!Object.keys(query).length) {
        delete options.qs;
    }
    options = Object.assign({}, options, option);
    try {
        return await this.helpers.request(options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.freshdeskApiRequest = freshdeskApiRequest;
async function freshdeskApiRequestAllItems(method, endpoint, body = {}, query = {}) {
    var _a;
    const returnData = [];
    let responseData;
    let uri;
    query.per_page = 100;
    do {
        responseData = await freshdeskApiRequest.call(this, method, endpoint, body, query, uri, {
            resolveWithFullResponse: true,
        });
        if (responseData.headers.link) {
            uri = responseData.headers.link.split(';')[0].replace('<', '').replace('>', '');
        }
        returnData.push.apply(returnData, responseData.body);
    } while ((_a = responseData.headers.link) === null || _a === void 0 ? void 0 : _a.includes('rel="next"'));
    return returnData;
}
exports.freshdeskApiRequestAllItems = freshdeskApiRequestAllItems;
function validateJSON(json) {
    let result;
    try {
        result = JSON.parse(json);
    }
    catch (exception) {
        result = [];
    }
    return result;
}
exports.validateJSON = validateJSON;
function capitalize(s) {
    if (typeof s !== 'string')
        return '';
    return s.charAt(0).toUpperCase() + s.slice(1);
}
exports.capitalize = capitalize;
