// n8n-nodes-base/dist/nodes/Google/Ads/GoogleAds.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoogleAds = void 0;
const CampaignDescription_1 = require("./CampaignDescription");
class GoogleAds {
    constructor() {
        this.description = {
            displayName: 'Google Ads',
            name: 'googleAds',
            icon: 'file:googleAds.svg',
            group: ['transform'],
            version: 1,
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            description: 'Use the Google Ads API',
            defaults: {
                name: 'Google Ads',
            },
            inputs: ['main'],
            outputs: ['main'],
            credentials: [
                {
                    name: 'googleAdsOAuth2Api',
                    required: true,
                    testedBy: {
                        request: {
                            method: 'GET',
                            url: '/v11/customers:listAccessibleCustomers',
                        },
                    },
                },
            ],
            requestDefaults: {
                returnFullResponse: true,
                baseURL: 'https://googleads.googleapis.com',
                headers: {
                    'developer-token': '={{$credentials.developerToken}}',
                },
            },
            properties: [
                {
                    displayName: 'Resource',
                    name: 'resource',
                    type: 'options',
                    noDataExpression: true,
                    options: [
                        {
                            name: 'Campaign',
                            value: 'campaign',
                        },
                    ],
                    default: 'campaign',
                },
                ...CampaignDescription_1.campaignOperations,
                {
                    displayName: 'Divide field names expressed with <i>micros</i> by 1,000,000 to get the actual value',
                    name: 'campaigsNotice',
                    type: 'notice',
                    default: '',
                    displayOptions: {
                        show: {
                            resource: ['campaign'],
                        },
                    },
                },
                ...CampaignDescription_1.campaignFields,
            ],
        };
    }
}
exports.GoogleAds = GoogleAds;
