// n8n/dist/eventbus/MessageEventBusDestination/MessageEventBusDestinationSentry.ee.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageEventBusDestinationSentry = exports.isMessageEventBusDestinationSentryOptions = void 0;
const MessageEventBusDestination_ee_1 = require("./MessageEventBusDestination.ee");
const Sentry = __importStar(require("@sentry/node"));
const n8n_workflow_1 = require("n8n-workflow");
const MessageEventBusHelper_1 = require("../MessageEventBus/MessageEventBusHelper");
const EventMessageGeneric_1 = require("../EventMessageClasses/EventMessageGeneric");
const constants_1 = require("../../constants");
const isMessageEventBusDestinationSentryOptions = (candidate) => {
    const o = candidate;
    if (!o)
        return false;
    return o.dsn !== undefined;
};
exports.isMessageEventBusDestinationSentryOptions = isMessageEventBusDestinationSentryOptions;
class MessageEventBusDestinationSentry extends MessageEventBusDestination_ee_1.MessageEventBusDestination {
    constructor(eventBusInstance, options) {
        var _a, _b;
        super(eventBusInstance, options);
        this.tracesSampleRate = 1.0;
        this.label = (_a = options.label) !== null && _a !== void 0 ? _a : 'Sentry DSN';
        this.__type = (_b = options.__type) !== null && _b !== void 0 ? _b : "$$MessageEventBusDestinationSentry";
        this.dsn = options.dsn;
        if (options.sendPayload)
            this.sendPayload = options.sendPayload;
        if (options.tracesSampleRate)
            this.tracesSampleRate = options.tracesSampleRate;
        const { ENVIRONMENT: environment } = process.env;
        this.sentryClient = new Sentry.NodeClient({
            dsn: this.dsn,
            tracesSampleRate: this.tracesSampleRate,
            environment,
            release: constants_1.N8N_VERSION,
            transport: Sentry.makeNodeTransport,
            integrations: Sentry.defaultIntegrations,
            stackParser: Sentry.defaultStackParser,
        });
    }
    async receiveFromEventBus(emitterPayload) {
        var _a, _b;
        const { msg, confirmCallback } = emitterPayload;
        let sendResult = false;
        if (!this.sentryClient)
            return sendResult;
        if (msg.eventName !== EventMessageGeneric_1.eventMessageGenericDestinationTestEvent) {
            if (!(0, MessageEventBusHelper_1.isLogStreamingEnabled)())
                return sendResult;
            if (!this.hasSubscribedToEvent(msg))
                return sendResult;
        }
        try {
            const payload = this.anonymizeAuditMessages ? msg.anonymize() : msg.payload;
            const scope = new Sentry.Scope();
            const level = (msg.eventName.toLowerCase().endsWith('error') ? 'error' : 'log');
            scope.setLevel(level);
            scope.setTags({
                event: msg.getEventName(),
                logger: (_a = this.label) !== null && _a !== void 0 ? _a : this.getId(),
                app: 'n8n',
            });
            if (this.sendPayload) {
                scope.setExtras(payload);
            }
            const sentryResult = this.sentryClient.captureMessage((_b = msg.message) !== null && _b !== void 0 ? _b : msg.eventName, level, { event_id: msg.id, data: payload }, scope);
            if (sentryResult) {
                confirmCallback(msg, { id: this.id, name: this.label });
                sendResult = true;
            }
        }
        catch (error) {
            if (error.message)
                n8n_workflow_1.LoggerProxy.debug(error.message);
        }
        return sendResult;
    }
    serialize() {
        const abstractSerialized = super.serialize();
        return {
            ...abstractSerialized,
            dsn: this.dsn,
            tracesSampleRate: this.tracesSampleRate,
            sendPayload: this.sendPayload,
        };
    }
    static deserialize(eventBusInstance, data) {
        if ('__type' in data &&
            data.__type === "$$MessageEventBusDestinationSentry" &&
            (0, exports.isMessageEventBusDestinationSentryOptions)(data)) {
            return new MessageEventBusDestinationSentry(eventBusInstance, data);
        }
        return null;
    }
    toString() {
        return JSON.stringify(this.serialize());
    }
    async close() {
        var _a;
        await super.close();
        await ((_a = this.sentryClient) === null || _a === void 0 ? void 0 : _a.close());
    }
}
exports.MessageEventBusDestinationSentry = MessageEventBusDestinationSentry;
