// n8n-nodes-base/dist/nodes/QuickBooks/descriptions/Item/ItemDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.itemFields = exports.itemOperations = void 0;
exports.itemOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        default: 'get',
        options: [
            {
                name: 'Get',
                value: 'get',
                action: 'Get an item',
            },
            {
                name: 'Get Many',
                value: 'getAll',
                action: 'Get many items',
            },
        ],
        displayOptions: {
            show: {
                resource: ['item'],
            },
        },
    },
];
exports.itemFields = [
    {
        displayName: 'Item ID',
        name: 'itemId',
        type: 'string',
        required: true,
        default: '',
        description: 'The ID of the item to retrieve',
        displayOptions: {
            show: {
                resource: ['item'],
                operation: ['get'],
            },
        },
    },
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        default: false,
        description: 'Whether to return all results or only up to a given limit',
        displayOptions: {
            show: {
                resource: ['item'],
                operation: ['getAll'],
            },
        },
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        default: 50,
        description: 'Max number of results to return',
        typeOptions: {
            minValue: 1,
            maxValue: 1000,
        },
        displayOptions: {
            show: {
                resource: ['item'],
                operation: ['getAll'],
                returnAll: [false],
            },
        },
    },
    {
        displayName: 'Filters',
        name: 'filters',
        type: 'collection',
        placeholder: 'Add Field',
        default: {},
        options: [
            {
                displayName: 'Query',
                name: 'query',
                type: 'string',
                default: '',
                placeholder: "WHERE Metadata.LastUpdatedTime > '2021-01-01'",
                description: 'The condition for selecting items. See the <a href="https://developer.intuit.com/app/developer/qbo/docs/develop/explore-the-quickbooks-online-api/data-queries">guide</a> for supported syntax.',
            },
        ],
        displayOptions: {
            show: {
                resource: ['item'],
                operation: ['getAll'],
            },
        },
    },
];
