// n8n-nodes-base/dist/nodes/Google/BigQuery/v2/actions/versionDescription.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.versionDescription = void 0;
const database = __importStar(require("./database/Database.resource"));
exports.versionDescription = {
    displayName: 'Google BigQuery',
    name: 'googleBigQuery',
    icon: 'file:googleBigQuery.svg',
    group: ['input'],
    version: 2,
    subtitle: '={{$parameter["operation"]}}',
    description: 'Consume Google BigQuery API',
    defaults: {
        name: 'Google BigQuery',
    },
    inputs: ['main'],
    outputs: ['main'],
    credentials: [
        {
            name: 'googleApi',
            required: true,
            displayOptions: {
                show: {
                    authentication: ['serviceAccount'],
                },
            },
        },
        {
            name: 'googleBigQueryOAuth2Api',
            required: true,
            displayOptions: {
                show: {
                    authentication: ['oAuth2'],
                },
            },
        },
    ],
    properties: [
        {
            displayName: 'Authentication',
            name: 'authentication',
            type: 'options',
            noDataExpression: true,
            options: [
                {
                    name: 'OAuth2 (recommended)',
                    value: 'oAuth2',
                },
                {
                    name: 'Service Account',
                    value: 'serviceAccount',
                },
            ],
            default: 'oAuth2',
        },
        {
            displayName: 'Resource',
            name: 'resource',
            type: 'hidden',
            noDataExpression: true,
            options: [
                {
                    name: 'Database',
                    value: 'database',
                },
            ],
            default: 'database',
        },
        ...database.description,
    ],
};
