// n8n-nodes-base/dist/nodes/Baserow/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TableFieldMapper = exports.toOptions = exports.getFieldNamesAndIds = exports.getJwtToken = exports.baserowApiRequestAllItems = exports.baserowApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function baserowApiRequest(method, endpoint, jwtToken, body = {}, qs = {}) {
    const credentials = (await this.getCredentials('baserowApi'));
    const options = {
        headers: {
            Authorization: `JWT ${jwtToken}`,
        },
        method,
        body,
        qs,
        uri: `${credentials.host}${endpoint}`,
        json: true,
    };
    if (Object.keys(qs).length === 0) {
        delete options.qs;
    }
    if (Object.keys(body).length === 0) {
        delete options.body;
    }
    try {
        return await this.helpers.request(options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.baserowApiRequest = baserowApiRequest;
async function baserowApiRequestAllItems(method, endpoint, jwtToken, body, qs = {}) {
    const returnData = [];
    let responseData;
    qs.page = 1;
    qs.size = 100;
    const returnAll = this.getNodeParameter('returnAll', 0, false);
    const limit = this.getNodeParameter('limit', 0, 0);
    do {
        responseData = await baserowApiRequest.call(this, method, endpoint, jwtToken, body, qs);
        returnData.push(...responseData.results);
        if (!returnAll && returnData.length > limit) {
            return returnData.slice(0, limit);
        }
        qs.page += 1;
    } while (responseData.next !== null);
    return returnData;
}
exports.baserowApiRequestAllItems = baserowApiRequestAllItems;
async function getJwtToken({ username, password, host }) {
    const options = {
        method: 'POST',
        body: {
            username,
            password,
        },
        uri: `${host}/api/user/token-auth/`,
        json: true,
    };
    try {
        const { token } = (await this.helpers.request(options));
        return token;
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.getJwtToken = getJwtToken;
async function getFieldNamesAndIds(tableId, jwtToken) {
    const endpoint = `/api/database/fields/table/${tableId}/`;
    const response = (await baserowApiRequest.call(this, 'GET', endpoint, jwtToken));
    return {
        names: response.map((field) => field.name),
        ids: response.map((field) => `field_${field.id}`),
    };
}
exports.getFieldNamesAndIds = getFieldNamesAndIds;
const toOptions = (items) => items.map(({ name, id }) => ({ name, value: id }));
exports.toOptions = toOptions;
class TableFieldMapper {
    constructor() {
        this.nameToIdMapping = {};
        this.idToNameMapping = {};
        this.mapIds = true;
    }
    async getTableFields(table, jwtToken) {
        const endpoint = `/api/database/fields/table/${table}/`;
        return baserowApiRequest.call(this, 'GET', endpoint, jwtToken);
    }
    createMappings(tableFields) {
        this.nameToIdMapping = this.createNameToIdMapping(tableFields);
        this.idToNameMapping = this.createIdToNameMapping(tableFields);
    }
    createIdToNameMapping(responseData) {
        return responseData.reduce((acc, cur) => {
            acc[`field_${cur.id}`] = cur.name;
            return acc;
        }, {});
    }
    createNameToIdMapping(responseData) {
        return responseData.reduce((acc, cur) => {
            acc[cur.name] = `field_${cur.id}`;
            return acc;
        }, {});
    }
    setField(field) {
        var _a;
        return this.mapIds ? field : (_a = this.nameToIdMapping[field]) !== null && _a !== void 0 ? _a : field;
    }
    idsToNames(obj) {
        Object.entries(obj).forEach(([key, value]) => {
            if (this.idToNameMapping[key] !== undefined) {
                delete obj[key];
                obj[this.idToNameMapping[key]] = value;
            }
        });
    }
    namesToIds(obj) {
        Object.entries(obj).forEach(([key, value]) => {
            if (this.nameToIdMapping[key] !== undefined) {
                delete obj[key];
                obj[this.nameToIdMapping[key]] = value;
            }
        });
    }
}
exports.TableFieldMapper = TableFieldMapper;
