// n8n-nodes-base/dist/nodes/Aws/CertificateManager/GenericFunctions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.awsApiRequestAllItems = exports.awsApiRequestREST = exports.awsApiRequest = void 0;
const lodash_get_1 = __importDefault(require("lodash.get"));
const n8n_workflow_1 = require("n8n-workflow");
async function awsApiRequest(service, method, path, body, query = {}, headers) {
    const credentials = await this.getCredentials('aws');
    const requestOptions = {
        qs: {
            service,
            path,
            ...query,
        },
        headers,
        method,
        url: '',
        body,
        region: credentials === null || credentials === void 0 ? void 0 : credentials.region,
    };
    try {
        return await this.helpers.requestWithAuthentication.call(this, 'aws', requestOptions);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.awsApiRequest = awsApiRequest;
async function awsApiRequestREST(service, method, path, body, query = {}, headers) {
    const response = await awsApiRequest.call(this, service, method, path, body, query, headers);
    try {
        return JSON.parse(response);
    }
    catch (e) {
        return response;
    }
}
exports.awsApiRequestREST = awsApiRequestREST;
async function awsApiRequestAllItems(propertyName, service, method, path, body, query = {}, headers = {}) {
    const returnData = [];
    let responseData;
    do {
        responseData = await awsApiRequestREST.call(this, service, method, path, body, query, headers);
        if (responseData.NextToken) {
            const data = (0, n8n_workflow_1.jsonParse)(body, {
                errorMessage: 'Response body is not valid JSON',
            });
            data.NextToken = responseData.NextToken;
        }
        returnData.push.apply(returnData, (0, lodash_get_1.default)(responseData, propertyName));
    } while (responseData.NextToken !== undefined);
    return returnData;
}
exports.awsApiRequestAllItems = awsApiRequestAllItems;
