// n8n-nodes-base/dist/nodes/Misp/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.throwOnInvalidUrl = exports.throwOnMissingSharingGroup = exports.throwOnEmptyUpdate = exports.mispApiRequestAllItems = exports.mispApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const url_1 = require("url");
async function mispApiRequest(method, endpoint, body = {}, qs = {}) {
    var _a;
    const { baseUrl, apiKey, allowUnauthorizedCerts } = (await this.getCredentials('mispApi'));
    const options = {
        headers: {
            Authorization: apiKey,
        },
        method,
        body,
        qs,
        uri: `${baseUrl}${endpoint}`,
        json: true,
        rejectUnauthorized: !allowUnauthorizedCerts,
    };
    if (!Object.keys(body).length) {
        delete options.body;
    }
    if (!Object.keys(qs).length) {
        delete options.qs;
    }
    try {
        return await this.helpers.request(options);
    }
    catch (error) {
        if (error.statusCode === 403) {
            error.statusCode = 400;
        }
        const errors = (_a = error === null || error === void 0 ? void 0 : error.error) === null || _a === void 0 ? void 0 : _a.errors;
        if (errors) {
            const key = Object.keys(errors)[0];
            if (key !== undefined) {
                let message = errors[key].join();
                if (message.includes(' nter')) {
                    message = message.replace(' nter', ' enter');
                }
                error.error.message = `${error.error.message}: ${message}`;
            }
        }
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.mispApiRequest = mispApiRequest;
async function mispApiRequestAllItems(endpoint) {
    const responseData = await mispApiRequest.call(this, 'GET', endpoint);
    const returnAll = this.getNodeParameter('returnAll', 0);
    if (!returnAll) {
        const limit = this.getNodeParameter('limit', 0);
        return responseData.slice(0, limit);
    }
    return responseData;
}
exports.mispApiRequestAllItems = mispApiRequestAllItems;
function throwOnEmptyUpdate(resource, updateFields) {
    if (!Object.keys(updateFields).length) {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Please enter at least one field to update for the ${resource}.`);
    }
}
exports.throwOnEmptyUpdate = throwOnEmptyUpdate;
const SHARING_GROUP_OPTION_ID = 4;
function throwOnMissingSharingGroup(fields) {
    if (fields.distribution === SHARING_GROUP_OPTION_ID && !fields.sharing_group_id) {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Please specify a sharing group');
    }
}
exports.throwOnMissingSharingGroup = throwOnMissingSharingGroup;
const isValidUrl = (str) => {
    try {
        new url_1.URL(str);
        return true;
    }
    catch (error) {
        return false;
    }
};
function throwOnInvalidUrl(str) {
    if (!isValidUrl(str)) {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Please specify a valid URL, protocol included. Example: https://site.com');
    }
}
exports.throwOnInvalidUrl = throwOnInvalidUrl;
