// n8n-nodes-base/dist/nodes/TheHive/TheHiveTrigger.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TheHiveTrigger = void 0;
const EventsDescription_1 = require("./descriptions/EventsDescription");
class TheHiveTrigger {
    constructor() {
        this.description = {
            displayName: 'TheHive Trigger',
            name: 'theHiveTrigger',
            icon: 'file:thehive.svg',
            group: ['trigger'],
            version: [1, 2],
            description: 'Starts the workflow when TheHive events occur',
            defaults: {
                name: 'TheHive Trigger',
            },
            inputs: [],
            outputs: ['main'],
            webhooks: [
                {
                    name: 'default',
                    httpMethod: 'POST',
                    responseMode: 'onReceived',
                    path: 'webhook',
                },
            ],
            properties: [...EventsDescription_1.eventsDescription],
        };
        this.webhookMethods = {
            default: {
                async checkExists() {
                    return true;
                },
                async create() {
                    return true;
                },
                async delete() {
                    return true;
                },
            },
        };
    }
    async webhook() {
        const bodyData = this.getBodyData();
        const events = this.getNodeParameter('events', []);
        if (!bodyData.operation || !bodyData.objectType) {
            return {};
        }
        const operation = bodyData.operation.replace('Creation', 'Create');
        const event = `${bodyData.objectType.toLowerCase()}_${operation.toLowerCase()}`;
        if (events.indexOf('*') === -1 && events.indexOf(event) === -1) {
            return {};
        }
        const returnData = [];
        returnData.push({
            event,
            body: this.getBodyData(),
            headers: this.getHeaderData(),
            query: this.getQueryData(),
        });
        return {
            workflowData: [this.helpers.returnJsonArray(returnData)],
        };
    }
}
exports.TheHiveTrigger = TheHiveTrigger;
