// n8n-nodes-base/dist/nodes/BambooHr/v1/transport/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.apiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function apiRequest(method, endpoint, body = {}, query = {}, option = {}) {
    var _a;
    const credentials = await this.getCredentials('bambooHrApi');
    const apiKey = credentials.apiKey;
    const subdomain = credentials.subdomain;
    const uri = `https://api.bamboohr.com/api/gateway.php/${subdomain}/v1/${endpoint}`;
    const options = {
        method,
        body,
        qs: query,
        url: uri,
        auth: {
            username: apiKey,
            password: 'x',
        },
        json: true,
    };
    if (Object.keys(option).length) {
        Object.assign(options, option);
    }
    if (!Object.keys(body).length) {
        delete options.body;
    }
    if (!Object.keys(query).length) {
        delete options.qs;
    }
    try {
        return await this.helpers.request(options);
    }
    catch (error) {
        const description = ((_a = error === null || error === void 0 ? void 0 : error.response) === null || _a === void 0 ? void 0 : _a.headers['x-bamboohr-error-messsage']) || '';
        const message = (error === null || error === void 0 ? void 0 : error.message) || '';
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error, { message, description });
    }
}
exports.apiRequest = apiRequest;
