// n8n-nodes-base/dist/nodes/Iterable/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.iterableApiRequestAllItems = exports.iterableApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function iterableApiRequest(method, resource, body = {}, qs = {}, uri, headers = {}) {
    const credentials = await this.getCredentials('iterableApi');
    const options = {
        headers: {
            'Content-Type': 'application/json',
            Api_Key: credentials.apiKey,
        },
        method,
        body,
        qs,
        uri: uri || `https://api.iterable.com/api${resource}`,
        json: true,
    };
    try {
        if (Object.keys(headers).length !== 0) {
            options.headers = Object.assign({}, options.headers, headers);
        }
        if (Object.keys(body).length === 0) {
            delete options.body;
        }
        return await this.helpers.request.call(this, options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.iterableApiRequest = iterableApiRequest;
async function iterableApiRequestAllItems(propertyName, method, endpoint, body = {}, query = {}) {
    const returnData = [];
    let responseData;
    query.maxResults = 100;
    do {
        responseData = await iterableApiRequest.call(this, method, endpoint, body, query);
        query.pageToken = responseData.nextPageToken;
        returnData.push.apply(returnData, responseData[propertyName]);
    } while (responseData.nextPageToken !== undefined && responseData.nextPageToken !== '');
    return returnData;
}
exports.iterableApiRequestAllItems = iterableApiRequestAllItems;
