// n8n-nodes-base/dist/nodes/Microsoft/Excel/v2/actions/worksheet/Worksheet.resource.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.description = exports.upsert = exports.update = exports.readRows = exports.getAll = exports.deleteWorksheet = exports.clear = exports.append = void 0;
const append = __importStar(require("./append.operation"));
exports.append = append;
const clear = __importStar(require("./clear.operation"));
exports.clear = clear;
const deleteWorksheet = __importStar(require("./deleteWorksheet.operation"));
exports.deleteWorksheet = deleteWorksheet;
const getAll = __importStar(require("./getAll.operation"));
exports.getAll = getAll;
const readRows = __importStar(require("./readRows.operation"));
exports.readRows = readRows;
const update = __importStar(require("./update.operation"));
exports.update = update;
const upsert = __importStar(require("./upsert.operation"));
exports.upsert = upsert;
exports.description = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                resource: ['worksheet'],
            },
        },
        options: [
            {
                name: 'Append',
                value: 'append',
                description: 'Append data to sheet',
                action: 'Append data to sheet',
            },
            {
                name: 'Append or Update',
                value: 'upsert',
                description: 'Append a new row or update the current one if it already exists (upsert)',
                action: 'Append or update a sheet',
            },
            {
                name: 'Clear',
                value: 'clear',
                description: 'Clear sheet',
                action: 'Clear sheet',
            },
            {
                name: 'Delete',
                value: 'deleteWorksheet',
                description: 'Delete sheet',
                action: 'Delete sheet',
            },
            {
                name: 'Get Many',
                value: 'getAll',
                description: 'Get a list of sheets',
                action: 'Get sheets',
            },
            {
                name: 'Get Rows',
                value: 'readRows',
                description: 'Retrieve a list of sheet rows',
                action: 'Get rows from sheet',
            },
            {
                name: 'Update',
                value: 'update',
                description: 'Update rows of a sheet or sheet range',
                action: 'Update sheet',
            },
        ],
        default: 'getAll',
    },
    ...append.description,
    ...clear.description,
    ...deleteWorksheet.description,
    ...getAll.description,
    ...readRows.description,
    ...update.description,
    ...upsert.description,
];
