// n8n-nodes-base/dist/nodes/Postgres/v2/methods/loadOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getColumnsWithoutColumnToMatchOn = exports.getColumnsMultiOptions = exports.getColumns = void 0;
const utils_1 = require("../helpers/utils");
const transport_1 = require("../transport");
async function getColumns() {
    const credentials = await this.getCredentials('postgres');
    const options = { nodeVersion: this.getNode().typeVersion };
    const { db } = (await transport_1.Connections.getInstance(credentials, options));
    const schema = this.getNodeParameter('schema', 0, {
        extractValue: true,
    });
    const table = this.getNodeParameter('table', 0, {
        extractValue: true,
    });
    try {
        const columns = await (0, utils_1.getTableSchema)(db, schema, table);
        return columns.map((column) => ({
            name: column.column_name,
            value: column.column_name,
            description: `Type: ${column.data_type.toUpperCase()}, Nullable: ${column.is_nullable}`,
        }));
    }
    catch (error) {
        throw error;
    }
}
exports.getColumns = getColumns;
async function getColumnsMultiOptions() {
    const returnData = await getColumns.call(this);
    const returnAll = { name: '*', value: '*', description: 'All columns' };
    return [returnAll, ...returnData];
}
exports.getColumnsMultiOptions = getColumnsMultiOptions;
async function getColumnsWithoutColumnToMatchOn() {
    const columnToMatchOn = this.getNodeParameter('columnToMatchOn');
    const returnData = await getColumns.call(this);
    return returnData.filter((column) => column.value !== columnToMatchOn);
}
exports.getColumnsWithoutColumnToMatchOn = getColumnsWithoutColumnToMatchOn;
