// n8n-nodes-base/dist/nodes/EmailReadImap/v1/EmailReadImapV1.node.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EmailReadImapV1 = exports.parseRawEmail = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const imap_simple_1 = require("imap-simple");
const mailparser_1 = require("mailparser");
const lodash_isempty_1 = __importDefault(require("lodash.isempty"));
const lodash_find_1 = __importDefault(require("lodash.find"));
async function parseRawEmail(messageEncoded, dataPropertyNameDownload) {
    const responseData = await (0, mailparser_1.simpleParser)(messageEncoded);
    const headers = {};
    for (const header of responseData.headerLines) {
        headers[header.key] = header.line;
    }
    responseData.headers = headers;
    responseData.headerLines = undefined;
    const binaryData = {};
    if (responseData.attachments) {
        for (let i = 0; i < responseData.attachments.length; i++) {
            const attachment = responseData.attachments[i];
            binaryData[`${dataPropertyNameDownload}${i}`] = await this.helpers.prepareBinaryData(attachment.content, attachment.filename, attachment.contentType);
        }
        responseData.attachments = undefined;
    }
    return {
        json: responseData,
        binary: Object.keys(binaryData).length ? binaryData : undefined,
    };
}
exports.parseRawEmail = parseRawEmail;
const versionDescription = {
    displayName: 'Email Trigger (IMAP)',
    name: 'emailReadImap',
    icon: 'fa:inbox',
    group: ['trigger'],
    version: 1,
    description: 'Triggers the workflow when a new email is received',
    eventTriggerDescription: 'Waiting for you to receive an email',
    defaults: {
        name: 'Email Trigger (IMAP)',
        color: '#44AA22',
    },
    inputs: [],
    outputs: ['main'],
    credentials: [
        {
            name: 'imap',
            required: true,
            testedBy: 'imapConnectionTest',
        },
    ],
    properties: [
        {
            displayName: 'Mailbox Name',
            name: 'mailbox',
            type: 'string',
            default: 'INBOX',
        },
        {
            displayName: 'Action',
            name: 'postProcessAction',
            type: 'options',
            options: [
                {
                    name: 'Mark as Read',
                    value: 'read',
                },
                {
                    name: 'Nothing',
                    value: 'nothing',
                },
            ],
            default: 'read',
            description: 'What to do after the email has been received. If "nothing" gets selected it will be processed multiple times.',
        },
        {
            displayName: 'Download Attachments',
            name: 'downloadAttachments',
            type: 'boolean',
            default: false,
            displayOptions: {
                show: {
                    format: ['simple'],
                },
            },
            description: 'Whether attachments of emails should be downloaded. Only set if needed as it increases processing.',
        },
        {
            displayName: 'Format',
            name: 'format',
            type: 'options',
            options: [
                {
                    name: 'RAW',
                    value: 'raw',
                    description: 'Returns the full email message data with body content in the raw field as a base64url encoded string; the payload field is not used',
                },
                {
                    name: 'Resolved',
                    value: 'resolved',
                    description: 'Returns the full email with all data resolved and attachments saved as binary data',
                },
                {
                    name: 'Simple',
                    value: 'simple',
                    description: 'Returns the full email; do not use if you wish to gather inline attachments',
                },
            ],
            default: 'simple',
            description: 'The format to return the message in',
        },
        {
            displayName: 'Property Prefix Name',
            name: 'dataPropertyAttachmentsPrefixName',
            type: 'string',
            default: 'attachment_',
            displayOptions: {
                show: {
                    format: ['resolved'],
                },
            },
            description: 'Prefix for name of the binary property to which to write the attachments. An index starting with 0 will be added. So if name is "attachment_" the first attachment is saved to "attachment_0"',
        },
        {
            displayName: 'Property Prefix Name',
            name: 'dataPropertyAttachmentsPrefixName',
            type: 'string',
            default: 'attachment_',
            displayOptions: {
                show: {
                    format: ['simple'],
                    downloadAttachments: [true],
                },
            },
            description: 'Prefix for name of the binary property to which to write the attachments. An index starting with 0 will be added. So if name is "attachment_" the first attachment is saved to "attachment_0"',
        },
        {
            displayName: 'Options',
            name: 'options',
            type: 'collection',
            placeholder: 'Add Option',
            default: {},
            options: [
                {
                    displayName: 'Custom Email Rules',
                    name: 'customEmailConfig',
                    type: 'string',
                    default: '["UNSEEN"]',
                    description: 'Custom email fetching rules. See <a href="https://github.com/mscdex/node-imap">node-imap</a>\'s search function for more details.',
                },
                {
                    displayName: 'Ignore SSL Issues',
                    name: 'allowUnauthorizedCerts',
                    type: 'boolean',
                    default: false,
                    description: 'Whether to connect even if SSL certificate validation is not possible',
                },
                {
                    displayName: 'Force Reconnect',
                    name: 'forceReconnect',
                    type: 'number',
                    default: 60,
                    description: 'Sets an interval (in minutes) to force a reconnection',
                },
            ],
        },
    ],
};
class EmailReadImapV1 {
    constructor(baseDescription) {
        this.methods = {
            credentialTest: {
                async imapConnectionTest(credential) {
                    const credentials = credential.data;
                    try {
                        const config = {
                            imap: {
                                user: credentials.user,
                                password: credentials.password,
                                host: credentials.host,
                                port: credentials.port,
                                tls: credentials.secure,
                                authTimeout: 20000,
                            },
                        };
                        const tlsOptions = {};
                        if (credentials.secure) {
                            tlsOptions.servername = credentials.host;
                        }
                        if (!(0, lodash_isempty_1.default)(tlsOptions)) {
                            config.imap.tlsOptions = tlsOptions;
                        }
                        const conn = (0, imap_simple_1.connect)(config).then(async (entry) => {
                            return entry;
                        });
                        (await conn).getBoxes((_err, _boxes) => { });
                    }
                    catch (error) {
                        console.log(error);
                        return {
                            status: 'Error',
                            message: error.message,
                        };
                    }
                    return {
                        status: 'OK',
                        message: 'Connection successful!',
                    };
                },
            },
        };
        this.description = {
            ...baseDescription,
            ...versionDescription,
        };
    }
    async trigger() {
        const credentials = await this.getCredentials('imap');
        const mailbox = this.getNodeParameter('mailbox');
        const postProcessAction = this.getNodeParameter('postProcessAction');
        const options = this.getNodeParameter('options', {});
        const staticData = this.getWorkflowStaticData('node');
        this.logger.debug('Loaded static data for node "EmailReadImap"', { staticData });
        let connection;
        const getText = async (parts, message, subtype) => {
            if (!message.attributes.struct) {
                return '';
            }
            const textParts = parts.filter((part) => {
                return (part.type.toUpperCase() === 'TEXT' && part.subtype.toUpperCase() === subtype.toUpperCase());
            });
            if (textParts.length === 0) {
                return '';
            }
            try {
                return await connection.getPartData(message, textParts[0]);
            }
            catch {
                return '';
            }
        };
        const getAttachment = async (imapConnection, parts, message) => {
            if (!message.attributes.struct) {
                return [];
            }
            const attachmentParts = parts.filter((part) => {
                return part.disposition && part.disposition.type.toUpperCase() === 'ATTACHMENT';
            });
            const attachmentPromises = [];
            let attachmentPromise;
            for (const attachmentPart of attachmentParts) {
                attachmentPromise = imapConnection
                    .getPartData(message, attachmentPart)
                    .then(async (partData) => {
                    return this.helpers.prepareBinaryData(partData, attachmentPart.disposition.params.filename);
                });
                attachmentPromises.push(attachmentPromise);
            }
            return Promise.all(attachmentPromises);
        };
        const getNewEmails = async (imapConnection, searchCriteria) => {
            const format = this.getNodeParameter('format', 0);
            let fetchOptions = {};
            if (format === 'simple' || format === 'raw') {
                fetchOptions = {
                    bodies: ['TEXT', 'HEADER'],
                    markSeen: false,
                    struct: true,
                };
            }
            else if (format === 'resolved') {
                fetchOptions = {
                    bodies: [''],
                    markSeen: false,
                    struct: true,
                };
            }
            const results = await imapConnection.search(searchCriteria, fetchOptions);
            const newEmails = [];
            let newEmail, messageHeader, messageBody;
            let attachments;
            let propertyName;
            const topLevelProperties = ['cc', 'date', 'from', 'subject', 'to'];
            if (format === 'resolved') {
                const dataPropertyAttachmentsPrefixName = this.getNodeParameter('dataPropertyAttachmentsPrefixName');
                for (const message of results) {
                    if (staticData.lastMessageUid !== undefined &&
                        message.attributes.uid <= staticData.lastMessageUid) {
                        continue;
                    }
                    if (staticData.lastMessageUid === undefined ||
                        staticData.lastMessageUid < message.attributes.uid) {
                        staticData.lastMessageUid = message.attributes.uid;
                    }
                    const part = (0, lodash_find_1.default)(message.parts, { which: '' });
                    if (part === undefined) {
                        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Email part could not be parsed.');
                    }
                    const parsedEmail = await parseRawEmail.call(this, part.body, dataPropertyAttachmentsPrefixName);
                    newEmails.push(parsedEmail);
                }
            }
            else if (format === 'simple') {
                const downloadAttachments = this.getNodeParameter('downloadAttachments');
                let dataPropertyAttachmentsPrefixName = '';
                if (downloadAttachments) {
                    dataPropertyAttachmentsPrefixName = this.getNodeParameter('dataPropertyAttachmentsPrefixName');
                }
                for (const message of results) {
                    if (staticData.lastMessageUid !== undefined &&
                        message.attributes.uid <= staticData.lastMessageUid) {
                        continue;
                    }
                    if (staticData.lastMessageUid === undefined ||
                        staticData.lastMessageUid < message.attributes.uid) {
                        staticData.lastMessageUid = message.attributes.uid;
                    }
                    const parts = (0, imap_simple_1.getParts)(message.attributes.struct);
                    newEmail = {
                        json: {
                            textHtml: await getText(parts, message, 'html'),
                            textPlain: await getText(parts, message, 'plain'),
                            metadata: {},
                        },
                    };
                    messageHeader = message.parts.filter((part) => {
                        return part.which === 'HEADER';
                    });
                    messageBody = messageHeader[0].body;
                    for (propertyName of Object.keys(messageBody)) {
                        if (messageBody[propertyName].length) {
                            if (topLevelProperties.includes(propertyName)) {
                                newEmail.json[propertyName] = messageBody[propertyName][0];
                            }
                            else {
                                newEmail.json.metadata[propertyName] =
                                    messageBody[propertyName][0];
                            }
                        }
                    }
                    if (downloadAttachments) {
                        attachments = await getAttachment(imapConnection, parts, message);
                        if (attachments.length) {
                            newEmail.binary = {};
                            for (let i = 0; i < attachments.length; i++) {
                                newEmail.binary[`${dataPropertyAttachmentsPrefixName}${i}`] = attachments[i];
                            }
                        }
                    }
                    newEmails.push(newEmail);
                }
            }
            else if (format === 'raw') {
                for (const message of results) {
                    if (staticData.lastMessageUid !== undefined &&
                        message.attributes.uid <= staticData.lastMessageUid) {
                        continue;
                    }
                    if (staticData.lastMessageUid === undefined ||
                        staticData.lastMessageUid < message.attributes.uid) {
                        staticData.lastMessageUid = message.attributes.uid;
                    }
                    const part = (0, lodash_find_1.default)(message.parts, { which: 'TEXT' });
                    if (part === undefined) {
                        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Email part could not be parsed.');
                    }
                    newEmail = {
                        json: {
                            raw: part.body,
                        },
                    };
                    newEmails.push(newEmail);
                }
            }
            if (postProcessAction === 'read') {
                const uidList = results.map((e) => e.attributes.uid);
                if (uidList.length > 0) {
                    await imapConnection.addFlags(uidList, '\\SEEN');
                }
            }
            return newEmails;
        };
        const returnedPromise = await this.helpers.createDeferredPromise();
        const establishConnection = async () => {
            let searchCriteria = ['UNSEEN'];
            if (options.customEmailConfig !== undefined) {
                try {
                    searchCriteria = JSON.parse(options.customEmailConfig);
                }
                catch (error) {
                    throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Custom email config is not valid JSON.');
                }
            }
            const config = {
                imap: {
                    user: credentials.user,
                    password: credentials.password,
                    host: credentials.host,
                    port: credentials.port,
                    tls: credentials.secure,
                    authTimeout: 20000,
                },
                onmail: async () => {
                    if (connection) {
                        if (staticData.lastMessageUid !== undefined) {
                            searchCriteria.push(['UID', `${staticData.lastMessageUid}:*`]);
                            this.logger.debug('Querying for new messages on node "EmailReadImap"', {
                                searchCriteria,
                            });
                        }
                        try {
                            const returnData = await getNewEmails(connection, searchCriteria);
                            if (returnData.length) {
                                this.emit([returnData]);
                            }
                        }
                        catch (error) {
                            this.logger.error('Email Read Imap node encountered an error fetching new emails', {
                                error,
                            });
                            await returnedPromise.promise().then(() => {
                                this.emitError(error);
                            });
                        }
                    }
                },
            };
            const tlsOptions = {};
            if (options.allowUnauthorizedCerts === true) {
                tlsOptions.rejectUnauthorized = false;
            }
            if (credentials.secure) {
                tlsOptions.servername = credentials.host;
            }
            if (!(0, lodash_isempty_1.default)(tlsOptions)) {
                config.imap.tlsOptions = tlsOptions;
            }
            return (0, imap_simple_1.connect)(config).then(async (conn) => {
                conn.on('error', async (error) => {
                    const errorCode = error.code.toUpperCase();
                    if (['ECONNRESET', 'EPIPE'].includes(errorCode)) {
                        this.logger.verbose(`IMAP connection was reset (${errorCode}) - reconnecting.`, {
                            error,
                        });
                        try {
                            connection = await establishConnection();
                            await connection.openBox(mailbox);
                            return;
                        }
                        catch (e) {
                            this.logger.error('IMAP reconnect did fail', { error: e });
                        }
                    }
                    else {
                        this.logger.error('Email Read Imap node encountered a connection error', { error });
                        this.emitError(error);
                    }
                });
                return conn;
            });
        };
        connection = await establishConnection();
        await connection.openBox(mailbox);
        let reconnectionInterval;
        if (options.forceReconnect !== undefined) {
            reconnectionInterval = setInterval(async () => {
                this.logger.verbose('Forcing reconnection of IMAP node.');
                connection.end();
                connection = await establishConnection();
                await connection.openBox(mailbox);
            }, options.forceReconnect * 1000 * 60);
        }
        async function closeFunction() {
            if (reconnectionInterval) {
                clearInterval(reconnectionInterval);
            }
            connection.end();
        }
        returnedPromise.resolve();
        return {
            closeFunction,
        };
    }
}
exports.EmailReadImapV1 = EmailReadImapV1;
