// n8n-nodes-base/dist/nodes/Aws/Textract/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateCredentials = exports.simplify = exports.awsApiRequestSOAP = exports.awsApiRequestREST = exports.awsApiRequest = void 0;
const url_1 = require("url");
const aws4_1 = require("aws4");
const xml2js_1 = require("xml2js");
const n8n_workflow_1 = require("n8n-workflow");
function getEndpointForService(service, credentials) {
    let endpoint;
    if (service === 'lambda' && credentials.lambdaEndpoint) {
        endpoint = credentials.lambdaEndpoint;
    }
    else if (service === 'sns' && credentials.snsEndpoint) {
        endpoint = credentials.snsEndpoint;
    }
    else {
        endpoint = `https://${service}.${credentials.region}.amazonaws.com`;
    }
    return endpoint.replace('{region}', credentials.region);
}
async function awsApiRequest(service, method, path, body, headers) {
    var _a, _b, _c, _d;
    const credentials = await this.getCredentials('aws');
    const requestOptions = {
        qs: {
            service,
            path,
        },
        method,
        body,
        url: '',
        headers,
        region: credentials === null || credentials === void 0 ? void 0 : credentials.region,
    };
    try {
        return await this.helpers.requestWithAuthentication.call(this, 'aws', requestOptions);
    }
    catch (error) {
        if (((_a = error === null || error === void 0 ? void 0 : error.response) === null || _a === void 0 ? void 0 : _a.data) || ((_b = error === null || error === void 0 ? void 0 : error.response) === null || _b === void 0 ? void 0 : _b.body)) {
            const errorMessage = ((_c = error === null || error === void 0 ? void 0 : error.response) === null || _c === void 0 ? void 0 : _c.data) || ((_d = error === null || error === void 0 ? void 0 : error.response) === null || _d === void 0 ? void 0 : _d.body);
            if (errorMessage.includes('AccessDeniedException')) {
                const user = JSON.parse(errorMessage).Message.split(' ')[1];
                throw new n8n_workflow_1.NodeApiError(this.getNode(), error, {
                    message: 'Unauthorized — please check your AWS policy configuration',
                    description: `Make sure an identity-based policy allows user ${user} to perform textract:AnalyzeExpense`,
                });
            }
        }
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.awsApiRequest = awsApiRequest;
async function awsApiRequestREST(service, method, path, body, headers) {
    const response = await awsApiRequest.call(this, service, method, path, body, headers);
    try {
        return JSON.parse(response);
    }
    catch (error) {
        return response;
    }
}
exports.awsApiRequestREST = awsApiRequestREST;
async function awsApiRequestSOAP(service, method, path, body, headers) {
    const response = await awsApiRequest.call(this, service, method, path, body, headers);
    try {
        return await new Promise((resolve, reject) => {
            (0, xml2js_1.parseString)(response, { explicitArray: false }, (err, data) => {
                if (err) {
                    return reject(err);
                }
                resolve(data);
            });
        });
    }
    catch (error) {
        return response;
    }
}
exports.awsApiRequestSOAP = awsApiRequestSOAP;
function simplify(data) {
    var _a, _b;
    const result = {};
    for (const document of data.ExpenseDocuments) {
        for (const field of document.SummaryFields) {
            result[((_a = field === null || field === void 0 ? void 0 : field.Type) === null || _a === void 0 ? void 0 : _a.Text) || ((_b = field === null || field === void 0 ? void 0 : field.LabelDetection) === null || _b === void 0 ? void 0 : _b.Text)] = field.ValueDetection.Text;
        }
    }
    return result;
}
exports.simplify = simplify;
async function validateCredentials(decryptedCredentials, service) {
    const credentials = decryptedCredentials;
    const endpoint = new url_1.URL(getEndpointForService(service, credentials) + '?Action=GetCallerIdentity&Version=2011-06-15');
    const signOpts = {
        host: endpoint.host,
        method: 'POST',
        path: '?Action=GetCallerIdentity&Version=2011-06-15',
    };
    const securityHeaders = {
        accessKeyId: `${credentials.accessKeyId}`.trim(),
        secretAccessKey: `${credentials.secretAccessKey}`.trim(),
        sessionToken: credentials.temporaryCredentials
            ? `${credentials.sessionToken}`.trim()
            : undefined,
    };
    (0, aws4_1.sign)(signOpts, securityHeaders);
    const options = {
        headers: signOpts.headers,
        method: 'POST',
        uri: endpoint.href,
        body: signOpts.body,
    };
    const response = await this.helpers.request(options);
    return new Promise((resolve, reject) => {
        (0, xml2js_1.parseString)(response, { explicitArray: false }, (err, data) => {
            if (err) {
                return reject(err);
            }
            resolve(data);
        });
    });
}
exports.validateCredentials = validateCredentials;
