// n8n/dist/decorators/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerController = exports.Middleware = exports.Delete = exports.Patch = exports.Put = exports.Post = exports.Get = exports.RestController = exports.NoAuthRequired = exports.Authorized = void 0;
var Authorized_1 = require("./Authorized");
Object.defineProperty(exports, "Authorized", { enumerable: true, get: function () { return Authorized_1.Authorized; } });
Object.defineProperty(exports, "NoAuthRequired", { enumerable: true, get: function () { return Authorized_1.NoAuthRequired; } });
var RestController_1 = require("./RestController");
Object.defineProperty(exports, "RestController", { enumerable: true, get: function () { return RestController_1.RestController; } });
var Route_1 = require("./Route");
Object.defineProperty(exports, "Get", { enumerable: true, get: function () { return Route_1.Get; } });
Object.defineProperty(exports, "Post", { enumerable: true, get: function () { return Route_1.Post; } });
Object.defineProperty(exports, "Put", { enumerable: true, get: function () { return Route_1.Put; } });
Object.defineProperty(exports, "Patch", { enumerable: true, get: function () { return Route_1.Patch; } });
Object.defineProperty(exports, "Delete", { enumerable: true, get: function () { return Route_1.Delete; } });
var Middleware_1 = require("./Middleware");
Object.defineProperty(exports, "Middleware", { enumerable: true, get: function () { return Middleware_1.Middleware; } });
var registerController_1 = require("./registerController");
Object.defineProperty(exports, "registerController", { enumerable: true, get: function () { return registerController_1.registerController; } });
