// n8n-nodes-base/dist/nodes/Aws/ELB/GenericFunctions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.awsApiRequestSOAPAllItems = exports.awsApiRequestSOAP = exports.awsApiRequestREST = exports.awsApiRequest = void 0;
const lodash_get_1 = __importDefault(require("lodash.get"));
const xml2js_1 = require("xml2js");
const n8n_workflow_1 = require("n8n-workflow");
async function awsApiRequest(service, method, path, body, query = {}, headers, _option = {}, _region) {
    const credentials = await this.getCredentials('aws');
    const requestOptions = {
        qs: {
            ...query,
            service,
            path,
        },
        headers,
        method,
        url: '',
        body,
        region: credentials === null || credentials === void 0 ? void 0 : credentials.region,
    };
    try {
        return await this.helpers.requestWithAuthentication.call(this, 'aws', requestOptions);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.awsApiRequest = awsApiRequest;
async function awsApiRequestREST(service, method, path, body, query = {}, headers, options = {}, region) {
    const response = await awsApiRequest.call(this, service, method, path, body, query, headers, options, region);
    try {
        return JSON.parse(response);
    }
    catch (e) {
        return response;
    }
}
exports.awsApiRequestREST = awsApiRequestREST;
async function awsApiRequestSOAP(service, method, path, body, query = {}, headers, option = {}, region) {
    const response = await awsApiRequest.call(this, service, method, path, body, query, headers, option, region);
    try {
        return await new Promise((resolve, reject) => {
            (0, xml2js_1.parseString)(response, { explicitArray: false }, (err, data) => {
                if (err) {
                    return reject(err);
                }
                resolve(data);
            });
        });
    }
    catch (e) {
        return e;
    }
}
exports.awsApiRequestSOAP = awsApiRequestSOAP;
async function awsApiRequestSOAPAllItems(propertyName, service, method, path, body, query = {}, headers = {}, option = {}, region) {
    const returnData = [];
    let responseData;
    const propertyNameArray = propertyName.split('.');
    do {
        responseData = await awsApiRequestSOAP.call(this, service, method, path, body, query, headers, option, region);
        if ((0, lodash_get_1.default)(responseData, `${propertyNameArray[0]}.${propertyNameArray[1]}.NextMarker`)) {
            query.Marker = (0, lodash_get_1.default)(responseData, `${propertyNameArray[0]}.${propertyNameArray[1]}.NextMarker`);
        }
        if ((0, lodash_get_1.default)(responseData, propertyName)) {
            if (Array.isArray((0, lodash_get_1.default)(responseData, propertyName))) {
                returnData.push.apply(returnData, (0, lodash_get_1.default)(responseData, propertyName));
            }
            else {
                returnData.push((0, lodash_get_1.default)(responseData, propertyName));
            }
        }
    } while ((0, lodash_get_1.default)(responseData, `${propertyNameArray[0]}.${propertyNameArray[1]}.NextMarker`) !== undefined);
    return returnData;
}
exports.awsApiRequestSOAPAllItems = awsApiRequestSOAPAllItems;
