// n8n-nodes-base/dist/nodes/N8n/WorkflowLocator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.workflowIdLocator = exports.searchWorkflows = void 0;
const GenericFunctions_1 = require("./GenericFunctions");
async function searchWorkflows(query) {
    const searchResults = (await GenericFunctions_1.apiRequestAllItems.call(this, 'GET', 'workflows', {}));
    const workflows = searchResults
        .map((w) => ({
        name: `${w.name} (#${w.id})`,
        value: w.id,
    }))
        .filter((w) => {
        var _a;
        return !query ||
            w.name.toLowerCase().includes(query.toLowerCase()) ||
            ((_a = w.value) === null || _a === void 0 ? void 0 : _a.toString()) === query;
    })
        .sort((a, b) => b.value - a.value);
    return {
        results: workflows,
    };
}
exports.searchWorkflows = searchWorkflows;
exports.workflowIdLocator = {
    displayName: 'Workflow',
    name: 'workflowId',
    type: 'resourceLocator',
    default: { mode: 'list', value: '' },
    description: 'Workflow to filter the executions by',
    modes: [
        {
            displayName: 'From List',
            name: 'list',
            type: 'list',
            placeholder: 'Select a Workflow...',
            initType: 'workflow',
            typeOptions: {
                searchListMethod: 'searchWorkflows',
                searchFilterRequired: false,
                searchable: true,
            },
        },
        {
            displayName: 'By URL',
            name: 'url',
            type: 'string',
            placeholder: 'https://myinstance.app.n8n.cloud/workflow/1',
            validation: [
                {
                    type: 'regex',
                    properties: {
                        regex: '.*/workflow/([0-9]{1,})',
                        errorMessage: 'Not a valid Workflow URL',
                    },
                },
            ],
            extractValue: {
                type: 'regex',
                regex: '.*/workflow/([0-9]{1,})',
            },
        },
        {
            displayName: 'ID',
            name: 'id',
            type: 'string',
            validation: [
                {
                    type: 'regex',
                    properties: {
                        regex: '[0-9]{1,}',
                        errorMessage: 'Not a valid Workflow ID',
                    },
                },
            ],
            placeholder: '1',
        },
    ],
};
