// n8n-nodes-base/dist/nodes/Formstack/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getSubmission = exports.getFields = exports.getForms = exports.apiRequestAllItems = exports.apiRequest = exports.FormstackFieldFormat = void 0;
const n8n_workflow_1 = require("n8n-workflow");
var FormstackFieldFormat;
(function (FormstackFieldFormat) {
    FormstackFieldFormat["ID"] = "id";
    FormstackFieldFormat["Label"] = "label";
    FormstackFieldFormat["Name"] = "name";
})(FormstackFieldFormat = exports.FormstackFieldFormat || (exports.FormstackFieldFormat = {}));
async function apiRequest(method, endpoint, body = {}, query = {}) {
    const authenticationMethod = this.getNodeParameter('authentication', 0);
    const options = {
        headers: {},
        method,
        body,
        qs: query || {},
        uri: `https://www.formstack.com/api/v2/${endpoint}`,
        json: true,
    };
    if (!Object.keys(body).length) {
        delete options.body;
    }
    try {
        if (authenticationMethod === 'accessToken') {
            const credentials = (await this.getCredentials('formstackApi'));
            options.headers.Authorization = `Bearer ${credentials.accessToken}`;
            return await this.helpers.request(options);
        }
        else {
            return await this.helpers.requestOAuth2.call(this, 'formstackOAuth2Api', options);
        }
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.apiRequest = apiRequest;
async function apiRequestAllItems(method, endpoint, body, dataKey, query) {
    if (query === undefined) {
        query = {};
    }
    query.per_page = 200;
    query.page = 0;
    const returnData = {
        items: [],
    };
    let responseData;
    do {
        query.page += 1;
        responseData = await apiRequest.call(this, method, endpoint, body, query);
        returnData.items.push.apply(returnData.items, responseData[dataKey]);
    } while (responseData.total !== undefined &&
        Math.ceil(responseData.total / query.per_page) > query.page);
    return returnData;
}
exports.apiRequestAllItems = apiRequestAllItems;
async function getForms() {
    const endpoint = 'form.json';
    const responseData = await apiRequestAllItems.call(this, 'GET', endpoint, {}, 'forms', {
        folders: false,
    });
    if (responseData.items === undefined) {
        throw new Error('No data got returned');
    }
    const returnData = [];
    for (const baseData of responseData.items) {
        returnData.push({
            name: baseData.name,
            value: baseData.id,
        });
    }
    return returnData;
}
exports.getForms = getForms;
async function getFields(formID) {
    const endpoint = `form/${formID}.json`;
    const responseData = await apiRequestAllItems.call(this, 'GET', endpoint, {}, 'fields');
    if (responseData.items === undefined) {
        throw new Error('No form fields meta data got returned');
    }
    const fields = responseData.items;
    const fieldMap = {};
    fields.forEach((field) => {
        fieldMap[field.id] = field;
    });
    return fieldMap;
}
exports.getFields = getFields;
async function getSubmission(uniqueId) {
    const endpoint = `submission/${uniqueId}.json`;
    const responseData = await apiRequestAllItems.call(this, 'GET', endpoint, {}, 'data');
    if (responseData.items === undefined) {
        throw new Error('No form fields meta data got returned');
    }
    return responseData.items;
}
exports.getSubmission = getSubmission;
