// n8n-nodes-base/dist/nodes/Shopify/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.keysToSnakeCase = exports.shopifyApiRequestAllItems = exports.shopifyApiRequest = void 0;
const change_case_1 = require("change-case");
async function shopifyApiRequest(method, resource, body = {}, query = {}, uri, option = {}) {
    const authenticationMethod = this.getNodeParameter('authentication', 0, 'oAuth2');
    let credentials;
    let credentialType = 'shopifyOAuth2Api';
    if (authenticationMethod === 'apiKey') {
        credentials = await this.getCredentials('shopifyApi');
        credentialType = 'shopifyApi';
    }
    else if (authenticationMethod === 'accessToken') {
        credentials = await this.getCredentials('shopifyAccessTokenApi');
        credentialType = 'shopifyAccessTokenApi';
    }
    else {
        credentials = await this.getCredentials('shopifyOAuth2Api');
    }
    const options = {
        method,
        qs: query,
        uri: uri || `https://${credentials.shopSubdomain}.myshopify.com/admin/api/2019-10${resource}`,
        body,
        json: true,
    };
    const oAuth2Options = {
        tokenType: 'Bearer',
        keyToIncludeInAccessTokenHeader: 'X-Shopify-Access-Token',
    };
    if (authenticationMethod === 'apiKey') {
        Object.assign(options, {
            auth: { username: credentials.apiKey, password: credentials.password },
        });
    }
    if (Object.keys(option).length !== 0) {
        Object.assign(options, option);
    }
    if (Object.keys(body).length === 0) {
        delete options.body;
    }
    if (Object.keys(query).length === 0) {
        delete options.qs;
    }
    return this.helpers.requestWithAuthentication.call(this, credentialType, options, {
        oauth2: oAuth2Options,
    });
}
exports.shopifyApiRequest = shopifyApiRequest;
async function shopifyApiRequestAllItems(propertyName, method, resource, body = {}, query = {}) {
    var _a;
    const returnData = [];
    for (const field in query) {
        if (query[field] === '') {
            delete query[field];
        }
    }
    let responseData;
    let uri;
    do {
        responseData = await shopifyApiRequest.call(this, method, resource, body, query, uri, {
            resolveWithFullResponse: true,
        });
        if (responseData.headers.link) {
            uri = responseData.headers.link.split(';')[0].replace('<', '').replace('>', '');
        }
        returnData.push.apply(returnData, responseData.body[propertyName]);
    } while ((_a = responseData.headers.link) === null || _a === void 0 ? void 0 : _a.includes('rel="next"'));
    return returnData;
}
exports.shopifyApiRequestAllItems = shopifyApiRequestAllItems;
function keysToSnakeCase(elements) {
    if (elements === undefined) {
        return [];
    }
    if (!Array.isArray(elements)) {
        elements = [elements];
    }
    for (const element of elements) {
        for (const key of Object.keys(element)) {
            if (key !== (0, change_case_1.snakeCase)(key)) {
                element[(0, change_case_1.snakeCase)(key)] = element[key];
                delete element[key];
            }
        }
    }
    return elements;
}
exports.keysToSnakeCase = keysToSnakeCase;
