// n8n-nodes-base/dist/nodes/Elastic/ElasticSecurity/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getVersion = exports.throwOnEmptyUpdate = exports.getConnector = exports.handleListing = exports.elasticSecurityApiRequestAllItems = exports.elasticSecurityApiRequest = exports.tolerateTrailingSlash = void 0;
const n8n_workflow_1 = require("n8n-workflow");
function tolerateTrailingSlash(baseUrl) {
    return baseUrl.endsWith('/') ? baseUrl.substr(0, baseUrl.length - 1) : baseUrl;
}
exports.tolerateTrailingSlash = tolerateTrailingSlash;
async function elasticSecurityApiRequest(method, endpoint, body = {}, qs = {}) {
    var _a, _b;
    const { username, password, baseUrl: rawBaseUrl, } = (await this.getCredentials('elasticSecurityApi'));
    const baseUrl = tolerateTrailingSlash(rawBaseUrl);
    const token = Buffer.from(`${username}:${password}`).toString('base64');
    const options = {
        headers: {
            Authorization: `Basic ${token}`,
            'kbn-xsrf': true,
        },
        method,
        body,
        qs,
        uri: `${baseUrl}/api${endpoint}`,
        json: true,
    };
    if (!Object.keys(body).length) {
        delete options.body;
    }
    if (!Object.keys(qs).length) {
        delete options.qs;
    }
    try {
        return await this.helpers.request(options);
    }
    catch (error) {
        if (((_a = error === null || error === void 0 ? void 0 : error.error) === null || _a === void 0 ? void 0 : _a.error) === 'Not Acceptable' && ((_b = error === null || error === void 0 ? void 0 : error.error) === null || _b === void 0 ? void 0 : _b.message)) {
            error.error.error = `${error.error.error}: ${error.error.message}`;
        }
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.elasticSecurityApiRequest = elasticSecurityApiRequest;
async function elasticSecurityApiRequestAllItems(method, endpoint, body = {}, qs = {}) {
    let _page = 1;
    const returnData = [];
    let responseData;
    const resource = this.getNodeParameter('resource', 0);
    do {
        responseData = await elasticSecurityApiRequest.call(this, method, endpoint, body, qs);
        _page++;
        const items = resource === 'case' ? responseData.cases : responseData;
        returnData.push(...items);
    } while (returnData.length < responseData.total);
    return returnData;
}
exports.elasticSecurityApiRequestAllItems = elasticSecurityApiRequestAllItems;
async function handleListing(method, endpoint, body = {}, qs = {}) {
    const returnAll = this.getNodeParameter('returnAll', 0);
    if (returnAll) {
        return elasticSecurityApiRequestAllItems.call(this, method, endpoint, body, qs);
    }
    const responseData = await elasticSecurityApiRequestAllItems.call(this, method, endpoint, body, qs);
    const limit = this.getNodeParameter('limit', 0);
    return responseData.slice(0, limit);
}
exports.handleListing = handleListing;
async function getConnector(connectorId) {
    const endpoint = `/actions/connector/${connectorId}`;
    const { id, name, connector_type_id: type, } = (await elasticSecurityApiRequest.call(this, 'GET', endpoint));
    return { id, name, type };
}
exports.getConnector = getConnector;
function throwOnEmptyUpdate(resource) {
    throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Please enter at least one field to update for the ${resource}`);
}
exports.throwOnEmptyUpdate = throwOnEmptyUpdate;
async function getVersion(endpoint) {
    const { version } = (await elasticSecurityApiRequest.call(this, 'GET', endpoint));
    if (!version) {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Cannot retrieve version for resource');
    }
    return version;
}
exports.getVersion = getVersion;
