// n8n-nodes-base/dist/nodes/Gotify/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.gotifyApiRequestAllItems = exports.gotifyApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function gotifyApiRequest(method, path, body = {}, qs = {}, uri, _option = {}) {
    const credentials = await this.getCredentials('gotifyApi');
    const options = {
        method,
        headers: {
            'X-Gotify-Key': method === 'POST' ? credentials.appApiToken : credentials.clientApiToken,
            accept: 'application/json',
        },
        body,
        qs,
        uri: uri || `${credentials.url}${path}`,
        json: true,
    };
    try {
        if (Object.keys(body).length === 0) {
            delete options.body;
        }
        return await this.helpers.request.call(this, options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.gotifyApiRequest = gotifyApiRequest;
async function gotifyApiRequestAllItems(propertyName, method, endpoint, body = {}, query = {}) {
    const returnData = [];
    let responseData;
    let uri;
    query.limit = 100;
    do {
        responseData = await gotifyApiRequest.call(this, method, endpoint, body, query, uri);
        if (responseData.paging.next) {
            uri = responseData.paging.next;
        }
        returnData.push.apply(returnData, responseData[propertyName]);
    } while (responseData.paging.next);
    return returnData;
}
exports.gotifyApiRequestAllItems = gotifyApiRequestAllItems;
