// n8n-nodes-base/dist/nodes/LinkedIn/LinkedIn.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LinkedIn = void 0;
const GenericFunctions_1 = require("./GenericFunctions");
const PostDescription_1 = require("./PostDescription");
class LinkedIn {
    constructor() {
        this.description = {
            displayName: 'LinkedIn',
            name: 'linkedIn',
            icon: 'file:linkedin.svg',
            group: ['input'],
            version: 1,
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            description: 'Consume LinkedIn API',
            defaults: {
                name: 'LinkedIn',
            },
            inputs: ['main'],
            outputs: ['main'],
            credentials: [
                {
                    name: 'linkedInOAuth2Api',
                    required: true,
                },
            ],
            properties: [
                {
                    displayName: 'Resource',
                    name: 'resource',
                    type: 'options',
                    noDataExpression: true,
                    options: [
                        {
                            name: 'Post',
                            value: 'post',
                        },
                    ],
                    default: 'post',
                },
                ...PostDescription_1.postOperations,
                ...PostDescription_1.postFields,
            ],
        };
        this.methods = {
            loadOptions: {
                async getPersonUrn() {
                    const returnData = [];
                    const person = await GenericFunctions_1.linkedInApiRequest.call(this, 'GET', '/me', {});
                    returnData.push({
                        name: `${person.localizedFirstName} ${person.localizedLastName}`,
                        value: person.id,
                    });
                    return returnData;
                },
            },
        };
    }
    async execute() {
        const items = this.getInputData();
        const returnData = [];
        let responseData;
        const resource = this.getNodeParameter('resource', 0);
        const operation = this.getNodeParameter('operation', 0);
        let body = {};
        for (let i = 0; i < items.length; i++) {
            try {
                if (resource === 'post') {
                    if (operation === 'create') {
                        const text = this.getNodeParameter('text', i);
                        const shareMediaCategory = this.getNodeParameter('shareMediaCategory', i);
                        const postAs = this.getNodeParameter('postAs', i);
                        const additionalFields = this.getNodeParameter('additionalFields', i);
                        let authorUrn = '';
                        let visibility = 'PUBLIC';
                        if (postAs === 'person') {
                            const personUrn = this.getNodeParameter('person', i);
                            visibility = additionalFields.visibility || 'PUBLIC';
                            authorUrn = `urn:li:person:${personUrn}`;
                        }
                        else {
                            const organizationUrn = this.getNodeParameter('organization', i);
                            authorUrn = `urn:li:organization:${organizationUrn}`;
                        }
                        let description = '';
                        let title = '';
                        let originalUrl = '';
                        body = {
                            author: authorUrn,
                            lifecycleState: 'PUBLISHED',
                            distribution: {
                                feedDistribution: 'MAIN_FEED',
                                targetEnties: [],
                                thirdPartyDistributionChannels: [],
                            },
                            visibility,
                        };
                        if (shareMediaCategory === 'IMAGE') {
                            if (additionalFields.description) {
                                description = additionalFields.description;
                            }
                            if (additionalFields.title) {
                                title = additionalFields.title;
                            }
                            const registerRequest = {
                                initializeUploadRequest: {
                                    owner: authorUrn,
                                },
                            };
                            const registerObject = await GenericFunctions_1.linkedInApiRequest.call(this, 'POST', '/images?action=initializeUpload', registerRequest);
                            const binaryPropertyName = this.getNodeParameter('binaryPropertyName', i);
                            this.helpers.assertBinaryData(i, binaryPropertyName);
                            const buffer = await this.helpers.getBinaryDataBuffer(i, binaryPropertyName);
                            const { uploadUrl, image } = registerObject.value;
                            await GenericFunctions_1.linkedInApiRequest.call(this, 'POST', uploadUrl, buffer, true);
                            const imageBody = {
                                content: {
                                    media: {
                                        title,
                                        id: image,
                                        description,
                                    },
                                },
                                commentary: text,
                            };
                            Object.assign(body, imageBody);
                        }
                        else if (shareMediaCategory === 'ARTICLE') {
                            if (additionalFields.description) {
                                description = additionalFields.description;
                            }
                            if (additionalFields.title) {
                                title = additionalFields.title;
                            }
                            if (additionalFields.originalUrl) {
                                originalUrl = additionalFields.originalUrl;
                            }
                            const articleBody = {
                                content: {
                                    article: {
                                        title,
                                        description,
                                        source: originalUrl,
                                    },
                                },
                                commentary: text,
                            };
                            Object.assign(body, articleBody);
                            if (description === '') {
                                delete body.description;
                            }
                            if (title === '') {
                                delete body.title;
                            }
                        }
                        else {
                            Object.assign(body, { commentary: text });
                        }
                        const endpoint = '/posts';
                        responseData = await GenericFunctions_1.linkedInApiRequest.call(this, 'POST', endpoint, body);
                    }
                }
                const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
                returnData.push(...executionData);
            }
            catch (error) {
                if (this.continueOnFail()) {
                    const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray({ error: error.message }), { itemData: { item: i } });
                    returnData.push(...executionData);
                    continue;
                }
                throw error;
            }
        }
        return this.prepareOutputData(returnData);
    }
}
exports.LinkedIn = LinkedIn;
