// n8n-nodes-base/dist/nodes/Microsoft/Excel/MicrosoftExcel.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MicrosoftExcel = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const MicrosoftExcelV1_node_1 = require("./v1/MicrosoftExcelV1.node");
const MicrosoftExcelV2_node_1 = require("./v2/MicrosoftExcelV2.node");
class MicrosoftExcel extends n8n_workflow_1.VersionedNodeType {
    constructor() {
        const baseDescription = {
            displayName: 'Microsoft Excel 365',
            name: 'microsoftExcel',
            icon: 'file:excel.svg',
            group: ['input'],
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            description: 'Consume Microsoft Excel API',
            defaultVersion: 2,
        };
        const nodeVersions = {
            1: new MicrosoftExcelV1_node_1.MicrosoftExcelV1(baseDescription),
            2: new MicrosoftExcelV2_node_1.MicrosoftExcelV2(baseDescription),
        };
        super(nodeVersions, baseDescription);
    }
}
exports.MicrosoftExcel = MicrosoftExcel;
