// n8n-nodes-base/dist/nodes/ReadBinaryFiles/ReadBinaryFiles.node.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReadBinaryFiles = void 0;
const fast_glob_1 = __importDefault(require("fast-glob"));
class ReadBinaryFiles {
    constructor() {
        this.description = {
            displayName: 'Read Binary Files',
            name: 'readBinaryFiles',
            icon: 'fa:file-import',
            group: ['input'],
            version: 1,
            description: 'Reads binary files from disk',
            defaults: {
                name: 'Read Binary Files',
                color: '#44AA44',
            },
            inputs: ['main'],
            outputs: ['main'],
            properties: [
                {
                    displayName: 'File Selector',
                    name: 'fileSelector',
                    type: 'string',
                    default: '',
                    required: true,
                    placeholder: '*.jpg',
                    description: 'Pattern for files to read',
                },
                {
                    displayName: 'Property Name',
                    name: 'dataPropertyName',
                    type: 'string',
                    default: 'data',
                    required: true,
                    description: 'Name of the binary property to which to write the data of the read files',
                },
            ],
        };
    }
    async execute() {
        const fileSelector = this.getNodeParameter('fileSelector', 0);
        const dataPropertyName = this.getNodeParameter('dataPropertyName', 0);
        const files = await (0, fast_glob_1.default)(fileSelector);
        const items = [];
        for (const filePath of files) {
            const stream = await this.helpers.createReadStream(filePath);
            items.push({
                binary: {
                    [dataPropertyName]: await this.helpers.prepareBinaryData(stream, filePath),
                },
                json: {},
                pairedItem: {
                    item: 0,
                },
            });
        }
        return this.prepareOutputData(items);
    }
}
exports.ReadBinaryFiles = ReadBinaryFiles;
