// n8n-nodes-base/dist/nodes/Google/BigQuery/GoogleBigQuery.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoogleBigQuery = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const GoogleBigQueryV1_node_1 = require("./v1/GoogleBigQueryV1.node");
const GoogleBigQueryV2_node_1 = require("./v2/GoogleBigQueryV2.node");
class GoogleBigQuery extends n8n_workflow_1.VersionedNodeType {
    constructor() {
        const baseDescription = {
            displayName: 'Google BigQuery',
            name: 'googleBigQuery',
            icon: 'file:googleBigQuery.svg',
            group: ['input'],
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            description: 'Consume Google BigQuery API',
            defaultVersion: 2,
        };
        const nodeVersions = {
            1: new GoogleBigQueryV1_node_1.GoogleBigQueryV1(baseDescription),
            2: new GoogleBigQueryV2_node_1.GoogleBigQueryV2(baseDescription),
        };
        super(nodeVersions, baseDescription);
    }
}
exports.GoogleBigQuery = GoogleBigQuery;
