// n8n-workflow/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NativeMethods = exports.ExpressionExtensions = exports.isINodePropertyOptionsList = exports.isINodePropertyCollectionList = exports.isINodePropertiesList = exports.isINodePropertyCollection = exports.isINodePropertyOptions = exports.isINodeProperties = exports.assert = exports.fileTypeFromMimeType = exports.sleep = exports.jsonStringify = exports.jsonParse = exports.deepCopy = exports.isObjectEmpty = exports.TelemetryHelpers = exports.ObservableObject = exports.NodeHelpers = exports.LoggerProxy = exports.ErrorReporterProxy = void 0;
const LoggerProxy = __importStar(require("./LoggerProxy"));
exports.LoggerProxy = LoggerProxy;
exports.ErrorReporterProxy = __importStar(require("./ErrorReporterProxy"));
const NodeHelpers = __importStar(require("./NodeHelpers"));
exports.NodeHelpers = NodeHelpers;
const ObservableObject = __importStar(require("./ObservableObject"));
exports.ObservableObject = ObservableObject;
const TelemetryHelpers = __importStar(require("./TelemetryHelpers"));
exports.TelemetryHelpers = TelemetryHelpers;
__exportStar(require("./Authentication"), exports);
__exportStar(require("./Constants"), exports);
__exportStar(require("./Cron"), exports);
__exportStar(require("./DeferredPromise"), exports);
__exportStar(require("./Interfaces"), exports);
__exportStar(require("./MessageEventBus"), exports);
__exportStar(require("./ExecutionStatus"), exports);
__exportStar(require("./Expression"), exports);
__exportStar(require("./ExpressionError"), exports);
__exportStar(require("./NodeErrors"), exports);
__exportStar(require("./NodeHelpers"), exports);
__exportStar(require("./RoutingNode"), exports);
__exportStar(require("./Workflow"), exports);
__exportStar(require("./WorkflowActivationError"), exports);
__exportStar(require("./WorkflowDataProxy"), exports);
__exportStar(require("./WorkflowErrors"), exports);
__exportStar(require("./WorkflowHooks"), exports);
__exportStar(require("./VersionedNodeType"), exports);
var utils_1 = require("./utils");
Object.defineProperty(exports, "isObjectEmpty", { enumerable: true, get: function () { return utils_1.isObjectEmpty; } });
Object.defineProperty(exports, "deepCopy", { enumerable: true, get: function () { return utils_1.deepCopy; } });
Object.defineProperty(exports, "jsonParse", { enumerable: true, get: function () { return utils_1.jsonParse; } });
Object.defineProperty(exports, "jsonStringify", { enumerable: true, get: function () { return utils_1.jsonStringify; } });
Object.defineProperty(exports, "sleep", { enumerable: true, get: function () { return utils_1.sleep; } });
Object.defineProperty(exports, "fileTypeFromMimeType", { enumerable: true, get: function () { return utils_1.fileTypeFromMimeType; } });
Object.defineProperty(exports, "assert", { enumerable: true, get: function () { return utils_1.assert; } });
var type_guards_1 = require("./type-guards");
Object.defineProperty(exports, "isINodeProperties", { enumerable: true, get: function () { return type_guards_1.isINodeProperties; } });
Object.defineProperty(exports, "isINodePropertyOptions", { enumerable: true, get: function () { return type_guards_1.isINodePropertyOptions; } });
Object.defineProperty(exports, "isINodePropertyCollection", { enumerable: true, get: function () { return type_guards_1.isINodePropertyCollection; } });
Object.defineProperty(exports, "isINodePropertiesList", { enumerable: true, get: function () { return type_guards_1.isINodePropertiesList; } });
Object.defineProperty(exports, "isINodePropertyCollectionList", { enumerable: true, get: function () { return type_guards_1.isINodePropertyCollectionList; } });
Object.defineProperty(exports, "isINodePropertyOptionsList", { enumerable: true, get: function () { return type_guards_1.isINodePropertyOptionsList; } });
var Extensions_1 = require("./Extensions");
Object.defineProperty(exports, "ExpressionExtensions", { enumerable: true, get: function () { return Extensions_1.ExpressionExtensions; } });
var NativeMethods_1 = require("./NativeMethods");
Object.defineProperty(exports, "NativeMethods", { enumerable: true, get: function () { return NativeMethods_1.NativeMethods; } });
