// n8n/dist/databases/utils/transformers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sqlite = exports.objectRetriever = exports.lowerCaser = exports.idStringifier = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const config_1 = __importDefault(require("../../config"));
exports.idStringifier = {
    from: (value) => value === null || value === void 0 ? void 0 : value.toString(),
    to: (value) => typeof value === 'string' ? Number(value) : value,
};
exports.lowerCaser = {
    from: (value) => value,
    to: (value) => (typeof value === 'string' ? value.toLowerCase() : value),
};
exports.objectRetriever = {
    to: (value) => value,
    from: (value) => (typeof value === 'string' ? (0, n8n_workflow_1.jsonParse)(value) : value),
};
const jsonColumn = {
    to: (value) => config_1.default.getEnv('database.type') === 'sqlite' ? JSON.stringify(value) : value,
    from: (value) => (typeof value === 'string' ? (0, n8n_workflow_1.jsonParse)(value) : value),
};
exports.sqlite = { jsonColumn };
