// n8n-nodes-base/dist/nodes/FormIo/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.formIoApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function getToken(credentials) {
    const base = credentials.domain || 'https://formio.form.io';
    const options = {
        headers: {
            'Content-Type': 'application/json',
        },
        method: 'POST',
        body: {
            data: {
                email: credentials.email,
                password: credentials.password,
            },
        },
        uri: `${base}/user/login`,
        json: true,
        resolveWithFullResponse: true,
    };
    try {
        const responseObject = await this.helpers.request(options);
        return responseObject.headers['x-jwt-token'];
    }
    catch (error) {
        throw new Error('Authentication Failed for Form.io. Please provide valid credentails/ endpoint details');
    }
}
async function formIoApiRequest(method, endpoint, body = {}, qs = {}) {
    const credentials = (await this.getCredentials('formIoApi'));
    const token = await getToken.call(this, credentials);
    const base = credentials.domain || 'https://api.form.io';
    const options = {
        headers: {
            'Content-Type': 'application/json',
            'x-jwt-token': token,
        },
        method,
        body,
        qs,
        uri: `${base}${endpoint}`,
        json: true,
    };
    try {
        return await this.helpers.request.call(this, options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.formIoApiRequest = formIoApiRequest;
