// n8n-nodes-base/dist/nodes/Zoho/GenericFunctions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.addGetAllFilterOptions = exports.getPicklistOptions = exports.capitalizeInitial = exports.getFields = exports.getModuleName = exports.toLoadOptions = exports.adjustProductPayload = exports.adjustVendorPayload = exports.adjustSalesOrderPayload = exports.adjustQuotePayload = exports.adjustPurchaseOrderPayload = exports.adjustLeadPayload = exports.adjustInvoicePayloadOnUpdate = exports.adjustInvoicePayload = exports.adjustDealPayload = exports.adjustContactPayload = exports.adjustAccountPayload = exports.adjustProductDetailsOnUpdate = exports.adjustProductDetails = exports.throwOnMissingProducts = exports.throwOnEmptyUpdate = exports.handleListing = exports.zohoApiRequestAllItems = exports.zohoApiRequest = exports.throwOnErrorStatus = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const lodash_flow_1 = __importDefault(require("lodash.flow"));
const lodash_sortby_1 = __importDefault(require("lodash.sortby"));
function throwOnErrorStatus(responseData) {
    var _a;
    if (((_a = responseData === null || responseData === void 0 ? void 0 : responseData.data) === null || _a === void 0 ? void 0 : _a[0].status) === 'error') {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), responseData);
    }
}
exports.throwOnErrorStatus = throwOnErrorStatus;
async function zohoApiRequest(method, endpoint, body = {}, qs = {}, uri) {
    var _a;
    const { oauthTokenData } = (await this.getCredentials('zohoOAuth2Api'));
    const options = {
        body: {
            data: [body],
        },
        method,
        qs,
        uri: uri || `${oauthTokenData.api_domain}/crm/v2${endpoint}`,
        json: true,
    };
    if (!Object.keys(body).length) {
        delete options.body;
    }
    if (!Object.keys(qs).length) {
        delete options.qs;
    }
    try {
        const responseData = await ((_a = this.helpers.requestOAuth2) === null || _a === void 0 ? void 0 : _a.call(this, 'zohoOAuth2Api', options));
        if (responseData === undefined)
            return [];
        throwOnErrorStatus.call(this, responseData);
        return responseData;
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.zohoApiRequest = zohoApiRequest;
async function zohoApiRequestAllItems(method, endpoint, body = {}, qs = {}) {
    const returnData = [];
    let responseData;
    qs.per_page = 200;
    qs.page = 1;
    do {
        responseData = await zohoApiRequest.call(this, method, endpoint, body, qs);
        if (Array.isArray(responseData) && !responseData.length)
            return returnData;
        returnData.push(...responseData.data);
        qs.page++;
    } while (responseData.info.more_records !== undefined && responseData.info.more_records === true);
    return returnData;
}
exports.zohoApiRequestAllItems = zohoApiRequestAllItems;
async function handleListing(method, endpoint, body = {}, qs = {}) {
    const returnAll = this.getNodeParameter('returnAll', 0);
    if (returnAll) {
        return zohoApiRequestAllItems.call(this, method, endpoint, body, qs);
    }
    const responseData = await zohoApiRequestAllItems.call(this, method, endpoint, body, qs);
    const limit = this.getNodeParameter('limit', 0);
    return responseData.slice(0, limit);
}
exports.handleListing = handleListing;
function throwOnEmptyUpdate(resource) {
    throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Please enter at least one field to update for the ${resource}.`);
}
exports.throwOnEmptyUpdate = throwOnEmptyUpdate;
function throwOnMissingProducts(resource, productDetails) {
    if (!productDetails.length) {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Please enter at least one product for the ${resource}.`);
    }
}
exports.throwOnMissingProducts = throwOnMissingProducts;
const omit = (propertyToOmit, { [propertyToOmit]: _, ...remainingObject }) => remainingObject;
const adjustProductDetails = (productDetails) => {
    return productDetails.map((p) => {
        return {
            ...omit('product', p),
            product: { id: p.id },
            quantity: p.quantity || 1,
        };
    });
};
exports.adjustProductDetails = adjustProductDetails;
const adjustProductDetailsOnUpdate = (allFields) => {
    if (!allFields.Product_Details)
        return allFields;
    return allFields.Product_Details.map((p) => {
        return {
            ...omit('product', p),
            product: { id: p.id },
            quantity: p.quantity || 1,
        };
    });
};
exports.adjustProductDetailsOnUpdate = adjustProductDetailsOnUpdate;
const adjustLocationFields = (locationType) => (allFields) => {
    const locationField = allFields[locationType];
    if (!locationField)
        return allFields;
    return {
        ...omit(locationType, allFields),
        ...locationField.address_fields,
    };
};
const adjustAddressFields = adjustLocationFields('Address');
const adjustBillingAddressFields = adjustLocationFields('Billing_Address');
const adjustMailingAddressFields = adjustLocationFields('Mailing_Address');
const adjustShippingAddressFields = adjustLocationFields('Shipping_Address');
const adjustOtherAddressFields = adjustLocationFields('Other_Address');
const adjustDateField = (dateType) => (allFields) => {
    const dateField = allFields[dateType];
    if (!dateField)
        return allFields;
    allFields[dateType] = dateField.split('T')[0];
    return allFields;
};
const adjustDateOfBirthField = adjustDateField('Date_of_Birth');
const adjustClosingDateField = adjustDateField('Closing_Date');
const adjustInvoiceDateField = adjustDateField('Invoice_Date');
const adjustDueDateField = adjustDateField('Due_Date');
const adjustPurchaseOrderDateField = adjustDateField('PO_Date');
const adjustValidTillField = adjustDateField('Valid_Till');
const adjustIdField = (idType, nameProperty) => (allFields) => {
    const idValue = allFields[idType];
    if (!idValue)
        return allFields;
    return {
        ...omit(idType, allFields),
        [nameProperty]: { id: idValue },
    };
};
const adjustAccountIdField = adjustIdField('accountId', 'Account_Name');
const adjustContactIdField = adjustIdField('contactId', 'Full_Name');
const adjustDealIdField = adjustIdField('dealId', 'Deal_Name');
const adjustCustomFields = (allFields) => {
    const { customFields, ...rest } = allFields;
    if (!(customFields === null || customFields === void 0 ? void 0 : customFields.customFields.length))
        return allFields;
    return customFields.customFields.reduce((acc, cur) => {
        acc[cur.fieldId] = cur.value;
        return acc;
    }, rest);
};
exports.adjustAccountPayload = (0, lodash_flow_1.default)(adjustBillingAddressFields, adjustShippingAddressFields, adjustCustomFields);
exports.adjustContactPayload = (0, lodash_flow_1.default)(adjustMailingAddressFields, adjustOtherAddressFields, adjustDateOfBirthField, adjustCustomFields);
exports.adjustDealPayload = (0, lodash_flow_1.default)(adjustClosingDateField, adjustCustomFields);
exports.adjustInvoicePayload = (0, lodash_flow_1.default)(adjustBillingAddressFields, adjustShippingAddressFields, adjustInvoiceDateField, adjustDueDateField, adjustAccountIdField, adjustCustomFields);
exports.adjustInvoicePayloadOnUpdate = (0, lodash_flow_1.default)(exports.adjustInvoicePayload, exports.adjustProductDetailsOnUpdate);
exports.adjustLeadPayload = (0, lodash_flow_1.default)(adjustAddressFields, adjustCustomFields);
exports.adjustPurchaseOrderPayload = (0, lodash_flow_1.default)(adjustBillingAddressFields, adjustShippingAddressFields, adjustDueDateField, adjustPurchaseOrderDateField, adjustCustomFields);
exports.adjustQuotePayload = (0, lodash_flow_1.default)(adjustBillingAddressFields, adjustShippingAddressFields, adjustValidTillField, adjustCustomFields);
exports.adjustSalesOrderPayload = (0, lodash_flow_1.default)(adjustBillingAddressFields, adjustShippingAddressFields, adjustDueDateField, adjustAccountIdField, adjustContactIdField, adjustDealIdField, adjustCustomFields);
exports.adjustVendorPayload = (0, lodash_flow_1.default)(adjustAddressFields, adjustCustomFields);
exports.adjustProductPayload = adjustCustomFields;
const toLoadOptions = (items, nameProperty) => items.map((item) => ({ name: item[nameProperty], value: item.id }));
exports.toLoadOptions = toLoadOptions;
function getModuleName(resource) {
    const map = {
        account: 'Accounts',
        contact: 'Contacts',
        deal: 'Deals',
        invoice: 'Invoices',
        lead: 'Leads',
        product: 'Products',
        purchaseOrder: 'Purchase_Orders',
        salesOrder: 'Sales_Orders',
        vendor: 'Vendors',
        quote: 'Quotes',
    };
    return map[resource];
}
exports.getModuleName = getModuleName;
async function getFields(resource, { onlyCustom } = { onlyCustom: false }) {
    const qs = { module: getModuleName(resource) };
    let { fields } = (await zohoApiRequest.call(this, 'GET', '/settings/fields', {}, qs));
    if (onlyCustom) {
        fields = fields.filter(({ custom_field }) => custom_field);
    }
    const options = fields.map(({ field_label, api_name }) => ({
        name: field_label,
        value: api_name,
    }));
    return (0, lodash_sortby_1.default)(options, (o) => o.name);
}
exports.getFields = getFields;
const capitalizeInitial = (str) => str[0].toUpperCase() + str.slice(1);
exports.capitalizeInitial = capitalizeInitial;
function getSectionApiName(resource) {
    if (resource === 'purchaseOrder')
        return 'Purchase Order Information';
    if (resource === 'salesOrder')
        return 'Sales Order Information';
    return `${(0, exports.capitalizeInitial)(resource)} Information`;
}
async function getPicklistOptions(resource, targetField) {
    var _a, _b;
    const qs = { module: getModuleName(resource) };
    const responseData = (await zohoApiRequest.call(this, 'GET', '/settings/layouts', {}, qs));
    const pickListOptions = (_b = (_a = responseData.layouts[0].sections
        .find((section) => section.api_name === getSectionApiName(resource))) === null || _a === void 0 ? void 0 : _a.fields.find((f) => f.api_name === targetField)) === null || _b === void 0 ? void 0 : _b.pick_list_values;
    if (!pickListOptions)
        return [];
    return pickListOptions.map((option) => ({
        name: option.display_value,
        value: option.actual_value,
    }));
}
exports.getPicklistOptions = getPicklistOptions;
const addGetAllFilterOptions = (qs, options) => {
    if (Object.keys(options).length) {
        const { fields, ...rest } = options;
        Object.assign(qs, fields && { fields: fields.join(',') }, rest);
    }
};
exports.addGetAllFilterOptions = addGetAllFilterOptions;
