// n8n-nodes-base/dist/nodes/Pipedrive/PipedriveTrigger.node.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PipedriveTrigger = void 0;
const GenericFunctions_1 = require("./GenericFunctions");
const basic_auth_1 = __importDefault(require("basic-auth"));
function authorizationError(resp, realm, responseCode, message) {
    if (message === undefined) {
        message = 'Authorization problem!';
        if (responseCode === 401) {
            message = 'Authorization is required!';
        }
        else if (responseCode === 403) {
            message = 'Authorization data is wrong!';
        }
    }
    resp.writeHead(responseCode, { 'WWW-Authenticate': `Basic realm="${realm}"` });
    resp.end(message);
    return {
        noWebhookResponse: true,
    };
}
class PipedriveTrigger {
    constructor() {
        this.description = {
            displayName: 'Pipedrive Trigger',
            name: 'pipedriveTrigger',
            icon: 'file:pipedrive.svg',
            group: ['trigger'],
            version: 1,
            description: 'Starts the workflow when Pipedrive events occur',
            defaults: {
                name: 'Pipedrive Trigger',
            },
            inputs: [],
            outputs: ['main'],
            credentials: [
                {
                    name: 'pipedriveApi',
                    required: true,
                    displayOptions: {
                        show: {
                            authentication: ['apiToken'],
                        },
                    },
                },
                {
                    name: 'pipedriveOAuth2Api',
                    required: true,
                    displayOptions: {
                        show: {
                            authentication: ['oAuth2'],
                        },
                    },
                },
                {
                    name: 'httpBasicAuth',
                    required: true,
                    displayOptions: {
                        show: {
                            incomingAuthentication: ['basicAuth'],
                        },
                    },
                },
            ],
            webhooks: [
                {
                    name: 'default',
                    httpMethod: 'POST',
                    responseMode: 'onReceived',
                    path: 'webhook',
                },
            ],
            properties: [
                {
                    displayName: 'Authentication',
                    name: 'authentication',
                    type: 'options',
                    options: [
                        {
                            name: 'API Token',
                            value: 'apiToken',
                        },
                        {
                            name: 'OAuth2',
                            value: 'oAuth2',
                        },
                    ],
                    default: 'apiToken',
                },
                {
                    displayName: 'Incoming Authentication',
                    name: 'incomingAuthentication',
                    type: 'options',
                    options: [
                        {
                            name: 'Basic Auth',
                            value: 'basicAuth',
                        },
                        {
                            name: 'None',
                            value: 'none',
                        },
                    ],
                    default: 'none',
                    description: 'If authentication should be activated for the webhook (makes it more secure)',
                },
                {
                    displayName: 'Action',
                    name: 'action',
                    type: 'options',
                    options: [
                        {
                            name: 'Added',
                            value: 'added',
                            description: 'Data got added',
                            action: 'Data was added',
                        },
                        {
                            name: 'All',
                            value: '*',
                            description: 'Any change',
                            action: 'Any change',
                        },
                        {
                            name: 'Deleted',
                            value: 'deleted',
                            description: 'Data got deleted',
                            action: 'Data was deleted',
                        },
                        {
                            name: 'Merged',
                            value: 'merged',
                            description: 'Data got merged',
                            action: 'Data was merged',
                        },
                        {
                            name: 'Updated',
                            value: 'updated',
                            description: 'Data got updated',
                            action: 'Data was updated',
                        },
                    ],
                    default: '*',
                    description: 'Type of action to receive notifications about',
                },
                {
                    displayName: 'Object',
                    name: 'object',
                    type: 'options',
                    options: [
                        {
                            name: 'Activity',
                            value: 'activity',
                        },
                        {
                            name: 'Activity Type',
                            value: 'activityType',
                        },
                        {
                            name: 'All',
                            value: '*',
                        },
                        {
                            name: 'Deal',
                            value: 'deal',
                        },
                        {
                            name: 'Note',
                            value: 'note',
                        },
                        {
                            name: 'Organization',
                            value: 'organization',
                        },
                        {
                            name: 'Person',
                            value: 'person',
                        },
                        {
                            name: 'Pipeline',
                            value: 'pipeline',
                        },
                        {
                            name: 'Product',
                            value: 'product',
                        },
                        {
                            name: 'Stage',
                            value: 'stage',
                        },
                        {
                            name: 'User',
                            value: 'user',
                        },
                    ],
                    default: '*',
                    description: 'Type of object to receive notifications about',
                },
            ],
        };
        this.webhookMethods = {
            default: {
                async checkExists() {
                    const webhookUrl = this.getNodeWebhookUrl('default');
                    const webhookData = this.getWorkflowStaticData('node');
                    const eventAction = this.getNodeParameter('action');
                    const eventObject = this.getNodeParameter('object');
                    const endpoint = '/webhooks';
                    const responseData = await GenericFunctions_1.pipedriveApiRequest.call(this, 'GET', endpoint, {});
                    if (responseData.data === undefined) {
                        return false;
                    }
                    for (const existingData of responseData.data) {
                        if (existingData.subscription_url === webhookUrl &&
                            existingData.event_action === eventAction &&
                            existingData.event_object === eventObject) {
                            webhookData.webhookId = existingData.id;
                            return true;
                        }
                    }
                    return false;
                },
                async create() {
                    const webhookUrl = this.getNodeWebhookUrl('default');
                    const incomingAuthentication = this.getNodeParameter('incomingAuthentication', 0);
                    const eventAction = this.getNodeParameter('action');
                    const eventObject = this.getNodeParameter('object');
                    const endpoint = '/webhooks';
                    const body = {
                        event_action: eventAction,
                        event_object: eventObject,
                        subscription_url: webhookUrl,
                        http_auth_user: undefined,
                        http_auth_password: undefined,
                    };
                    if (incomingAuthentication === 'basicAuth') {
                        let httpBasicAuth;
                        try {
                            httpBasicAuth = await this.getCredentials('httpBasicAuth');
                        }
                        catch (error) {
                        }
                        if (httpBasicAuth === undefined || !httpBasicAuth.user || !httpBasicAuth.password) {
                            return false;
                        }
                        body.http_auth_user = httpBasicAuth.user;
                        body.http_auth_password = httpBasicAuth.password;
                    }
                    const responseData = await GenericFunctions_1.pipedriveApiRequest.call(this, 'POST', endpoint, body);
                    if (responseData.data === undefined || responseData.data.id === undefined) {
                        return false;
                    }
                    const webhookData = this.getWorkflowStaticData('node');
                    webhookData.webhookId = responseData.data.id;
                    return true;
                },
                async delete() {
                    const webhookData = this.getWorkflowStaticData('node');
                    if (webhookData.webhookId !== undefined) {
                        const endpoint = `/webhooks/${webhookData.webhookId}`;
                        const body = {};
                        try {
                            await GenericFunctions_1.pipedriveApiRequest.call(this, 'DELETE', endpoint, body);
                        }
                        catch (error) {
                            return false;
                        }
                        delete webhookData.webhookId;
                        delete webhookData.webhookEvents;
                    }
                    return true;
                },
            },
        };
    }
    async webhook() {
        const req = this.getRequestObject();
        const resp = this.getResponseObject();
        const realm = 'Webhook';
        const incomingAuthentication = this.getNodeParameter('incomingAuthentication', 0);
        if (incomingAuthentication === 'basicAuth') {
            let httpBasicAuth;
            try {
                httpBasicAuth = await this.getCredentials('httpBasicAuth');
            }
            catch (error) {
            }
            if (httpBasicAuth === undefined || !httpBasicAuth.user || !httpBasicAuth.password) {
                return authorizationError(resp, realm, 500, 'No authentication data defined on node!');
            }
            const basicAuthData = (0, basic_auth_1.default)(req);
            if (basicAuthData === undefined) {
                return authorizationError(resp, realm, 401);
            }
            if (basicAuthData.name !== httpBasicAuth.user ||
                basicAuthData.pass !== httpBasicAuth.password) {
                return authorizationError(resp, realm, 403);
            }
        }
        return {
            workflowData: [this.helpers.returnJsonArray(req.body)],
        };
    }
}
exports.PipedriveTrigger = PipedriveTrigger;
