// n8n-nodes-base/dist/nodes/Github/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.githubApiRequestAllItems = exports.getFileSha = exports.githubApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function githubApiRequest(method, endpoint, body, query, option = {}) {
    const options = {
        method,
        headers: {
            'User-Agent': 'n8n',
        },
        body,
        qs: query,
        uri: '',
        json: true,
    };
    if (Object.keys(option).length !== 0) {
        Object.assign(options, option);
    }
    try {
        const authenticationMethod = this.getNodeParameter('authentication', 0, 'accessToken');
        let credentialType = '';
        if (authenticationMethod === 'accessToken') {
            const credentials = await this.getCredentials('githubApi');
            credentialType = 'githubApi';
            const baseUrl = credentials.server || 'https://api.github.com';
            options.uri = `${baseUrl}${endpoint}`;
        }
        else {
            const credentials = await this.getCredentials('githubOAuth2Api');
            credentialType = 'githubOAuth2Api';
            const baseUrl = credentials.server || 'https://api.github.com';
            options.uri = `${baseUrl}${endpoint}`;
        }
        return await this.helpers.requestWithAuthentication.call(this, credentialType, options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.githubApiRequest = githubApiRequest;
async function getFileSha(owner, repository, filePath, branch) {
    const getBody = {};
    if (branch !== undefined) {
        getBody.branch = branch;
    }
    const getEndpoint = `/repos/${owner}/${repository}/contents/${encodeURI(filePath)}`;
    const responseData = await githubApiRequest.call(this, 'GET', getEndpoint, getBody, {});
    if (responseData.sha === undefined) {
        throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Could not get the SHA of the file.');
    }
    return responseData.sha;
}
exports.getFileSha = getFileSha;
async function githubApiRequestAllItems(method, endpoint, body = {}, query = {}) {
    var _a;
    const returnData = [];
    let responseData;
    query.per_page = 100;
    query.page = 1;
    do {
        responseData = await githubApiRequest.call(this, method, endpoint, body, query, {
            resolveWithFullResponse: true,
        });
        query.page++;
        returnData.push.apply(returnData, responseData.body);
    } while ((_a = responseData.headers.link) === null || _a === void 0 ? void 0 : _a.includes('next'));
    return returnData;
}
exports.githubApiRequestAllItems = githubApiRequestAllItems;
