// n8n-nodes-base/dist/nodes/Slack/V1/UserDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.userFields = exports.userOperations = void 0;
exports.userOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                resource: ['user'],
            },
        },
        options: [
            {
                name: 'Info',
                value: 'info',
                description: 'Get information about a user',
                action: 'Get information about a user',
            },
            {
                name: 'Get Many',
                value: 'getAll',
                description: 'Get a list of many users',
                action: 'Get many users',
            },
            {
                name: 'Get Presence',
                value: 'getPresence',
                description: 'Get online status of a user',
                action: "Get a user's presence status",
            },
        ],
        default: 'info',
    },
];
exports.userFields = [
    {
        displayName: 'User ID',
        name: 'user',
        type: 'string',
        typeOptions: {
            loadOptionsMethod: 'getUsers',
        },
        default: '',
        displayOptions: {
            show: {
                operation: ['info'],
                resource: ['user'],
            },
        },
        required: true,
        description: 'The ID of the user to get information about',
    },
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        displayOptions: {
            show: {
                resource: ['user'],
                operation: ['getAll'],
            },
        },
        default: false,
        description: 'Whether to return all results or only up to a given limit',
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        displayOptions: {
            show: {
                resource: ['user'],
                operation: ['getAll'],
                returnAll: [false],
            },
        },
        typeOptions: {
            minValue: 1,
            maxValue: 100,
        },
        default: 50,
        description: 'Max number of results to return',
    },
    {
        displayName: 'User ID',
        name: 'user',
        type: 'string',
        typeOptions: {
            loadOptionsMethod: 'getUsers',
        },
        default: '',
        displayOptions: {
            show: {
                operation: ['getPresence'],
                resource: ['user'],
            },
        },
        required: true,
        description: 'The ID of the user to get the online status of',
    },
];
