// n8n-nodes-base/dist/nodes/SyncroMSP/v1/actions/rmm/getAll/description.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rmmGetAllDescription = void 0;
exports.rmmGetAllDescription = [
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        displayOptions: {
            show: {
                resource: ['rmm'],
                operation: ['getAll'],
            },
        },
        default: false,
        noDataExpression: true,
        description: 'Whether to return all results or only up to a given limit',
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        typeOptions: {
            minValue: 1,
        },
        displayOptions: {
            show: {
                resource: ['rmm'],
                operation: ['getAll'],
                returnAll: [false],
            },
        },
        default: 25,
        description: 'Max number of results to return',
    },
    {
        displayName: 'Filters',
        name: 'filters',
        type: 'collection',
        placeholder: 'Add Filter',
        displayOptions: {
            show: {
                resource: ['rmm'],
                operation: ['getAll'],
            },
        },
        default: {},
        options: [
            {
                displayName: 'Status',
                name: 'status',
                type: 'options',
                options: [
                    {
                        name: 'Active',
                        value: 'active',
                    },
                    {
                        name: 'All',
                        value: 'all',
                    },
                    {
                        name: 'Resolved',
                        value: 'resolved',
                    },
                ],
                default: 'all',
            },
        ],
    },
];
