// n8n-nodes-base/dist/nodes/MoveBinaryData/MoveBinaryData.node.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MoveBinaryData = void 0;
const lodash_get_1 = __importDefault(require("lodash.get"));
const lodash_set_1 = __importDefault(require("lodash.set"));
const lodash_unset_1 = __importDefault(require("lodash.unset"));
const pretty_bytes_1 = __importDefault(require("pretty-bytes"));
const n8n_workflow_1 = require("n8n-workflow");
const iconv_lite_1 = __importDefault(require("iconv-lite"));
iconv_lite_1.default.encodingExists('utf8');
const bomAware = [];
const encodeDecodeOptions = [];
const encodings = iconv_lite_1.default.encodings;
Object.keys(encodings).forEach((encoding) => {
    if (!(encoding.startsWith('_') || typeof encodings[encoding] === 'string')) {
        if (encodings[encoding].bomAware) {
            bomAware.push(encoding);
        }
        encodeDecodeOptions.push({ name: encoding, value: encoding });
    }
});
encodeDecodeOptions.sort((a, b) => {
    if (a.name < b.name) {
        return -1;
    }
    if (a.name > b.name) {
        return 1;
    }
    return 0;
});
class MoveBinaryData {
    constructor() {
        this.description = {
            displayName: 'Move Binary Data',
            name: 'moveBinaryData',
            icon: 'fa:exchange-alt',
            group: ['transform'],
            version: 1,
            subtitle: '={{$parameter["mode"]==="binaryToJson" ? "Binary to JSON" : "JSON to Binary"}}',
            description: 'Move data between binary and JSON properties',
            defaults: {
                name: 'Move Binary Data',
                color: '#7722CC',
            },
            inputs: ['main'],
            outputs: ['main'],
            properties: [
                {
                    displayName: 'Mode',
                    name: 'mode',
                    type: 'options',
                    options: [
                        {
                            name: 'Binary to JSON',
                            value: 'binaryToJson',
                            description: 'Move data from Binary to JSON',
                        },
                        {
                            name: 'JSON to Binary',
                            value: 'jsonToBinary',
                            description: 'Move data from JSON to Binary',
                        },
                    ],
                    default: 'binaryToJson',
                    description: 'From and to where data should be moved',
                },
                {
                    displayName: 'Set All Data',
                    name: 'setAllData',
                    type: 'boolean',
                    displayOptions: {
                        show: {
                            mode: ['binaryToJson'],
                        },
                    },
                    default: true,
                    description: 'Whether all JSON data should be replaced with the data retrieved from binary key. Else the data will be written to a single key.',
                },
                {
                    displayName: 'Source Key',
                    name: 'sourceKey',
                    type: 'string',
                    displayOptions: {
                        show: {
                            mode: ['binaryToJson'],
                        },
                    },
                    default: 'data',
                    required: true,
                    placeholder: 'data',
                    description: 'The name of the binary key to get data from. It is also possible to define deep keys by using dot-notation like for example: "level1.level2.currentKey".',
                },
                {
                    displayName: 'Destination Key',
                    name: 'destinationKey',
                    type: 'string',
                    displayOptions: {
                        show: {
                            mode: ['binaryToJson'],
                            setAllData: [false],
                        },
                    },
                    default: 'data',
                    required: true,
                    placeholder: '',
                    description: 'The name the JSON key to copy data to. It is also possible to define deep keys by using dot-notation like for example: "level1.level2.newKey".',
                },
                {
                    displayName: 'Convert All Data',
                    name: 'convertAllData',
                    type: 'boolean',
                    displayOptions: {
                        show: {
                            mode: ['jsonToBinary'],
                        },
                    },
                    default: true,
                    description: 'Whether all JSON data should be converted to binary. Else only the data of one key will be converted.',
                },
                {
                    displayName: 'Source Key',
                    name: 'sourceKey',
                    type: 'string',
                    displayOptions: {
                        show: {
                            convertAllData: [false],
                            mode: ['jsonToBinary'],
                        },
                    },
                    default: 'data',
                    required: true,
                    placeholder: 'data',
                    description: 'The name of the JSON key to get data from. It is also possible to define deep keys by using dot-notation like for example: "level1.level2.currentKey".',
                },
                {
                    displayName: 'Destination Key',
                    name: 'destinationKey',
                    type: 'string',
                    displayOptions: {
                        show: {
                            mode: ['jsonToBinary'],
                        },
                    },
                    default: 'data',
                    required: true,
                    placeholder: 'data',
                    description: 'The name the binary key to copy data to. It is also possible to define deep keys by using dot-notation like for example: "level1.level2.newKey".',
                },
                {
                    displayName: 'Options',
                    name: 'options',
                    type: 'collection',
                    placeholder: 'Add Option',
                    default: {},
                    options: [
                        {
                            displayName: 'Data Is Base64',
                            name: 'dataIsBase64',
                            type: 'boolean',
                            displayOptions: {
                                hide: {
                                    useRawData: [true],
                                },
                                show: {
                                    '/mode': ['jsonToBinary'],
                                    '/convertAllData': [false],
                                },
                            },
                            default: false,
                            description: 'Whether to keep the binary data as base64 string',
                        },
                        {
                            displayName: 'Encoding',
                            name: 'encoding',
                            type: 'options',
                            options: encodeDecodeOptions,
                            displayOptions: {
                                show: {
                                    '/mode': ['binaryToJson', 'jsonToBinary'],
                                },
                            },
                            default: 'utf8',
                            description: 'Set the encoding of the data stream',
                        },
                        {
                            displayName: 'Strip BOM',
                            name: 'stripBOM',
                            displayOptions: {
                                show: {
                                    '/mode': ['binaryToJson'],
                                    encoding: bomAware,
                                },
                            },
                            type: 'boolean',
                            default: true,
                        },
                        {
                            displayName: 'Add BOM',
                            name: 'addBOM',
                            displayOptions: {
                                show: {
                                    '/mode': ['jsonToBinary'],
                                    encoding: bomAware,
                                },
                            },
                            type: 'boolean',
                            default: false,
                        },
                        {
                            displayName: 'File Name',
                            name: 'fileName',
                            type: 'string',
                            displayOptions: {
                                show: {
                                    '/mode': ['jsonToBinary'],
                                },
                            },
                            default: '',
                            placeholder: 'example.json',
                            description: 'The file name to set',
                        },
                        {
                            displayName: 'JSON Parse',
                            name: 'jsonParse',
                            type: 'boolean',
                            displayOptions: {
                                hide: {
                                    keepAsBase64: [true],
                                },
                                show: {
                                    '/mode': ['binaryToJson'],
                                    '/setAllData': [false],
                                },
                            },
                            default: false,
                            description: 'Whether to run JSON parse on the data to get proper object data',
                        },
                        {
                            displayName: 'Keep Source',
                            name: 'keepSource',
                            type: 'boolean',
                            default: false,
                            description: 'Whether the source key should be kept. By default it will be deleted.',
                        },
                        {
                            displayName: 'Keep As Base64',
                            name: 'keepAsBase64',
                            type: 'boolean',
                            displayOptions: {
                                hide: {
                                    jsonParse: [true],
                                },
                                show: {
                                    '/mode': ['binaryToJson'],
                                    '/setAllData': [false],
                                },
                            },
                            default: false,
                            description: 'Whether to keep the binary data as base64 string',
                        },
                        {
                            displayName: 'Mime Type',
                            name: 'mimeType',
                            type: 'string',
                            displayOptions: {
                                show: {
                                    '/mode': ['jsonToBinary'],
                                },
                            },
                            default: 'application/json',
                            placeholder: 'application/json',
                            description: 'The mime-type to set. By default will the mime-type for JSON be set.',
                        },
                        {
                            displayName: 'Use Raw Data',
                            name: 'useRawData',
                            type: 'boolean',
                            displayOptions: {
                                hide: {
                                    dataIsBase64: [true],
                                },
                                show: {
                                    '/mode': ['jsonToBinary'],
                                },
                            },
                            default: false,
                            description: 'Whether to use data as is and do not JSON.stringify it',
                        },
                    ],
                },
            ],
        };
    }
    async execute() {
        const items = this.getInputData();
        const mode = this.getNodeParameter('mode', 0);
        const returnData = [];
        let item;
        let newItem;
        let options;
        for (let itemIndex = 0; itemIndex < items.length; itemIndex++) {
            item = items[itemIndex];
            options = this.getNodeParameter('options', itemIndex, {});
            newItem = {
                json: {},
                pairedItem: {
                    item: itemIndex,
                },
            };
            if (mode === 'binaryToJson') {
                const setAllData = this.getNodeParameter('setAllData', itemIndex);
                const sourceKey = this.getNodeParameter('sourceKey', itemIndex);
                const value = (0, lodash_get_1.default)(item.binary, sourceKey);
                if (value === undefined) {
                    continue;
                }
                const encoding = options.encoding || 'utf8';
                const buffer = await this.helpers.getBinaryDataBuffer(itemIndex, sourceKey);
                let convertedValue;
                if (setAllData) {
                    convertedValue = iconv_lite_1.default.decode(buffer, encoding, {
                        stripBOM: options.stripBOM,
                    });
                    newItem.json = (0, n8n_workflow_1.jsonParse)(convertedValue);
                }
                else {
                    newItem.json = (0, n8n_workflow_1.deepCopy)(item.json);
                    if (options.keepAsBase64 !== true) {
                        convertedValue = iconv_lite_1.default.decode(buffer, encoding, {
                            stripBOM: options.stripBOM,
                        });
                    }
                    else {
                        convertedValue = Buffer.from(buffer).toString(n8n_workflow_1.BINARY_ENCODING);
                    }
                    if (options.jsonParse) {
                        convertedValue = (0, n8n_workflow_1.jsonParse)(convertedValue);
                    }
                    const destinationKey = this.getNodeParameter('destinationKey', itemIndex, '');
                    (0, lodash_set_1.default)(newItem.json, destinationKey, convertedValue);
                }
                if (options.keepSource === true) {
                    newItem.binary = item.binary;
                }
                else {
                    newItem.binary = (0, n8n_workflow_1.deepCopy)(item.binary);
                    (0, lodash_unset_1.default)(newItem.binary, sourceKey);
                }
            }
            else if (mode === 'jsonToBinary') {
                const convertAllData = this.getNodeParameter('convertAllData', itemIndex);
                const destinationKey = this.getNodeParameter('destinationKey', itemIndex);
                const encoding = options.encoding || 'utf8';
                let value = item.json;
                if (!convertAllData) {
                    const sourceKey = this.getNodeParameter('sourceKey', itemIndex);
                    value = (0, lodash_get_1.default)(item.json, sourceKey);
                }
                if (value === undefined) {
                    continue;
                }
                if (item.binary !== undefined) {
                    newItem.binary = (0, n8n_workflow_1.deepCopy)(item.binary);
                }
                else {
                    newItem.binary = {};
                }
                const mimeType = options.mimeType || 'application/json';
                const convertedValue = {
                    data: '',
                    mimeType,
                    fileType: (0, n8n_workflow_1.fileTypeFromMimeType)(mimeType),
                };
                if (options.dataIsBase64 !== true) {
                    if (options.useRawData !== true || typeof value === 'object') {
                        value = JSON.stringify(value);
                    }
                    convertedValue.fileSize = (0, pretty_bytes_1.default)(value.length);
                    convertedValue.data = iconv_lite_1.default
                        .encode(value, encoding, { addBOM: options.addBOM })
                        .toString(n8n_workflow_1.BINARY_ENCODING);
                }
                else {
                    convertedValue.data = value;
                }
                if (options.fileName) {
                    convertedValue.fileName = options.fileName;
                }
                (0, lodash_set_1.default)(newItem.binary, destinationKey, convertedValue);
                if (options.keepSource === true) {
                    newItem.json = item.json;
                }
                else {
                    if (convertAllData) {
                        newItem.json = {};
                    }
                    else {
                        newItem.json = (0, n8n_workflow_1.deepCopy)(item.json);
                        const sourceKey = this.getNodeParameter('sourceKey', itemIndex);
                        (0, lodash_unset_1.default)(newItem.json, sourceKey);
                    }
                }
            }
            else {
                throw new n8n_workflow_1.NodeOperationError(this.getNode(), `The operation "${mode}" is not known!`, {
                    itemIndex,
                });
            }
            returnData.push(newItem);
        }
        return [returnData];
    }
}
exports.MoveBinaryData = MoveBinaryData;
