// n8n-nodes-base/dist/nodes/Mautic/CampaignContactDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.campaignContactFields = exports.campaignContactOperations = void 0;
exports.campaignContactOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                resource: ['campaignContact'],
            },
        },
        options: [
            {
                name: 'Add',
                value: 'add',
                description: 'Add contact to a campaign',
                action: 'Add a campaign contact',
            },
            {
                name: 'Remove',
                value: 'remove',
                description: 'Remove contact from a campaign',
                action: 'Remove a campaign contact',
            },
        ],
        default: 'add',
    },
];
exports.campaignContactFields = [
    {
        displayName: 'Contact ID',
        name: 'contactId',
        type: 'string',
        required: true,
        displayOptions: {
            show: {
                resource: ['campaignContact'],
                operation: ['add', 'remove'],
            },
        },
        default: '',
    },
    {
        displayName: 'Campaign Name or ID',
        name: 'campaignId',
        type: 'options',
        description: 'Choose from the list, or specify an ID using an <a href="https://docs.n8n.io/code-examples/expressions/">expression</a>',
        required: true,
        displayOptions: {
            show: {
                resource: ['campaignContact'],
                operation: ['add', 'remove'],
            },
        },
        typeOptions: {
            loadOptionsMethod: 'getCampaigns',
        },
        default: '',
    },
];
