// n8n-nodes-base/dist/nodes/ApiTemplateIo/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.downloadImage = exports.validateJSON = exports.loadResource = exports.apiTemplateIoApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function apiTemplateIoApiRequest(method, endpoint, qs = {}, body = {}) {
    const options = {
        headers: {
            'user-agent': 'n8n',
            Accept: 'application/json',
        },
        uri: `https://api.apitemplate.io/v1${endpoint}`,
        method,
        qs,
        body,
        followRedirect: true,
        followAllRedirects: true,
        json: true,
    };
    if (!Object.keys(body).length) {
        delete options.body;
    }
    if (!Object.keys(qs).length) {
        delete options.qs;
    }
    try {
        const response = await this.helpers.requestWithAuthentication.call(this, 'apiTemplateIoApi', options);
        if (response.status === 'error') {
            throw new n8n_workflow_1.NodeApiError(this.getNode(), response.message);
        }
        return response;
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.apiTemplateIoApiRequest = apiTemplateIoApiRequest;
async function loadResource(resource) {
    const target = resource === 'image' ? ['JPEG', 'PNG'] : ['PDF'];
    const templates = await apiTemplateIoApiRequest.call(this, 'GET', '/list-templates');
    const filtered = templates.filter(({ format }) => target.includes(format));
    return filtered.map(({ format, name, id }) => ({
        name: `${name} (${format})`,
        value: id,
    }));
}
exports.loadResource = loadResource;
function validateJSON(json) {
    let result;
    if (typeof json === 'object') {
        return json;
    }
    try {
        result = JSON.parse(json);
    }
    catch (exception) {
        result = undefined;
    }
    return result;
}
exports.validateJSON = validateJSON;
async function downloadImage(url) {
    return this.helpers.request({
        uri: url,
        method: 'GET',
        json: false,
        encoding: null,
    });
}
exports.downloadImage = downloadImage;
