// n8n-nodes-base/dist/nodes/Google/YouTube/VideoCategoryDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.videoCategoryFields = exports.videoCategoryOperations = void 0;
exports.videoCategoryOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                resource: ['videoCategory'],
            },
        },
        options: [
            {
                name: 'Get Many',
                value: 'getAll',
                description: 'Retrieve many video categories',
                action: 'Get many video categories',
            },
        ],
        default: 'getAll',
    },
];
exports.videoCategoryFields = [
    {
        displayName: 'Region Code',
        name: 'regionCode',
        type: 'options',
        description: 'Choose from the list, or specify an ID using an <a href="https://docs.n8n.io/code-examples/expressions/">expression</a>',
        required: true,
        displayOptions: {
            show: {
                operation: ['getAll'],
                resource: ['videoCategory'],
            },
        },
        typeOptions: {
            loadOptionsMethod: 'getCountriesCodes',
        },
        default: '',
    },
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        displayOptions: {
            show: {
                operation: ['getAll'],
                resource: ['videoCategory'],
            },
        },
        default: false,
        description: 'Whether to return all results or only up to a given limit',
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        displayOptions: {
            show: {
                operation: ['getAll'],
                resource: ['videoCategory'],
                returnAll: [false],
            },
        },
        typeOptions: {
            minValue: 1,
            maxValue: 50,
        },
        default: 25,
        description: 'Max number of results to return',
    },
];
