// n8n-nodes-base/dist/nodes/Google/YouTube/YouTube.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.YouTube = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const GenericFunctions_1 = require("./GenericFunctions");
const ChannelDescription_1 = require("./ChannelDescription");
const PlaylistDescription_1 = require("./PlaylistDescription");
const PlaylistItemDescription_1 = require("./PlaylistItemDescription");
const VideoDescription_1 = require("./VideoDescription");
const VideoCategoryDescription_1 = require("./VideoCategoryDescription");
const CountryCodes_1 = require("./CountryCodes");
const UPLOAD_CHUNK_SIZE = 1024 * 1024;
class YouTube {
    constructor() {
        this.description = {
            displayName: 'YouTube',
            name: 'youTube',
            icon: 'file:youTube.png',
            group: ['input'],
            version: 1,
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            description: 'Consume YouTube API',
            defaults: {
                name: 'YouTube',
            },
            inputs: ['main'],
            outputs: ['main'],
            credentials: [
                {
                    name: 'youTubeOAuth2Api',
                    required: true,
                },
            ],
            properties: [
                {
                    displayName: 'Resource',
                    name: 'resource',
                    type: 'options',
                    noDataExpression: true,
                    options: [
                        {
                            name: 'Channel',
                            value: 'channel',
                        },
                        {
                            name: 'Playlist',
                            value: 'playlist',
                        },
                        {
                            name: 'Playlist Item',
                            value: 'playlistItem',
                        },
                        {
                            name: 'Video',
                            value: 'video',
                        },
                        {
                            name: 'Video Category',
                            value: 'videoCategory',
                        },
                    ],
                    default: 'channel',
                },
                ...ChannelDescription_1.channelOperations,
                ...ChannelDescription_1.channelFields,
                ...PlaylistDescription_1.playlistOperations,
                ...PlaylistDescription_1.playlistFields,
                ...PlaylistItemDescription_1.playlistItemOperations,
                ...PlaylistItemDescription_1.playlistItemFields,
                ...VideoDescription_1.videoOperations,
                ...VideoDescription_1.videoFields,
                ...VideoCategoryDescription_1.videoCategoryOperations,
                ...VideoCategoryDescription_1.videoCategoryFields,
            ],
        };
        this.methods = {
            loadOptions: {
                async getLanguages() {
                    const returnData = [];
                    const languages = await GenericFunctions_1.googleApiRequestAllItems.call(this, 'items', 'GET', '/youtube/v3/i18nLanguages');
                    for (const language of languages) {
                        const languageName = language.id.toUpperCase();
                        const languageId = language.id;
                        returnData.push({
                            name: languageName,
                            value: languageId,
                        });
                    }
                    return returnData;
                },
                async getCountriesCodes() {
                    const returnData = [];
                    for (const countryCode of CountryCodes_1.countriesCodes) {
                        const countryCodeName = `${countryCode.name} - ${countryCode.alpha2}`;
                        const countryCodeId = countryCode.alpha2;
                        returnData.push({
                            name: countryCodeName,
                            value: countryCodeId,
                        });
                    }
                    return returnData;
                },
                async getVideoCategories() {
                    const countryCode = this.getCurrentNodeParameter('regionCode');
                    const returnData = [];
                    const qs = {};
                    qs.regionCode = countryCode;
                    qs.part = 'snippet';
                    const categories = await GenericFunctions_1.googleApiRequestAllItems.call(this, 'items', 'GET', '/youtube/v3/videoCategories', {}, qs);
                    for (const category of categories) {
                        const categoryName = category.snippet.title;
                        const categoryId = category.id;
                        returnData.push({
                            name: categoryName,
                            value: categoryId,
                        });
                    }
                    return returnData;
                },
                async getPlaylists() {
                    const returnData = [];
                    const qs = {};
                    qs.part = 'snippet';
                    qs.mine = true;
                    const playlists = await GenericFunctions_1.googleApiRequestAllItems.call(this, 'items', 'GET', '/youtube/v3/playlists', {}, qs);
                    for (const playlist of playlists) {
                        const playlistName = playlist.snippet.title;
                        const playlistId = playlist.id;
                        returnData.push({
                            name: playlistName,
                            value: playlistId,
                        });
                    }
                    return returnData;
                },
            },
        };
    }
    async execute() {
        var _a, _b, _c, _d;
        const items = this.getInputData();
        const returnData = [];
        const length = items.length;
        const qs = {};
        let responseData;
        const resource = this.getNodeParameter('resource', 0);
        const operation = this.getNodeParameter('operation', 0);
        for (let i = 0; i < length; i++) {
            try {
                if (resource === 'channel') {
                    if (operation === 'get') {
                        let part = this.getNodeParameter('part', i);
                        const channelId = this.getNodeParameter('channelId', i);
                        if (part.includes('*')) {
                            part = [
                                'brandingSettings',
                                'contentDetails',
                                'contentOwnerDetails',
                                'id',
                                'localizations',
                                'snippet',
                                'statistics',
                                'status',
                                'topicDetails',
                            ];
                        }
                        qs.part = part.join(',');
                        qs.id = channelId;
                        responseData = await GenericFunctions_1.googleApiRequest.call(this, 'GET', '/youtube/v3/channels', {}, qs);
                        responseData = responseData.items;
                    }
                    if (operation === 'getAll') {
                        const returnAll = this.getNodeParameter('returnAll', i);
                        let part = this.getNodeParameter('part', i);
                        const options = this.getNodeParameter('options', i);
                        const filters = this.getNodeParameter('filters', i);
                        if (part.includes('*')) {
                            part = [
                                'brandingSettings',
                                'contentDetails',
                                'contentOwnerDetails',
                                'id',
                                'localizations',
                                'snippet',
                                'statistics',
                                'status',
                                'topicDetails',
                            ];
                        }
                        qs.part = part.join(',');
                        Object.assign(qs, options, filters);
                        qs.mine = true;
                        if (qs.categoryId || qs.forUsername || qs.id || qs.managedByMe) {
                            delete qs.mine;
                        }
                        if (returnAll) {
                            responseData = await GenericFunctions_1.googleApiRequestAllItems.call(this, 'items', 'GET', '/youtube/v3/channels', {}, qs);
                        }
                        else {
                            qs.maxResults = this.getNodeParameter('limit', i);
                            responseData = await GenericFunctions_1.googleApiRequest.call(this, 'GET', '/youtube/v3/channels', {}, qs);
                            responseData = responseData.items;
                        }
                    }
                    if (operation === 'update') {
                        const channelId = this.getNodeParameter('channelId', i);
                        const updateFields = this.getNodeParameter('updateFields', i);
                        const body = {
                            id: channelId,
                            brandingSettings: {
                                channel: {},
                                image: {},
                            },
                        };
                        qs.part = 'brandingSettings';
                        if (updateFields.onBehalfOfContentOwner) {
                            qs.onBehalfOfContentOwner = updateFields.onBehalfOfContentOwner;
                        }
                        if (updateFields.brandingSettingsUi) {
                            const channelSettingsValues = updateFields.brandingSettingsUi
                                .channelSettingsValues;
                            const channelSettings = {};
                            if (channelSettingsValues === null || channelSettingsValues === void 0 ? void 0 : channelSettingsValues.channel) {
                                const channelSettingsOptions = channelSettingsValues.channel;
                                if (channelSettingsOptions.country) {
                                    channelSettings.country = channelSettingsOptions.country;
                                }
                                if (channelSettingsOptions.description) {
                                    channelSettings.description = channelSettingsOptions.description;
                                }
                                if (channelSettingsOptions.defaultLanguage) {
                                    channelSettings.defaultLanguage = channelSettingsOptions.defaultLanguage;
                                }
                                if (channelSettingsOptions.defaultTab) {
                                    channelSettings.defaultTab = channelSettingsOptions.defaultTab;
                                }
                                if (channelSettingsOptions.featuredChannelsTitle) {
                                    channelSettings.featuredChannelsTitle =
                                        channelSettingsOptions.featuredChannelsTitle;
                                }
                                if (channelSettingsOptions.featuredChannelsUrls) {
                                    channelSettings.featuredChannelsUrls =
                                        channelSettingsOptions.featuredChannelsUrls;
                                }
                                if (channelSettingsOptions.keywords) {
                                    channelSettings.keywords = channelSettingsOptions.keywords;
                                }
                                if (channelSettingsOptions.moderateComments) {
                                    channelSettings.moderateComments =
                                        channelSettingsOptions.moderateComments;
                                }
                                if (channelSettingsOptions.profileColor) {
                                    channelSettings.profileColor = channelSettingsOptions.profileColor;
                                }
                                if (channelSettingsOptions.profileColor) {
                                    channelSettings.profileColor = channelSettingsOptions.profileColor;
                                }
                                if (channelSettingsOptions.showRelatedChannels) {
                                    channelSettings.showRelatedChannels =
                                        channelSettingsOptions.showRelatedChannels;
                                }
                                if (channelSettingsOptions.showBrowseView) {
                                    channelSettings.showBrowseView = channelSettingsOptions.showBrowseView;
                                }
                                if (channelSettingsOptions.trackingAnalyticsAccountId) {
                                    channelSettings.trackingAnalyticsAccountId =
                                        channelSettingsOptions.trackingAnalyticsAccountId;
                                }
                                if (channelSettingsOptions.unsubscribedTrailer) {
                                    channelSettings.unsubscribedTrailer =
                                        channelSettingsOptions.unsubscribedTrailer;
                                }
                            }
                            const imageSettingsValues = updateFields.brandingSettingsUi
                                .imageSettingsValues;
                            const imageSettings = {};
                            if (imageSettingsValues === null || imageSettingsValues === void 0 ? void 0 : imageSettingsValues.image) {
                                const imageSettingsOptions = imageSettings.image;
                                if (imageSettingsOptions.bannerExternalUrl) {
                                    imageSettings.bannerExternalUrl =
                                        imageSettingsOptions.bannerExternalUrl;
                                }
                                if (imageSettingsOptions.trackingImageUrl) {
                                    imageSettings.trackingImageUrl = imageSettingsOptions.trackingImageUrl;
                                }
                                if (imageSettingsOptions.watchIconImageUrl) {
                                    imageSettings.watchIconImageUrl =
                                        imageSettingsOptions.watchIconImageUrl;
                                }
                            }
                            body.brandingSettings.channel = channelSettings;
                            body.brandingSettings.image = imageSettings;
                        }
                        responseData = await GenericFunctions_1.googleApiRequest.call(this, 'PUT', '/youtube/v3/channels', body, qs);
                    }
                    if (operation === 'uploadBanner') {
                        const channelId = this.getNodeParameter('channelId', i);
                        const binaryProperty = this.getNodeParameter('binaryProperty', i);
                        const binaryData = this.helpers.assertBinaryData(i, binaryProperty);
                        const body = await this.helpers.getBinaryDataBuffer(i, binaryProperty);
                        const requestOptions = {
                            headers: {
                                ...(binaryData.mimeType ? { 'Content-Type': binaryData.mimeType } : {}),
                            },
                            json: false,
                        };
                        const response = await GenericFunctions_1.googleApiRequest.call(this, 'POST', '/upload/youtube/v3/channelBanners/insert', body, qs, undefined, requestOptions);
                        const { url } = JSON.parse(response);
                        qs.part = 'brandingSettings';
                        responseData = await GenericFunctions_1.googleApiRequest.call(this, 'PUT', '/youtube/v3/channels', {
                            id: channelId,
                            brandingSettings: {
                                image: {
                                    bannerExternalUrl: url,
                                },
                            },
                        }, qs);
                    }
                }
                if (resource === 'playlist') {
                    if (operation === 'get') {
                        let part = this.getNodeParameter('part', i);
                        const playlistId = this.getNodeParameter('playlistId', i);
                        const options = this.getNodeParameter('options', i);
                        if (part.includes('*')) {
                            part = ['contentDetails', 'id', 'localizations', 'player', 'snippet', 'status'];
                        }
                        qs.part = part.join(',');
                        qs.id = playlistId;
                        Object.assign(qs, options);
                        responseData = await GenericFunctions_1.googleApiRequest.call(this, 'GET', '/youtube/v3/playlists', {}, qs);
                        responseData = responseData.items;
                    }
                    if (operation === 'getAll') {
                        const returnAll = this.getNodeParameter('returnAll', i);
                        let part = this.getNodeParameter('part', i);
                        const options = this.getNodeParameter('options', i);
                        const filters = this.getNodeParameter('filters', i);
                        if (part.includes('*')) {
                            part = ['contentDetails', 'id', 'localizations', 'player', 'snippet', 'status'];
                        }
                        qs.part = part.join(',');
                        Object.assign(qs, options, filters);
                        qs.mine = true;
                        if (qs.channelId || qs.id) {
                            delete qs.mine;
                        }
                        if (returnAll) {
                            responseData = await GenericFunctions_1.googleApiRequestAllItems.call(this, 'items', 'GET', '/youtube/v3/playlists', {}, qs);
                        }
                        else {
                            qs.maxResults = this.getNodeParameter('limit', i);
                            responseData = await GenericFunctions_1.googleApiRequest.call(this, 'GET', '/youtube/v3/playlists', {}, qs);
                            responseData = responseData.items;
                        }
                    }
                    if (operation === 'create') {
                        const title = this.getNodeParameter('title', i);
                        const options = this.getNodeParameter('options', i);
                        qs.part = 'snippet';
                        const body = {
                            snippet: {
                                title,
                            },
                        };
                        if (options.tags) {
                            body.snippet.tags = options.tags.split(',');
                        }
                        if (options.description) {
                            body.snippet.privacyStatus = options.privacyStatus;
                        }
                        if (options.defaultLanguage) {
                            body.snippet.defaultLanguage = options.defaultLanguage;
                        }
                        if (options.onBehalfOfContentOwner) {
                            qs.onBehalfOfContentOwner = options.onBehalfOfContentOwner;
                        }
                        if (options.onBehalfOfContentOwnerChannel) {
                            qs.onBehalfOfContentOwnerChannel = options.onBehalfOfContentOwnerChannel;
                        }
                        responseData = await GenericFunctions_1.googleApiRequest.call(this, 'POST', '/youtube/v3/playlists', body, qs);
                    }
                    if (operation === 'update') {
                        const playlistId = this.getNodeParameter('playlistId', i);
                        const title = this.getNodeParameter('title', i);
                        const updateFields = this.getNodeParameter('updateFields', i);
                        qs.part = 'snippet, status';
                        const body = {
                            id: playlistId,
                            snippet: {
                                title,
                            },
                            status: {},
                        };
                        if (updateFields.tags) {
                            body.snippet.tags = updateFields.tags.split(',');
                        }
                        if (updateFields.privacyStatus) {
                            body.status.privacyStatus = updateFields.privacyStatus;
                        }
                        if (updateFields.description) {
                            body.snippet.description = updateFields.description;
                        }
                        if (updateFields.defaultLanguage) {
                            body.snippet.defaultLanguage = updateFields.defaultLanguage;
                        }
                        if (updateFields.onBehalfOfContentOwner) {
                            qs.onBehalfOfContentOwner = updateFields.onBehalfOfContentOwner;
                        }
                        responseData = await GenericFunctions_1.googleApiRequest.call(this, 'PUT', '/youtube/v3/playlists', body, qs);
                    }
                    if (operation === 'delete') {
                        const playlistId = this.getNodeParameter('playlistId', i);
                        const options = this.getNodeParameter('options', i);
                        const body = {
                            id: playlistId,
                        };
                        if (options.onBehalfOfContentOwner) {
                            qs.onBehalfOfContentOwner = options.onBehalfOfContentOwner;
                        }
                        responseData = await GenericFunctions_1.googleApiRequest.call(this, 'DELETE', '/youtube/v3/playlists', body);
                        responseData = { success: true };
                    }
                }
                if (resource === 'playlistItem') {
                    if (operation === 'get') {
                        let part = this.getNodeParameter('part', i);
                        const playlistItemId = this.getNodeParameter('playlistItemId', i);
                        const options = this.getNodeParameter('options', i);
                        if (part.includes('*')) {
                            part = ['contentDetails', 'id', 'snippet', 'status'];
                        }
                        qs.part = part.join(',');
                        qs.id = playlistItemId;
                        Object.assign(qs, options);
                        responseData = await GenericFunctions_1.googleApiRequest.call(this, 'GET', '/youtube/v3/playlistItems', {}, qs);
                        responseData = responseData.items;
                    }
                    if (operation === 'getAll') {
                        const returnAll = this.getNodeParameter('returnAll', i);
                        let part = this.getNodeParameter('part', i);
                        const options = this.getNodeParameter('options', i);
                        const playlistId = this.getNodeParameter('playlistId', i);
                        if (part.includes('*')) {
                            part = ['contentDetails', 'id', 'snippet', 'status'];
                        }
                        qs.playlistId = playlistId;
                        qs.part = part.join(',');
                        Object.assign(qs, options);
                        if (returnAll) {
                            responseData = await GenericFunctions_1.googleApiRequestAllItems.call(this, 'items', 'GET', '/youtube/v3/playlistItems', {}, qs);
                        }
                        else {
                            qs.maxResults = this.getNodeParameter('limit', i);
                            responseData = await GenericFunctions_1.googleApiRequest.call(this, 'GET', '/youtube/v3/playlistItems', {}, qs);
                            responseData = responseData.items;
                        }
                    }
                    if (operation === 'add') {
                        const playlistId = this.getNodeParameter('playlistId', i);
                        const videoId = this.getNodeParameter('videoId', i);
                        const options = this.getNodeParameter('options', i);
                        qs.part = 'snippet, contentDetails';
                        const body = {
                            snippet: {
                                playlistId,
                                resourceId: {
                                    kind: 'youtube#video',
                                    videoId,
                                },
                            },
                            contentDetails: {},
                        };
                        if (options.position) {
                            body.snippet.position = options.position;
                        }
                        if (options.note) {
                            body.contentDetails.note = options.note;
                        }
                        if (options.startAt) {
                            body.contentDetails.startAt = options.startAt;
                        }
                        if (options.endAt) {
                            body.contentDetails.endAt = options.endAt;
                        }
                        if (options.onBehalfOfContentOwner) {
                            qs.onBehalfOfContentOwner = options.onBehalfOfContentOwner;
                        }
                        responseData = await GenericFunctions_1.googleApiRequest.call(this, 'POST', '/youtube/v3/playlistItems', body, qs);
                    }
                    if (operation === 'delete') {
                        const playlistItemId = this.getNodeParameter('playlistItemId', i);
                        const options = this.getNodeParameter('options', i);
                        const body = {
                            id: playlistItemId,
                        };
                        if (options.onBehalfOfContentOwner) {
                            qs.onBehalfOfContentOwner = options.onBehalfOfContentOwner;
                        }
                        responseData = await GenericFunctions_1.googleApiRequest.call(this, 'DELETE', '/youtube/v3/playlistItems', body);
                        responseData = { success: true };
                    }
                }
                if (resource === 'video') {
                    if (operation === 'getAll') {
                        const returnAll = this.getNodeParameter('returnAll', i);
                        const options = this.getNodeParameter('options', i);
                        const filters = this.getNodeParameter('filters', i);
                        qs.part = 'snippet';
                        qs.type = 'video';
                        qs.forMine = true;
                        Object.assign(qs, options, filters);
                        if (Object.keys(filters).length > 0) {
                            delete qs.forMine;
                        }
                        if (qs.relatedToVideoId && qs.forDeveloper !== undefined) {
                            throw new n8n_workflow_1.NodeOperationError(this.getNode(), "When using the parameter 'related to video' the parameter 'for developer' cannot be set", { itemIndex: i });
                        }
                        if (returnAll) {
                            responseData = await GenericFunctions_1.googleApiRequestAllItems.call(this, 'items', 'GET', '/youtube/v3/search', {}, qs);
                        }
                        else {
                            qs.maxResults = this.getNodeParameter('limit', i);
                            responseData = await GenericFunctions_1.googleApiRequest.call(this, 'GET', '/youtube/v3/search', {}, qs);
                            responseData = responseData.items;
                        }
                    }
                    if (operation === 'get') {
                        let part = this.getNodeParameter('part', i);
                        const videoId = this.getNodeParameter('videoId', i);
                        const options = this.getNodeParameter('options', i);
                        if (part.includes('*')) {
                            part = [
                                'contentDetails',
                                'id',
                                'liveStreamingDetails',
                                'localizations',
                                'player',
                                'recordingDetails',
                                'snippet',
                                'statistics',
                                'status',
                                'topicDetails',
                            ];
                        }
                        qs.part = part.join(',');
                        qs.id = videoId;
                        Object.assign(qs, options);
                        responseData = await GenericFunctions_1.googleApiRequest.call(this, 'GET', '/youtube/v3/videos', {}, qs);
                        responseData = responseData.items;
                    }
                    if (operation === 'upload') {
                        const title = this.getNodeParameter('title', i);
                        const categoryId = this.getNodeParameter('categoryId', i);
                        const options = this.getNodeParameter('options', i);
                        const binaryProperty = this.getNodeParameter('binaryProperty', i);
                        const binaryData = this.helpers.assertBinaryData(i, binaryProperty);
                        let mimeType;
                        let contentLength;
                        let fileContent;
                        if (binaryData.id) {
                            fileContent = this.helpers.getBinaryStream(binaryData.id, UPLOAD_CHUNK_SIZE);
                            const metadata = await this.helpers.getBinaryMetadata(binaryData.id);
                            contentLength = metadata.fileSize;
                            mimeType = (_a = metadata.mimeType) !== null && _a !== void 0 ? _a : binaryData.mimeType;
                        }
                        else {
                            fileContent = Buffer.from(binaryData.data, n8n_workflow_1.BINARY_ENCODING);
                            contentLength = fileContent.length;
                            mimeType = binaryData.mimeType;
                        }
                        const payload = {
                            snippet: {
                                title,
                                categoryId,
                                description: options.description,
                                tags: (_b = options.tags) === null || _b === void 0 ? void 0 : _b.split(','),
                                defaultLanguage: options.defaultLanguage,
                            },
                            status: {
                                privacyStatus: options.privacyStatus,
                                embeddable: options.embeddable,
                                publicStatsViewable: options.publicStatsViewable,
                                publishAt: options.publishAt,
                                selfDeclaredMadeForKids: options.selfDeclaredMadeForKids,
                                license: options.license,
                            },
                            recordingDetails: {
                                recordingDate: options.recordingDate,
                            },
                        };
                        const resumableUpload = await GenericFunctions_1.googleApiRequest.call(this, 'POST', '/upload/youtube/v3/videos', payload, {
                            uploadType: 'resumable',
                            part: 'snippet,status,recordingDetails',
                            notifySubscribers: (_c = options.notifySubscribers) !== null && _c !== void 0 ? _c : false,
                        }, undefined, {
                            headers: {
                                'X-Upload-Content-Length': contentLength,
                                'X-Upload-Content-Type': mimeType,
                            },
                            json: true,
                            resolveWithFullResponse: true,
                        });
                        const uploadUrl = resumableUpload.headers.location;
                        let uploadId;
                        let offset = 0;
                        for await (const chunk of fileContent) {
                            const nextOffset = offset + Number(chunk.length);
                            try {
                                const response = await this.helpers.httpRequest({
                                    method: 'PUT',
                                    url: uploadUrl,
                                    headers: {
                                        'Content-Length': chunk.length,
                                        'Content-Range': `bytes ${offset}-${nextOffset - 1}/${contentLength}`,
                                    },
                                    body: chunk,
                                });
                                uploadId = response.id;
                            }
                            catch (error) {
                                if (((_d = error.response) === null || _d === void 0 ? void 0 : _d.status) !== 308)
                                    throw error;
                            }
                            offset = nextOffset;
                        }
                        responseData = { uploadId, ...resumableUpload.body };
                    }
                    if (operation === 'update') {
                        const id = this.getNodeParameter('videoId', i);
                        const title = this.getNodeParameter('title', i);
                        const categoryId = this.getNodeParameter('categoryId', i);
                        const updateFields = this.getNodeParameter('updateFields', i);
                        qs.part = 'snippet, status, recordingDetails';
                        const body = {
                            id,
                            snippet: {
                                title,
                                categoryId,
                            },
                            status: {},
                            recordingDetails: {},
                        };
                        if (updateFields.description) {
                            body.snippet.description = updateFields.description;
                        }
                        if (updateFields.privacyStatus) {
                            body.status.privacyStatus = updateFields.privacyStatus;
                        }
                        if (updateFields.tags) {
                            body.snippet.tags = updateFields.tags.split(',');
                        }
                        if (updateFields.embeddable) {
                            body.status.embeddable = updateFields.embeddable;
                        }
                        if (updateFields.publicStatsViewable) {
                            body.status.publicStatsViewable = updateFields.publicStatsViewable;
                        }
                        if (updateFields.publishAt) {
                            body.status.publishAt = updateFields.publishAt;
                        }
                        if (updateFields.selfDeclaredMadeForKids) {
                            body.status.selfDeclaredMadeForKids = updateFields.selfDeclaredMadeForKids;
                        }
                        if (updateFields.recordingDate) {
                            body.recordingDetails.recordingDate = updateFields.recordingDate;
                        }
                        if (updateFields.license) {
                            body.status.license = updateFields.license;
                        }
                        if (updateFields.defaultLanguage) {
                            body.snippet.defaultLanguage = updateFields.defaultLanguage;
                        }
                        responseData = await GenericFunctions_1.googleApiRequest.call(this, 'PUT', '/youtube/v3/videos', body, qs);
                    }
                    if (operation === 'delete') {
                        const videoId = this.getNodeParameter('videoId', i);
                        const options = this.getNodeParameter('options', i);
                        const body = {
                            id: videoId,
                        };
                        if (options.onBehalfOfContentOwner) {
                            qs.onBehalfOfContentOwner = options.onBehalfOfContentOwner;
                        }
                        responseData = await GenericFunctions_1.googleApiRequest.call(this, 'DELETE', '/youtube/v3/videos', body);
                        responseData = { success: true };
                    }
                    if (operation === 'rate') {
                        const videoId = this.getNodeParameter('videoId', i);
                        const rating = this.getNodeParameter('rating', i);
                        const body = {
                            id: videoId,
                            rating,
                        };
                        responseData = await GenericFunctions_1.googleApiRequest.call(this, 'POST', '/youtube/v3/videos/rate', body);
                        responseData = { success: true };
                    }
                }
                if (resource === 'videoCategory') {
                    if (operation === 'getAll') {
                        const returnAll = this.getNodeParameter('returnAll', i);
                        const regionCode = this.getNodeParameter('regionCode', i);
                        qs.regionCode = regionCode;
                        qs.part = 'snippet';
                        responseData = await GenericFunctions_1.googleApiRequest.call(this, 'GET', '/youtube/v3/videoCategories', {}, qs);
                        responseData = responseData.items;
                        if (!returnAll) {
                            const limit = this.getNodeParameter('limit', i);
                            responseData = responseData.splice(0, limit);
                        }
                    }
                }
            }
            catch (error) {
                if (this.continueOnFail()) {
                    const executionErrorData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray({ error: error.message }), { itemData: { item: i } });
                    returnData.push(...executionErrorData);
                    continue;
                }
                throw error;
            }
            const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
            returnData.push(...executionData);
        }
        return this.prepareOutputData(returnData);
    }
}
exports.YouTube = YouTube;
