// n8n/dist/controllers/ldap.controller.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LdapController = void 0;
const lodash_pick_1 = __importDefault(require("lodash.pick"));
const decorators_1 = require("../decorators");
const helpers_1 = require("../Ldap/helpers");
const LdapService_ee_1 = require("../Ldap/LdapService.ee");
const LdapSync_ee_1 = require("../Ldap/LdapSync.ee");
const ResponseHelper_1 = require("../ResponseHelper");
const constants_1 = require("../Ldap/constants");
const InternalHooks_1 = require("../InternalHooks");
let LdapController = class LdapController {
    constructor(ldapService, ldapSync, internalHooks) {
        this.ldapService = ldapService;
        this.ldapSync = ldapSync;
        this.internalHooks = internalHooks;
    }
    async getConfig() {
        return (0, helpers_1.getLdapConfig)();
    }
    async testConnection() {
        try {
            await this.ldapService.testConnection();
        }
        catch (error) {
            throw new ResponseHelper_1.BadRequestError(error.message);
        }
    }
    async updateConfig(req) {
        try {
            await (0, helpers_1.updateLdapConfig)(req.body);
        }
        catch (error) {
            throw new ResponseHelper_1.BadRequestError(error.message);
        }
        const data = await (0, helpers_1.getLdapConfig)();
        void this.internalHooks.onUserUpdatedLdapSettings({
            user_id: req.user.id,
            ...(0, lodash_pick_1.default)(data, constants_1.NON_SENSIBLE_LDAP_CONFIG_PROPERTIES),
        });
        return data;
    }
    async getLdapSync(req) {
        const { page = '0', perPage = '20' } = req.query;
        return (0, helpers_1.getLdapSynchronizations)(parseInt(page, 10), parseInt(perPage, 10));
    }
    async syncLdap(req) {
        try {
            await this.ldapSync.run(req.body.type);
        }
        catch (error) {
            throw new ResponseHelper_1.BadRequestError(error.message);
        }
    }
};
__decorate([
    (0, decorators_1.Get)('/config'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], LdapController.prototype, "getConfig", null);
__decorate([
    (0, decorators_1.Post)('/test-connection'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], LdapController.prototype, "testConnection", null);
__decorate([
    (0, decorators_1.Put)('/config'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], LdapController.prototype, "updateConfig", null);
__decorate([
    (0, decorators_1.Get)('/sync'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], LdapController.prototype, "getLdapSync", null);
__decorate([
    (0, decorators_1.Post)('/sync'),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], LdapController.prototype, "syncLdap", null);
LdapController = __decorate([
    (0, decorators_1.Authorized)(['global', 'owner']),
    (0, decorators_1.RestController)('/ldap'),
    __metadata("design:paramtypes", [LdapService_ee_1.LdapService,
        LdapSync_ee_1.LdapSync,
        InternalHooks_1.InternalHooks])
], LdapController);
exports.LdapController = LdapController;
