// n8n-nodes-base/dist/nodes/Microsoft/Excel/v2/actions/versionDescription.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.versionDescription = void 0;
const table = __importStar(require("./table/Table.resource"));
const workbook = __importStar(require("./workbook/Workbook.resource"));
const worksheet = __importStar(require("./worksheet/Worksheet.resource"));
exports.versionDescription = {
    displayName: 'Microsoft Excel 365',
    name: 'microsoftExcel',
    icon: 'file:excel.svg',
    group: ['input'],
    version: 2,
    subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
    description: 'Consume Microsoft Excel API',
    defaults: {
        name: 'Microsoft Excel 365',
    },
    inputs: ['main'],
    outputs: ['main'],
    credentials: [
        {
            name: 'microsoftExcelOAuth2Api',
            required: true,
        },
    ],
    properties: [
        {
            displayName: 'This node connects to the Microsoft 365 cloud platform. Use the \'Spreadsheet File\' node to directly manipulate spreadsheet files (.xls, .csv, etc). <a href="/templates/890" target="_blank">More info</a>.',
            name: 'notice',
            type: 'notice',
            default: '',
        },
        {
            displayName: 'Resource',
            name: 'resource',
            type: 'options',
            noDataExpression: true,
            options: [
                {
                    name: 'Table',
                    value: 'table',
                    description: 'Represents an Excel table',
                },
                {
                    name: 'Workbook',
                    value: 'workbook',
                    description: 'A workbook is the top level object which contains one or more worksheets',
                },
                {
                    name: 'Sheet',
                    value: 'worksheet',
                    description: 'A sheet is a grid of cells which can contain data, tables, charts, etc',
                },
            ],
            default: 'workbook',
        },
        ...table.description,
        ...workbook.description,
        ...worksheet.description,
    ],
};
