// n8n-nodes-base/dist/nodes/HighLevel/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTimezones = exports.getUsers = exports.getPipelineStages = exports.highLevelApiPagination = exports.splitTagsPreSendAction = exports.taskUpdatePreSendAction = exports.opportunityUpdatePreSendAction = exports.highLevelApiRequest = exports.dateTimeToEpochPreSendAction = exports.validEmailAndPhonePreSendAction = exports.contactIdentifierPreSendAction = exports.dueDatePreSendAction = exports.taskPostReceiceAction = exports.isPhoneValid = exports.isEmailValid = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const luxon_1 = require("luxon");
const VALID_EMAIL_REGEX = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
const VALID_PHONE_REGEX = /((?:\+|00)[17](?: |\-)?|(?:\+|00)[1-9]\d{0,2}(?: |\-)?|(?:\+|00)1\-\d{3}(?: |\-)?)?(0\d|\([0-9]{3}\)|[1-9]{0,3})(?:((?: |\-)[0-9]{2}){4}|((?:[0-9]{2}){4})|((?: |\-)[0-9]{3}(?: |\-)[0-9]{4})|([0-9]{7}))/;
function isEmailValid(email) {
    return VALID_EMAIL_REGEX.test(String(email).toLowerCase());
}
exports.isEmailValid = isEmailValid;
function isPhoneValid(phone) {
    return VALID_PHONE_REGEX.test(String(phone));
}
exports.isPhoneValid = isPhoneValid;
function dateToIsoSupressMillis(dateTime) {
    const options = { suppressMilliseconds: true };
    return luxon_1.DateTime.fromISO(dateTime).toISO(options);
}
async function taskPostReceiceAction(items, _response) {
    const contactId = this.getNodeParameter('contactId');
    items.forEach((item) => (item.json.contactId = contactId));
    return items;
}
exports.taskPostReceiceAction = taskPostReceiceAction;
async function dueDatePreSendAction(requestOptions) {
    let dueDateParam = this.getNodeParameter('dueDate', null);
    if (!dueDateParam) {
        const fields = this.getNodeParameter('updateFields');
        dueDateParam = fields.dueDate;
    }
    if (!dueDateParam) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), {}, { message: 'dueDate is required', description: 'dueDate is required' });
    }
    const dueDate = dateToIsoSupressMillis(dueDateParam);
    requestOptions.body = (requestOptions.body || {});
    Object.assign(requestOptions.body, { dueDate });
    return requestOptions;
}
exports.dueDatePreSendAction = dueDatePreSendAction;
async function contactIdentifierPreSendAction(requestOptions) {
    requestOptions.body = (requestOptions.body || {});
    let identifier = this.getNodeParameter('contactIdentifier', null);
    if (!identifier) {
        const fields = this.getNodeParameter('updateFields');
        identifier = fields.contactIdentifier;
    }
    if (isEmailValid(identifier)) {
        Object.assign(requestOptions.body, { email: identifier });
    }
    else if (isPhoneValid(identifier)) {
        Object.assign(requestOptions.body, { phone: identifier });
    }
    else {
        Object.assign(requestOptions.body, { contactId: identifier });
    }
    return requestOptions;
}
exports.contactIdentifierPreSendAction = contactIdentifierPreSendAction;
async function validEmailAndPhonePreSendAction(requestOptions) {
    const body = (requestOptions.body || {});
    if (body.email && !isEmailValid(body.email)) {
        const message = `email "${body.email}" has invalid format`;
        throw new n8n_workflow_1.NodeApiError(this.getNode(), {}, { message, description: message });
    }
    if (body.phone && !isPhoneValid(body.phone)) {
        const message = `phone "${body.phone}" has invalid format`;
        throw new n8n_workflow_1.NodeApiError(this.getNode(), {}, { message, description: message });
    }
    return requestOptions;
}
exports.validEmailAndPhonePreSendAction = validEmailAndPhonePreSendAction;
async function dateTimeToEpochPreSendAction(requestOptions) {
    const qs = (requestOptions.qs || {});
    const toEpoch = (dt) => new Date(dt).getTime();
    if (qs.startDate)
        qs.startDate = toEpoch(qs.startDate);
    if (qs.endDate)
        qs.endDate = toEpoch(qs.endDate);
    return requestOptions;
}
exports.dateTimeToEpochPreSendAction = dateTimeToEpochPreSendAction;
async function highLevelApiRequest(method, resource, body = {}, qs = {}, uri, option = {}) {
    let options = {
        method,
        body,
        qs,
        uri: uri || `https://rest.gohighlevel.com/v1${resource}`,
        json: true,
    };
    if (!Object.keys(body).length) {
        delete options.body;
    }
    if (!Object.keys(qs).length) {
        delete options.qs;
    }
    options = Object.assign({}, options, option);
    return this.helpers.requestWithAuthentication.call(this, 'highLevelApi', options);
}
exports.highLevelApiRequest = highLevelApiRequest;
async function opportunityUpdatePreSendAction(requestOptions) {
    const body = (requestOptions.body || {});
    if (!body.status || !body.title) {
        const pipelineId = this.getNodeParameter('pipelineId');
        const opportunityId = this.getNodeParameter('opportunityId');
        const resource = `/pipelines/${pipelineId}/opportunities/${opportunityId}`;
        const responseData = await highLevelApiRequest.call(this, 'GET', resource);
        body.status = body.status || responseData.status;
        body.title = body.title || responseData.name;
        requestOptions.body = body;
    }
    return requestOptions;
}
exports.opportunityUpdatePreSendAction = opportunityUpdatePreSendAction;
async function taskUpdatePreSendAction(requestOptions) {
    const body = (requestOptions.body || {});
    if (!body.title || !body.dueDate) {
        const contactId = this.getNodeParameter('contactId');
        const taskId = this.getNodeParameter('taskId');
        const resource = `/contacts/${contactId}/tasks/${taskId}`;
        const responseData = await highLevelApiRequest.call(this, 'GET', resource);
        body.title = body.title || responseData.title;
        body.dueDate = body.dueDate || dateToIsoSupressMillis(responseData.dueDate);
        requestOptions.body = body;
    }
    return requestOptions;
}
exports.taskUpdatePreSendAction = taskUpdatePreSendAction;
async function splitTagsPreSendAction(requestOptions) {
    const body = (requestOptions.body || {});
    if (body.tags) {
        if (Array.isArray(body.tags))
            return requestOptions;
        body.tags = body.tags.split(',').map((tag) => tag.trim());
    }
    return requestOptions;
}
exports.splitTagsPreSendAction = splitTagsPreSendAction;
async function highLevelApiPagination(requestData) {
    const responseData = [];
    const resource = this.getNodeParameter('resource');
    const returnAll = this.getNodeParameter('returnAll', false);
    const resourceMapping = {
        contact: 'contacts',
        opportunity: 'opportunities',
    };
    const rootProperty = resourceMapping[resource];
    requestData.options.qs = requestData.options.qs || {};
    if (returnAll)
        requestData.options.qs.limit = 100;
    let responseTotal = 0;
    do {
        const pageResponseData = await this.makeRoutingRequest(requestData);
        const items = pageResponseData[0].json[rootProperty];
        items.forEach((item) => responseData.push({ json: item }));
        const meta = pageResponseData[0].json.meta;
        const startAfterId = meta.startAfterId;
        const startAfter = meta.startAfter;
        requestData.options.qs = { startAfterId, startAfter };
        responseTotal = meta.total || 0;
    } while (returnAll && responseTotal > responseData.length);
    return responseData;
}
exports.highLevelApiPagination = highLevelApiPagination;
async function getPipelineStages() {
    const pipelineId = this.getCurrentNodeParameter('pipelineId');
    const responseData = await highLevelApiRequest.call(this, 'GET', '/pipelines');
    const pipelines = responseData.pipelines;
    const pipeline = pipelines.find((p) => p.id === pipelineId);
    if (pipeline) {
        const options = pipeline.stages.map((stage) => {
            const name = stage.name;
            const value = stage.id;
            return { name, value };
        });
        return options;
    }
    return [];
}
exports.getPipelineStages = getPipelineStages;
async function getUsers() {
    const responseData = await highLevelApiRequest.call(this, 'GET', '/users');
    const users = responseData.users;
    const options = users.map((user) => {
        const name = user.name;
        const value = user.id;
        return { name, value };
    });
    return options;
}
exports.getUsers = getUsers;
async function getTimezones() {
    const responseData = await highLevelApiRequest.call(this, 'GET', '/timezones');
    const timezones = responseData.timezones;
    return timezones.map((zone) => ({
        name: zone,
        value: zone,
    }));
}
exports.getTimezones = getTimezones;
