// n8n-nodes-base/dist/nodes/HighLevel/HighLevel.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HighLevel = void 0;
const ContactDescription_1 = require("./description/ContactDescription");
const OpportunityDescription_1 = require("./description/OpportunityDescription");
const TaskDescription_1 = require("./description/TaskDescription");
const GenericFunctions_1 = require("./GenericFunctions");
const ressources = [
    {
        displayName: 'Resource',
        name: 'resource',
        type: 'options',
        noDataExpression: true,
        options: [
            {
                name: 'Contact',
                value: 'contact',
            },
            {
                name: 'Opportunity',
                value: 'opportunity',
            },
            {
                name: 'Task',
                value: 'task',
            },
        ],
        default: 'contact',
        required: true,
    },
];
class HighLevel {
    constructor() {
        this.description = {
            displayName: 'HighLevel',
            name: 'highLevel',
            icon: 'file:highLevel.svg',
            group: ['transform'],
            version: 1,
            description: 'Consume HighLevel API',
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            defaults: {
                name: 'HighLevel',
            },
            inputs: ['main'],
            outputs: ['main'],
            credentials: [
                {
                    name: 'highLevelApi',
                    required: true,
                },
            ],
            requestDefaults: {
                baseURL: 'https://rest.gohighlevel.com/v1',
                headers: {
                    Accept: 'application/json',
                    'Content-Type': 'application/json',
                },
            },
            requestOperations: {
                pagination: GenericFunctions_1.highLevelApiPagination,
            },
            properties: [
                ...ressources,
                ...ContactDescription_1.contactOperations,
                ...ContactDescription_1.contactNotes,
                ...ContactDescription_1.contactFields,
                ...OpportunityDescription_1.opportunityOperations,
                ...OpportunityDescription_1.opportunityFields,
                ...TaskDescription_1.taskOperations,
                ...TaskDescription_1.taskFields,
            ],
        };
        this.methods = {
            loadOptions: {
                getPipelineStages: GenericFunctions_1.getPipelineStages,
                getUsers: GenericFunctions_1.getUsers,
                getTimezones: GenericFunctions_1.getTimezones,
            },
        };
    }
}
exports.HighLevel = HighLevel;
