// n8n-nodes-base/dist/nodes/Notion/SearchFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDatabases = void 0;
const GenericFunctions_1 = require("./GenericFunctions");
async function getDatabases(filter) {
    var _a;
    const returnData = [];
    const body = {
        page_size: 100,
        query: filter,
        filter: { property: 'object', value: 'database' },
    };
    const databases = await GenericFunctions_1.notionApiRequestAllItems.call(this, 'results', 'POST', '/search', body);
    for (const database of databases) {
        returnData.push({
            name: ((_a = database.title[0]) === null || _a === void 0 ? void 0 : _a.plain_text) || database.id,
            value: database.id,
            url: database.url,
        });
    }
    returnData.sort((a, b) => {
        if (a.name.toLocaleLowerCase() < b.name.toLocaleLowerCase()) {
            return -1;
        }
        if (a.name.toLocaleLowerCase() > b.name.toLocaleLowerCase()) {
            return 1;
        }
        return 0;
    });
    return { results: returnData };
}
exports.getDatabases = getDatabases;
