// n8n-nodes-base/dist/nodes/BambooHr/v1/actions/employeeDocument/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.descriptions = exports.upload = exports.update = exports.getAll = exports.download = exports.del = void 0;
const del = __importStar(require("./del"));
exports.del = del;
const download = __importStar(require("./download"));
exports.download = download;
const getAll = __importStar(require("./getAll"));
exports.getAll = getAll;
const update = __importStar(require("./update"));
exports.update = update;
const upload = __importStar(require("./upload"));
exports.upload = upload;
exports.descriptions = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                resource: ['employeeDocument'],
            },
        },
        options: [
            {
                name: 'Delete',
                value: 'delete',
                description: 'Delete an employee document',
                action: 'Delete an employee document',
            },
            {
                name: 'Download',
                value: 'download',
                description: 'Download an employee document',
                action: 'Download an employee document',
            },
            {
                name: 'Get Many',
                value: 'getAll',
                description: 'Get many employee documents',
                action: 'Get many employee documents',
            },
            {
                name: 'Update',
                value: 'update',
                description: 'Update an employee document',
                action: 'Update an employee document',
            },
            {
                name: 'Upload',
                value: 'upload',
                description: 'Upload an employee document',
                action: 'Upload an employee document',
            },
        ],
        default: 'delete',
    },
    ...del.description,
    ...download.description,
    ...getAll.description,
    ...update.description,
    ...upload.description,
];
