// n8n-nodes-base/dist/nodes/Copper/descriptions/PersonDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.personFields = exports.personOperations = void 0;
const sharedFields_1 = require("../utils/sharedFields");
exports.personOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                resource: ['person'],
            },
        },
        options: [
            {
                name: 'Create',
                value: 'create',
                action: 'Create a person',
            },
            {
                name: 'Delete',
                value: 'delete',
                action: 'Delete a person',
            },
            {
                name: 'Get',
                value: 'get',
                action: 'Get a person',
            },
            {
                name: 'Get Many',
                value: 'getAll',
                action: 'Get many people',
            },
            {
                name: 'Update',
                value: 'update',
                action: 'Update a person',
            },
        ],
        default: 'create',
    },
];
exports.personFields = [
    {
        displayName: 'Name',
        name: 'name',
        description: 'Name of the person to create',
        type: 'string',
        required: true,
        default: '',
        displayOptions: {
            show: {
                resource: ['person'],
                operation: ['create'],
            },
        },
    },
    {
        displayName: 'Additional Fields',
        name: 'additionalFields',
        type: 'collection',
        placeholder: 'Add Field',
        default: {},
        displayOptions: {
            show: {
                resource: ['person'],
                operation: ['create'],
            },
        },
        options: [
            sharedFields_1.addressFixedCollection,
            {
                displayName: 'Details',
                name: 'details',
                type: 'string',
                default: '',
                description: 'Description to set for the person',
            },
            {
                displayName: 'Email Domain',
                name: 'email_domain',
                type: 'string',
                default: '',
            },
            sharedFields_1.emailsFixedCollection,
            sharedFields_1.phoneNumbersFixedCollection,
        ],
    },
    {
        displayName: 'Person ID',
        name: 'personId',
        description: 'ID of the person to delete',
        type: 'string',
        required: true,
        default: '',
        displayOptions: {
            show: {
                resource: ['person'],
                operation: ['delete'],
            },
        },
    },
    {
        displayName: 'Person ID',
        name: 'personId',
        description: 'ID of the person to retrieve',
        type: 'string',
        required: true,
        default: '',
        displayOptions: {
            show: {
                resource: ['person'],
                operation: ['get'],
            },
        },
    },
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        default: false,
        description: 'Whether to return all results or only up to a given limit',
        displayOptions: {
            show: {
                resource: ['person'],
                operation: ['getAll'],
            },
        },
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        default: 5,
        description: 'Max number of results to return',
        typeOptions: {
            minValue: 1,
            maxValue: 1000,
        },
        displayOptions: {
            show: {
                resource: ['person'],
                operation: ['getAll'],
                returnAll: [false],
            },
        },
    },
    {
        displayName: 'Filters',
        name: 'filterFields',
        type: 'collection',
        placeholder: 'Add Field',
        default: {},
        displayOptions: {
            show: {
                resource: ['person'],
                operation: ['getAll'],
            },
        },
        options: [
            {
                displayName: 'Name',
                name: 'name',
                type: 'string',
                default: '',
                description: 'Name of the person to filter by',
            },
        ],
    },
    {
        displayName: 'Person ID',
        name: 'personId',
        description: 'ID of the person to update',
        type: 'string',
        required: true,
        default: '',
        displayOptions: {
            show: {
                resource: ['person'],
                operation: ['update'],
            },
        },
    },
    {
        displayName: 'Update Fields',
        name: 'updateFields',
        type: 'collection',
        placeholder: 'Add Field',
        default: {},
        displayOptions: {
            show: {
                resource: ['person'],
                operation: ['update'],
            },
        },
        options: [
            sharedFields_1.addressFixedCollection,
            {
                displayName: 'Details',
                name: 'details',
                type: 'string',
                default: '',
                description: 'Description to set for the person',
            },
            {
                displayName: 'Email Domain',
                name: 'email_domain',
                type: 'string',
                default: '',
            },
            sharedFields_1.emailsFixedCollection,
            {
                displayName: 'Name',
                name: 'name',
                type: 'string',
                default: '',
                description: 'Name to set for the person',
            },
            sharedFields_1.phoneNumbersFixedCollection,
        ],
    },
];
