// n8n-nodes-base/dist/nodes/Clockify/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.clockifyApiRequestAllItems = exports.clockifyApiRequest = void 0;
async function clockifyApiRequest(method, resource, body = {}, qs = {}, uri, _option = {}) {
    const BASE_URL = 'https://api.clockify.me/api/v1';
    const options = {
        headers: {
            'Content-Type': 'application/json',
        },
        method,
        qs,
        body,
        uri: `${BASE_URL}/${resource}`,
        json: true,
        useQuerystring: true,
    };
    return this.helpers.requestWithAuthentication.call(this, 'clockifyApi', options);
}
exports.clockifyApiRequest = clockifyApiRequest;
async function clockifyApiRequestAllItems(method, endpoint, body = {}, query = {}) {
    const returnData = [];
    let responseData;
    query['page-size'] = 50;
    query.page = 1;
    do {
        responseData = await clockifyApiRequest.call(this, method, endpoint, body, query);
        returnData.push.apply(returnData, responseData);
        const limit = query.limit;
        if (limit && returnData.length >= limit) {
            return returnData;
        }
        query.page++;
    } while (responseData.length !== 0);
    return returnData;
}
exports.clockifyApiRequestAllItems = clockifyApiRequestAllItems;
