// n8n-nodes-base/dist/nodes/Clockify/WorkpaceInterfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var AdminOnlyPagesEnum;
(function (AdminOnlyPagesEnum) {
    AdminOnlyPagesEnum["PROJECT"] = "PROJECT";
    AdminOnlyPagesEnum["TEAM"] = "TEAM";
    AdminOnlyPagesEnum["REPORTS"] = "REPORTS";
})(AdminOnlyPagesEnum || (AdminOnlyPagesEnum = {}));
var DaysOfWeekEnum;
(function (DaysOfWeekEnum) {
    DaysOfWeekEnum["MONDAY"] = "MONDAY";
    DaysOfWeekEnum["TUESDAY"] = "TUESDAY";
    DaysOfWeekEnum["WEDNESDAY"] = "WEDNESDAY";
    DaysOfWeekEnum["THURSDAY"] = "THURSDAY";
    DaysOfWeekEnum["FRIDAY"] = "FRIDAY";
    DaysOfWeekEnum["SATURDAY"] = "SATURDAY";
    DaysOfWeekEnum["SUNDAY"] = "SUNDAY";
})(DaysOfWeekEnum || (DaysOfWeekEnum = {}));
var DatePeriodEnum;
(function (DatePeriodEnum) {
    DatePeriodEnum["DAYS"] = "DAYS";
    DatePeriodEnum["WEEKS"] = "WEEKS";
    DatePeriodEnum["MONTHS"] = "MONTHS";
})(DatePeriodEnum || (DatePeriodEnum = {}));
var AutomaticLockTypeEnum;
(function (AutomaticLockTypeEnum) {
    AutomaticLockTypeEnum["WEEKLY"] = "WEEKLY";
    AutomaticLockTypeEnum["MONTHLY"] = "MONTHLY";
    AutomaticLockTypeEnum["OLDER_THAN"] = "OLDER_THAN";
})(AutomaticLockTypeEnum || (AutomaticLockTypeEnum = {}));
