// n8n/dist/CredentialsOverwrites.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CredentialsOverwrites = void 0;
const config_1 = __importDefault(require("./config"));
const n8n_workflow_1 = require("n8n-workflow");
class CredentialsOverwritesClass {
    constructor(credentialTypes) {
        this.credentialTypes = credentialTypes;
        this.overwriteData = {};
        this.resolvedTypes = [];
        const data = config_1.default.getEnv('credentials.overwrite.data');
        const overwriteData = (0, n8n_workflow_1.jsonParse)(data, {
            errorMessage: 'The credentials-overwrite is not valid JSON.',
        });
        this.setData(overwriteData);
    }
    setData(overwriteData) {
        this.resolvedTypes.length = 0;
        this.overwriteData = overwriteData;
        for (const type in overwriteData) {
            const overwrites = this.getOverwrites(type);
            if (overwrites && Object.keys(overwrites).length) {
                this.overwriteData[type] = overwrites;
            }
        }
    }
    applyOverwrite(type, data) {
        const overwrites = this.get(type);
        if (overwrites === undefined) {
            return data;
        }
        const returnData = (0, n8n_workflow_1.deepCopy)(data);
        for (const key of Object.keys(overwrites)) {
            if ([null, undefined, ''].includes(returnData[key])) {
                returnData[key] = overwrites[key];
            }
        }
        return returnData;
    }
    getOverwrites(type) {
        if (this.resolvedTypes.includes(type)) {
            return this.overwriteData[type];
        }
        if (!this.credentialTypes.recognizes(type)) {
            n8n_workflow_1.LoggerProxy.warn(`Unknown credential type ${type} in Credential overwrites`);
            return;
        }
        const credentialTypeData = this.credentialTypes.getByName(type);
        if (credentialTypeData.extends === undefined) {
            this.resolvedTypes.push(type);
            return this.overwriteData[type];
        }
        const overwrites = {};
        for (const credentialsTypeName of credentialTypeData.extends) {
            Object.assign(overwrites, this.getOverwrites(credentialsTypeName));
        }
        if (this.overwriteData[type] !== undefined) {
            Object.assign(overwrites, this.overwriteData[type]);
        }
        this.resolvedTypes.push(type);
        return overwrites;
    }
    get(name) {
        const parentTypes = this.credentialTypes.getParentTypes(name);
        return [name, ...parentTypes]
            .reverse()
            .map((type) => this.overwriteData[type])
            .filter((type) => !!type)
            .reduce((acc, current) => Object.assign(acc, current), {});
    }
    getAll() {
        return this.overwriteData;
    }
}
let credentialsOverwritesInstance;
function CredentialsOverwrites(credentialTypes) {
    if (!credentialsOverwritesInstance) {
        if (credentialTypes) {
            credentialsOverwritesInstance = new CredentialsOverwritesClass(credentialTypes);
        }
        else {
            throw new Error('CredentialsOverwrites not initialized yet');
        }
    }
    return credentialsOverwritesInstance;
}
exports.CredentialsOverwrites = CredentialsOverwrites;
