// n8n/dist/Queue.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Queue = void 0;
const typedi_1 = require("typedi");
const config_1 = __importDefault(require("./config"));
const ActiveExecutions_1 = require("./ActiveExecutions");
const WebhookHelpers = __importStar(require("./WebhookHelpers"));
let Queue = class Queue {
    constructor(activeExecutions) {
        this.activeExecutions = activeExecutions;
    }
    async init() {
        const prefix = config_1.default.getEnv('queue.bull.prefix');
        const redisOptions = config_1.default.getEnv('queue.bull.redis');
        const { default: Bull } = await Promise.resolve().then(() => __importStar(require('bull')));
        this.jobQueue = new Bull('jobs', { prefix, redis: redisOptions, enableReadyCheck: false });
        this.jobQueue.on('global:progress', (jobId, progress) => {
            this.activeExecutions.resolveResponsePromise(progress.executionId, WebhookHelpers.decodeWebhookResponse(progress.response));
        });
    }
    async add(jobData, jobOptions) {
        return this.jobQueue.add(jobData, jobOptions);
    }
    async getJob(jobId) {
        return this.jobQueue.getJob(jobId);
    }
    async getJobs(jobTypes) {
        return this.jobQueue.getJobs(jobTypes);
    }
    getBullObjectInstance() {
        return this.jobQueue;
    }
    async stopJob(job) {
        if (await job.isActive()) {
            await job.progress(-1);
            return true;
        }
        try {
            await job.remove();
            return true;
        }
        catch (e) {
            await job.progress(-1);
        }
        return false;
    }
};
Queue = __decorate([
    (0, typedi_1.Service)(),
    __metadata("design:paramtypes", [ActiveExecutions_1.ActiveExecutions])
], Queue);
exports.Queue = Queue;
