// n8n-nodes-base/dist/nodes/MailerLite/MailerLiteTrigger.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MailerLiteTrigger = void 0;
const GenericFunctions_1 = require("./GenericFunctions");
class MailerLiteTrigger {
    constructor() {
        this.description = {
            displayName: 'MailerLite Trigger',
            name: 'mailerLiteTrigger',
            icon: 'file:mailerLite.png',
            group: ['trigger'],
            version: 1,
            description: 'Starts the workflow when MailerLite events occur',
            defaults: {
                name: 'MailerLite Trigger',
            },
            inputs: [],
            outputs: ['main'],
            credentials: [
                {
                    name: 'mailerLiteApi',
                    required: true,
                },
            ],
            webhooks: [
                {
                    name: 'default',
                    httpMethod: 'POST',
                    responseMode: 'onReceived',
                    path: 'webhook',
                },
            ],
            properties: [
                {
                    displayName: 'Event',
                    name: 'event',
                    type: 'options',
                    options: [
                        {
                            name: 'Campaign Sent',
                            value: 'campaign.sent',
                            description: 'Fired when campaign is sent',
                        },
                        {
                            name: 'Subscriber Added Throught Webform',
                            value: 'subscriber.added_through_webform',
                            description: 'Fired when a subscriber is added though a form',
                        },
                        {
                            name: 'Subscriber Added to Group',
                            value: 'subscriber.add_to_group',
                            description: 'Fired when a subscriber is added to a group',
                        },
                        {
                            name: 'Subscriber Autonomation Completed',
                            value: 'subscriber.automation_complete',
                            description: 'Fired when subscriber finishes automation',
                        },
                        {
                            name: 'Subscriber Autonomation Triggered',
                            value: 'subscriber.automation_triggered',
                            description: 'Fired when subscriber starts automation',
                        },
                        {
                            name: 'Subscriber Bounced',
                            value: 'subscriber.bounced',
                            description: 'Fired when an email address bounces',
                        },
                        {
                            name: 'Subscriber Complained',
                            value: 'subscriber.complaint',
                            description: 'Fired when subscriber marks a campaign as a spam',
                        },
                        {
                            name: 'Subscriber Created',
                            value: 'subscriber.create',
                            description: 'Fired when a new subscriber is added to an account',
                        },
                        {
                            name: 'Subscriber Removed From Group',
                            value: 'subscriber.remove_from_group',
                            description: 'Fired when a subscriber is removed from a group',
                        },
                        {
                            name: 'Subscriber Unsubscribe',
                            value: 'subscriber.unsubscribe',
                            description: 'Fired when a subscriber becomes unsubscribed',
                        },
                        {
                            name: 'Subscriber Updated',
                            value: 'subscriber.update',
                            description: "Fired when any of the subscriber's custom fields are updated",
                        },
                    ],
                    required: true,
                    default: [],
                    description: 'The events to listen to',
                },
            ],
        };
        this.webhookMethods = {
            default: {
                async checkExists() {
                    const webhookUrl = this.getNodeWebhookUrl('default');
                    const webhookData = this.getWorkflowStaticData('node');
                    const event = this.getNodeParameter('event');
                    const endpoint = '/webhooks';
                    const { webhooks } = await GenericFunctions_1.mailerliteApiRequest.call(this, 'GET', endpoint, {});
                    for (const webhook of webhooks) {
                        if (webhook.url === webhookUrl && webhook.event === event) {
                            webhookData.webhookId = webhook.id;
                            return true;
                        }
                    }
                    return false;
                },
                async create() {
                    const webhookData = this.getWorkflowStaticData('node');
                    const webhookUrl = this.getNodeWebhookUrl('default');
                    const event = this.getNodeParameter('event');
                    const endpoint = '/webhooks';
                    const body = {
                        url: webhookUrl,
                        event,
                    };
                    const responseData = await GenericFunctions_1.mailerliteApiRequest.call(this, 'POST', endpoint, body);
                    if (responseData.id === undefined) {
                        return false;
                    }
                    webhookData.webhookId = responseData.id;
                    return true;
                },
                async delete() {
                    const webhookData = this.getWorkflowStaticData('node');
                    if (webhookData.webhookId !== undefined) {
                        const endpoint = `/webhooks/${webhookData.webhookId}`;
                        try {
                            await GenericFunctions_1.mailerliteApiRequest.call(this, 'DELETE', endpoint);
                        }
                        catch (error) {
                            return false;
                        }
                        delete webhookData.webhookId;
                    }
                    return true;
                },
            },
        };
    }
    async webhook() {
        const body = this.getBodyData();
        const events = body.events;
        return {
            workflowData: [this.helpers.returnJsonArray(events)],
        };
    }
}
exports.MailerLiteTrigger = MailerLiteTrigger;
