// n8n-nodes-base/dist/nodes/Microsoft/Excel/v2/methods/listSearch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getWorksheetTables = exports.getWorksheetsList = exports.searchWorkbooks = void 0;
const transport_1 = require("../transport");
async function searchWorkbooks(filter, paginationToken) {
    const q = filter ? encodeURI(`.xlsx AND ${filter}`) : '.xlsx';
    let response = {};
    if (paginationToken) {
        response = await transport_1.microsoftApiRequest.call(this, 'GET', '', undefined, undefined, paginationToken);
    }
    else {
        response = await transport_1.microsoftApiRequest.call(this, 'GET', `/drive/root/search(q='${q}')`, undefined, {
            select: 'id,name,webUrl',
            $top: 100,
        });
    }
    return {
        results: response.value.map((workbook) => {
            return {
                name: workbook.name.replace('.xlsx', ''),
                value: workbook.id,
                url: workbook.webUrl,
            };
        }),
        paginationToken: response['@odata.nextLink'],
    };
}
exports.searchWorkbooks = searchWorkbooks;
async function getWorksheetsList() {
    const workbookRLC = this.getNodeParameter('workbook');
    const workbookId = workbookRLC.value;
    let workbookURL = workbookRLC.cachedResultUrl;
    if (workbookURL.includes('1drv.ms')) {
        workbookURL = `https://onedrive.live.com/edit.aspx?resid=${workbookId}`;
    }
    let response = {};
    response = await transport_1.microsoftApiRequest.call(this, 'GET', `/drive/items/${workbookId}/workbook/worksheets`, undefined, {
        select: 'id,name',
    });
    return {
        results: response.value.map((worksheet) => ({
            name: worksheet.name,
            value: worksheet.id,
            url: `${workbookURL}&activeCell=${encodeURIComponent(worksheet.name)}!A1`,
        })),
    };
}
exports.getWorksheetsList = getWorksheetsList;
async function getWorksheetTables() {
    const workbookRLC = this.getNodeParameter('workbook');
    const workbookId = workbookRLC.value;
    let workbookURL = workbookRLC.cachedResultUrl;
    if (workbookURL.includes('1drv.ms')) {
        workbookURL = `https://onedrive.live.com/edit.aspx?resid=${workbookId}`;
    }
    const worksheetId = this.getNodeParameter('worksheet', undefined, {
        extractValue: true,
    });
    let response = {};
    response = await transport_1.microsoftApiRequest.call(this, 'GET', `/drive/items/${workbookId}/workbook/worksheets/${worksheetId}/tables`, undefined);
    const results = [];
    for (const table of response.value) {
        const name = table.name;
        const value = table.id;
        const { address } = await transport_1.microsoftApiRequest.call(this, 'GET', `/drive/items/${workbookId}/workbook/worksheets/${worksheetId}/tables/${value}/range`, undefined, {
            select: 'address',
        });
        const [sheetName, sheetRange] = address.split('!');
        const url = `${workbookURL}&activeCell=${encodeURIComponent(sheetName)}${sheetRange ? '!' + sheetRange : ''}`;
        results.push({ name, value, url });
    }
    return { results };
}
exports.getWorksheetTables = getWorksheetTables;
