// n8n-nodes-base/dist/nodes/N8nTrainingCustomerMessenger/N8nTrainingCustomerMessenger.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.N8nTrainingCustomerMessenger = void 0;
class N8nTrainingCustomerMessenger {
    constructor() {
        this.description = {
            displayName: 'Customer Messenger (n8n training)',
            name: 'n8nTrainingCustomerMessenger',
            icon: 'file:n8nTrainingCustomerMessenger.svg',
            group: ['transform'],
            version: 1,
            description: 'Dummy node used for n8n training',
            defaults: {
                name: 'Customer Messenger (n8n training)',
            },
            inputs: ['main'],
            outputs: ['main'],
            properties: [
                {
                    displayName: 'Customer ID',
                    name: 'customerId',
                    type: 'string',
                    required: true,
                    default: '',
                },
                {
                    displayName: 'Message',
                    name: 'message',
                    type: 'string',
                    required: true,
                    typeOptions: {
                        rows: 4,
                    },
                    default: '',
                },
            ],
        };
    }
    async execute() {
        const items = this.getInputData();
        const returnData = [];
        const length = items.length;
        let responseData;
        for (let i = 0; i < length; i++) {
            const customerId = this.getNodeParameter('customerId', i);
            const message = this.getNodeParameter('message', i);
            responseData = { output: `Sent message to customer ${customerId}:  ${message}` };
            const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
            returnData.push(...executionData);
        }
        return this.prepareOutputData(returnData);
    }
}
exports.N8nTrainingCustomerMessenger = N8nTrainingCustomerMessenger;
