// n8n-nodes-base/dist/nodes/SyncroMSP/v1/actions/rmm/create/description.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.rmmCreateDescription = void 0;
exports.rmmCreateDescription = [
    {
        displayName: 'Asset ID',
        name: 'assetId',
        type: 'string',
        displayOptions: {
            show: {
                resource: ['rmm'],
                operation: ['create'],
            },
        },
        default: '',
    },
    {
        displayName: 'Customer ID',
        name: 'customerId',
        type: 'string',
        displayOptions: {
            show: {
                resource: ['rmm'],
                operation: ['create'],
            },
        },
        default: '',
    },
    {
        displayName: 'Description',
        name: 'description',
        type: 'string',
        displayOptions: {
            show: {
                resource: ['rmm'],
                operation: ['create'],
            },
        },
        default: '',
    },
    {
        displayName: 'Additional Fields',
        name: 'additionalFields',
        type: 'collection',
        placeholder: 'Add Field',
        displayOptions: {
            show: {
                resource: ['rmm'],
                operation: ['create'],
            },
        },
        default: {},
        options: [
            {
                displayName: 'Resolved',
                name: 'resolved',
                type: 'boolean',
                default: false,
            },
        ],
    },
];
