// n8n-nodes-base/dist/nodes/Notion/v2/VersionDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.versionDescription = void 0;
const DatabaseDescription_1 = require("../DatabaseDescription");
const UserDescription_1 = require("../UserDescription");
const PageDescription_1 = require("../PageDescription");
const BlockDescription_1 = require("../BlockDescription");
const DatabasePageDescription_1 = require("../DatabasePageDescription");
exports.versionDescription = {
    displayName: 'Notion',
    name: 'notion',
    icon: 'file:notion.svg',
    group: ['output'],
    version: 2,
    subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
    description: 'Consume Notion API',
    defaults: {
        name: 'Notion (Beta)',
    },
    inputs: ['main'],
    outputs: ['main'],
    credentials: [
        {
            name: 'notionApi',
            required: true,
        },
    ],
    properties: [
        {
            displayName: 'In Notion, make sure to <a href="https://www.notion.so/help/add-and-manage-connections-with-the-api" target="_blank">add your connection</a> to the pages you want to access.',
            name: 'notionNotice',
            type: 'notice',
            default: '',
        },
        {
            displayName: 'Version',
            name: 'version',
            type: 'hidden',
            default: 2,
        },
        {
            displayName: 'Resource',
            name: 'resource',
            type: 'options',
            noDataExpression: true,
            options: [
                {
                    name: 'Block',
                    value: 'block',
                },
                {
                    name: 'Database',
                    value: 'database',
                },
                {
                    name: 'Database Page',
                    value: 'databasePage',
                },
                {
                    name: 'Page',
                    value: 'page',
                },
                {
                    name: 'User',
                    value: 'user',
                },
            ],
            default: 'page',
        },
        ...BlockDescription_1.blockOperations,
        ...BlockDescription_1.blockFields,
        ...DatabaseDescription_1.databaseOperations,
        ...DatabaseDescription_1.databaseFields,
        ...DatabasePageDescription_1.databasePageOperations,
        ...DatabasePageDescription_1.databasePageFields,
        ...PageDescription_1.pageOperations,
        ...PageDescription_1.pageFields,
        ...UserDescription_1.userOperations,
        ...UserDescription_1.userFields,
    ],
};
