// n8n-nodes-base/dist/nodes/Splunk/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getId = exports.populate = exports.setCount = exports.formatResults = exports.formatFeed = exports.splunkApiRequest = exports.toUnixEpoch = exports.extractErrorDescription = exports.parseXml = exports.formatSearch = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const xml2js_1 = require("xml2js");
function compactEntryContent(splunkObject) {
    if (typeof splunkObject !== 'object') {
        return {};
    }
    if (Array.isArray(splunkObject)) {
        return splunkObject.reduce((acc, cur) => {
            acc = { ...acc, ...compactEntryContent(cur) };
            return acc;
        }, {});
    }
    if (splunkObject['s:dict']) {
        const obj = splunkObject['s:dict']['s:key'];
        return { [splunkObject.$.name]: compactEntryContent(obj) };
    }
    if (splunkObject['s:list']) {
        const items = splunkObject['s:list']['s:item'];
        return { [splunkObject.$.name]: items };
    }
    if (splunkObject._) {
        return {
            [splunkObject.$.name]: splunkObject._,
        };
    }
    return {
        [splunkObject.$.name]: '',
    };
}
function formatEntryContent(content) {
    return content['s:dict']['s:key'].reduce((acc, cur) => {
        acc = { ...acc, ...compactEntryContent(cur) };
        return acc;
    }, {});
}
function formatEntry(entry) {
    const { content, link, ...rest } = entry;
    const formattedEntry = { ...rest, ...formatEntryContent(content) };
    if (formattedEntry.id) {
        formattedEntry.entryUrl = formattedEntry.id;
        formattedEntry.id = formattedEntry.id.split('/').pop();
    }
    return formattedEntry;
}
function formatSearch(responseData) {
    const { entry: entries } = responseData;
    if (!entries)
        return [];
    return Array.isArray(entries) ? entries.map(formatEntry) : [formatEntry(entries)];
}
exports.formatSearch = formatSearch;
async function parseXml(xml) {
    return new Promise((resolve, reject) => {
        (0, xml2js_1.parseString)(xml, { explicitArray: false }, (error, result) => {
            error ? reject(error) : resolve(result);
        });
    });
}
exports.parseXml = parseXml;
function extractErrorDescription(rawError) {
    var _a;
    const messages = (_a = rawError.response) === null || _a === void 0 ? void 0 : _a.messages;
    return messages ? { [messages.msg.$.type.toLowerCase()]: messages.msg._ } : rawError;
}
exports.extractErrorDescription = extractErrorDescription;
function toUnixEpoch(timestamp) {
    return Date.parse(timestamp) / 1000;
}
exports.toUnixEpoch = toUnixEpoch;
async function splunkApiRequest(method, endpoint, body = {}, qs = {}) {
    var _a;
    const { authToken, baseUrl, allowUnauthorizedCerts } = (await this.getCredentials('splunkApi'));
    const options = {
        headers: {
            Authorization: `Bearer ${authToken}`,
            'Content-Type': 'application/x-www-form-urlencoded',
        },
        method,
        form: body,
        qs,
        uri: `${baseUrl}${endpoint}`,
        json: true,
        rejectUnauthorized: !allowUnauthorizedCerts,
        useQuerystring: true,
    };
    if (!Object.keys(body).length) {
        delete options.body;
    }
    if (!Object.keys(qs).length) {
        delete options.qs;
    }
    try {
        return await this.helpers.request(options).then(parseXml);
    }
    catch (error) {
        if (((_a = error === null || error === void 0 ? void 0 : error.cause) === null || _a === void 0 ? void 0 : _a.code) === 'ECONNREFUSED') {
            throw new n8n_workflow_1.NodeApiError(this.getNode(), { ...error, code: 401 });
        }
        const rawError = (await parseXml(error.error));
        error = extractErrorDescription(rawError);
        if ('fatal' in error) {
            error = { error: error.fatal };
        }
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.splunkApiRequest = splunkApiRequest;
function formatFeed(responseData) {
    const { entry: entries } = responseData.feed;
    if (!entries)
        return [];
    return Array.isArray(entries) ? entries.map(formatEntry) : [formatEntry(entries)];
}
exports.formatFeed = formatFeed;
function compactResult(splunkObject) {
    var _a, _b, _c;
    if (typeof splunkObject !== 'object') {
        return {};
    }
    if (Array.isArray(splunkObject === null || splunkObject === void 0 ? void 0 : splunkObject.value) && ((_a = splunkObject === null || splunkObject === void 0 ? void 0 : splunkObject.value[0]) === null || _a === void 0 ? void 0 : _a.text)) {
        return {
            [splunkObject.$.k]: splunkObject.value.map((v) => v.text).join(','),
        };
    }
    if (!((_b = splunkObject === null || splunkObject === void 0 ? void 0 : splunkObject.$) === null || _b === void 0 ? void 0 : _b.k) || !((_c = splunkObject === null || splunkObject === void 0 ? void 0 : splunkObject.value) === null || _c === void 0 ? void 0 : _c.text)) {
        return {};
    }
    return {
        [splunkObject.$.k]: splunkObject.value.text,
    };
}
function formatResult(field) {
    return field.reduce((acc, cur) => {
        acc = { ...acc, ...compactResult(cur) };
        return acc;
    }, {});
}
function formatResults(responseData) {
    const results = responseData.results.result;
    if (!results)
        return [];
    return Array.isArray(results)
        ? results.map((r) => formatResult(r.field))
        : [formatResult(results.field)];
}
exports.formatResults = formatResults;
function setCount(qs) {
    qs.count = this.getNodeParameter('returnAll', 0) ? 0 : this.getNodeParameter('limit', 0);
}
exports.setCount = setCount;
function populate(source, destination) {
    if (Object.keys(source).length) {
        Object.assign(destination, source);
    }
}
exports.populate = populate;
function getId(i, idType, endpoint) {
    const id = this.getNodeParameter(idType, i);
    return id.includes(endpoint) ? id.split(endpoint).pop() : id;
}
exports.getId = getId;
