// n8n-nodes-base/dist/nodes/Uplead/CompanyDesciption.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.companyFields = exports.companyOperations = void 0;
exports.companyOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                resource: ['company'],
            },
        },
        options: [
            {
                name: 'Enrich',
                value: 'enrich',
                action: 'Enrich a company',
            },
        ],
        default: 'enrich',
    },
];
exports.companyFields = [
    {
        displayName: 'Company',
        name: 'company',
        type: 'string',
        default: '',
        displayOptions: {
            show: {
                resource: ['company'],
                operation: ['enrich'],
            },
        },
        description: 'The name of the company (e.g – amazon)',
    },
    {
        displayName: 'Domain',
        name: 'domain',
        type: 'string',
        default: '',
        displayOptions: {
            show: {
                resource: ['company'],
                operation: ['enrich'],
            },
        },
        description: 'The domain name (e.g – amazon.com)',
    },
];
