// n8n-nodes-base/dist/nodes/Google/Sheet/GoogleSheets.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GoogleSheets = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const GoogleSheetsV1_node_1 = require("./v1/GoogleSheetsV1.node");
const GoogleSheetsV2_node_1 = require("./v2/GoogleSheetsV2.node");
class GoogleSheets extends n8n_workflow_1.VersionedNodeType {
    constructor() {
        const baseDescription = {
            displayName: 'Google Sheets ',
            name: 'googleSheets',
            icon: 'file:googleSheets.svg',
            group: ['input', 'output'],
            defaultVersion: 3,
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            description: 'Read, update and write data to Google Sheets',
        };
        const nodeVersions = {
            1: new GoogleSheetsV1_node_1.GoogleSheetsV1(baseDescription),
            2: new GoogleSheetsV1_node_1.GoogleSheetsV1(baseDescription),
            3: new GoogleSheetsV2_node_1.GoogleSheetsV2(baseDescription),
        };
        super(nodeVersions, baseDescription);
    }
}
exports.GoogleSheets = GoogleSheets;
