// n8n-nodes-base/dist/nodes/Venafi/ProtectCloud/GenericFunctions.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.encryptPassphrase = exports.venafiApiRequestAllItems = exports.venafiApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const lodash_get_1 = __importDefault(require("lodash.get"));
const nacl_factory = __importStar(require("js-nacl"));
async function venafiApiRequest(method, resource, body = {}, qs = {}, uri, option = {}) {
    const operation = this.getNodeParameter('operation', 0);
    const options = {
        headers: {
            Accept: 'application/json',
            'content-type': 'application/json',
        },
        method,
        body,
        qs,
        uri: `https://api.venafi.cloud${resource}`,
        json: true,
    };
    if (Object.keys(option).length) {
        Object.assign(options, option);
    }
    if (operation === 'download') {
        if (!resource.endsWith('keystore')) {
            delete options.headers.Accept;
            delete options.headers['content-type'];
        }
    }
    try {
        if (Object.keys(body).length === 0) {
            delete options.body;
        }
        return await this.helpers.requestWithAuthentication.call(this, 'venafiTlsProtectCloudApi', options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.venafiApiRequest = venafiApiRequest;
async function venafiApiRequestAllItems(propertyName, method, endpoint, body = {}, query = {}) {
    var _a;
    const returnData = [];
    let responseData;
    do {
        responseData = await venafiApiRequest.call(this, method, endpoint, body, query);
        endpoint = (0, lodash_get_1.default)(responseData, '_links[0].Next');
        returnData.push.apply(returnData, responseData[propertyName]);
    } while ((_a = responseData._links) === null || _a === void 0 ? void 0 : _a[0].Next);
    return returnData;
}
exports.venafiApiRequestAllItems = venafiApiRequestAllItems;
async function encryptPassphrase(certificateId, passphrase, storePassphrase) {
    let dekHash = '';
    const dekResponse = await venafiApiRequest.call(this, 'GET', `/outagedetection/v1/certificates/${certificateId}`);
    if (dekResponse.dekHash) {
        dekHash = dekResponse.dekHash;
    }
    let pubKey = '';
    const pubKeyResponse = await venafiApiRequest.call(this, 'GET', `/v1/edgeencryptionkeys/${dekHash}`);
    if (pubKeyResponse.key) {
        pubKey = pubKeyResponse.key;
    }
    let encryptedKeyPass = '';
    let encryptedKeyStorePass = '';
    const promise = async () => {
        return new Promise((resolve, reject) => {
            nacl_factory.instantiate((nacl) => {
                try {
                    const passphraseUTF8 = nacl.encode_utf8(passphrase);
                    const keyPassBuffer = nacl.crypto_box_seal(passphraseUTF8, Buffer.from(pubKey, 'base64'));
                    encryptedKeyPass = Buffer.from(keyPassBuffer).toString('base64');
                    const storePassphraseUTF8 = nacl.encode_utf8(storePassphrase);
                    const keyStorePassBuffer = nacl.crypto_box_seal(storePassphraseUTF8, Buffer.from(pubKey, 'base64'));
                    encryptedKeyStorePass = Buffer.from(keyStorePassBuffer).toString('base64');
                    return resolve([encryptedKeyPass, encryptedKeyStorePass]);
                }
                catch (error) {
                    return reject(error);
                }
            });
        });
    };
    return promise();
}
exports.encryptPassphrase = encryptPassphrase;
