// n8n-nodes-base/dist/nodes/WooCommerce/GenericFunctions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.adjustMetadata = exports.setFields = exports.toSnakeCase = exports.setMetadata = exports.getAutomaticSecret = exports.woocommerceApiRequestAllItems = exports.woocommerceApiRequest = void 0;
const crypto_1 = require("crypto");
const change_case_1 = require("change-case");
const lodash_omit_1 = __importDefault(require("lodash.omit"));
async function woocommerceApiRequest(method, resource, body = {}, qs = {}, uri, option = {}) {
    const credentials = await this.getCredentials('wooCommerceApi');
    let options = {
        method,
        qs,
        body,
        uri: uri || `${credentials.url}/wp-json/wc/v3${resource}`,
        json: true,
    };
    if (!Object.keys(body).length) {
        delete options.form;
    }
    options = Object.assign({}, options, option);
    return this.helpers.requestWithAuthentication.call(this, 'wooCommerceApi', options);
}
exports.woocommerceApiRequest = woocommerceApiRequest;
async function woocommerceApiRequestAllItems(method, endpoint, body = {}, query = {}) {
    var _a;
    const returnData = [];
    let responseData;
    let uri;
    query.per_page = 100;
    do {
        responseData = await woocommerceApiRequest.call(this, method, endpoint, body, query, uri, {
            resolveWithFullResponse: true,
        });
        const links = responseData.headers.link.split(',');
        const nextLink = links.find((link) => link.indexOf('rel="next"') !== -1);
        if (nextLink) {
            uri = nextLink.split(';')[0].replace(/<(.*)>/, '$1');
        }
        returnData.push.apply(returnData, responseData.body);
    } while ((_a = responseData.headers.link) === null || _a === void 0 ? void 0 : _a.includes('rel="next"'));
    return returnData;
}
exports.woocommerceApiRequestAllItems = woocommerceApiRequestAllItems;
function getAutomaticSecret(credentials) {
    const data = `${credentials.consumerKey},${credentials.consumerSecret}`;
    return (0, crypto_1.createHash)('md5').update(data).digest('hex');
}
exports.getAutomaticSecret = getAutomaticSecret;
function setMetadata(data) {
    var _a;
    for (let i = 0; i < data.length; i++) {
        if ((_a = data[i].metadataUi) === null || _a === void 0 ? void 0 : _a.metadataValues) {
            data[i].meta_data = data[i].metadataUi.metadataValues;
            delete data[i].metadataUi;
        }
        else {
            delete data[i].metadataUi;
        }
    }
}
exports.setMetadata = setMetadata;
function toSnakeCase(data) {
    if (!Array.isArray(data)) {
        data = [data];
    }
    let remove = false;
    for (let i = 0; i < data.length; i++) {
        for (const key of Object.keys(data[i])) {
            if (data[i][(0, change_case_1.snakeCase)(key)] === undefined) {
                remove = true;
            }
            data[i][(0, change_case_1.snakeCase)(key)] = data[i][key];
            if (remove) {
                delete data[i][key];
                remove = false;
            }
        }
    }
}
exports.toSnakeCase = toSnakeCase;
function setFields(fieldsToSet, body) {
    for (const fields in fieldsToSet) {
        if (fields === 'tags') {
            body.tags = fieldsToSet[fields].map((tag) => ({ id: parseInt(tag, 10) }));
        }
        else {
            body[(0, change_case_1.snakeCase)(fields.toString())] = fieldsToSet[fields];
        }
    }
}
exports.setFields = setFields;
function adjustMetadata(fields) {
    if (!fields.meta_data)
        return fields;
    return {
        ...(0, lodash_omit_1.default)(fields, ['meta_data']),
        meta_data: fields.meta_data.meta_data_fields,
    };
}
exports.adjustMetadata = adjustMetadata;
