// n8n-nodes-base/dist/nodes/Storyblok/StoryManagementDescription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.storyManagementFields = exports.storyManagementOperations = void 0;
exports.storyManagementOperations = [
    {
        displayName: 'Operation',
        name: 'operation',
        type: 'options',
        noDataExpression: true,
        displayOptions: {
            show: {
                source: ['managementApi'],
                resource: ['story'],
            },
        },
        options: [
            {
                name: 'Delete',
                value: 'delete',
                description: 'Delete a story',
                action: 'Delete a story',
            },
            {
                name: 'Get',
                value: 'get',
                description: 'Get a story',
                action: 'Get a story',
            },
            {
                name: 'Get Many',
                value: 'getAll',
                description: 'Get many stories',
                action: 'Get many stories',
            },
            {
                name: 'Publish',
                value: 'publish',
                description: 'Publish a story',
                action: 'Publish a story',
            },
            {
                name: 'Unpublish',
                value: 'unpublish',
                description: 'Unpublish a story',
                action: 'Unpublish a story',
            },
        ],
        default: 'get',
    },
];
exports.storyManagementFields = [
    {
        displayName: 'Space Name or ID',
        name: 'space',
        type: 'options',
        typeOptions: {
            loadOptionsMethod: 'getSpaces',
        },
        default: '',
        required: true,
        displayOptions: {
            show: {
                source: ['managementApi'],
                resource: ['story'],
                operation: ['delete'],
            },
        },
        description: 'The name of the space. Choose from the list, or specify an ID using an <a href="https://docs.n8n.io/code-examples/expressions/">expression</a>.',
    },
    {
        displayName: 'Story ID',
        name: 'storyId',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                source: ['managementApi'],
                resource: ['story'],
                operation: ['delete'],
            },
        },
        description: 'Numeric ID of the story',
    },
    {
        displayName: 'Space Name or ID',
        name: 'space',
        type: 'options',
        typeOptions: {
            loadOptionsMethod: 'getSpaces',
        },
        default: '',
        required: true,
        displayOptions: {
            show: {
                source: ['managementApi'],
                resource: ['story'],
                operation: ['get'],
            },
        },
        description: 'The name of the space. Choose from the list, or specify an ID using an <a href="https://docs.n8n.io/code-examples/expressions/">expression</a>.',
    },
    {
        displayName: 'Story ID',
        name: 'storyId',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                source: ['managementApi'],
                resource: ['story'],
                operation: ['get'],
            },
        },
        description: 'Numeric ID of the story',
    },
    {
        displayName: 'Space Name or ID',
        name: 'space',
        type: 'options',
        typeOptions: {
            loadOptionsMethod: 'getSpaces',
        },
        default: '',
        required: true,
        displayOptions: {
            show: {
                source: ['managementApi'],
                resource: ['story'],
                operation: ['getAll'],
            },
        },
        description: 'The name of the space. Choose from the list, or specify an ID using an <a href="https://docs.n8n.io/code-examples/expressions/">expression</a>.',
    },
    {
        displayName: 'Return All',
        name: 'returnAll',
        type: 'boolean',
        displayOptions: {
            show: {
                source: ['managementApi'],
                resource: ['story'],
                operation: ['getAll'],
            },
        },
        default: false,
        description: 'Whether to return all results or only up to a given limit',
    },
    {
        displayName: 'Limit',
        name: 'limit',
        type: 'number',
        displayOptions: {
            show: {
                source: ['managementApi'],
                resource: ['story'],
                operation: ['getAll'],
                returnAll: [false],
            },
        },
        typeOptions: {
            minValue: 1,
            maxValue: 100,
        },
        default: 50,
        description: 'Max number of results to return',
    },
    {
        displayName: 'Filters',
        name: 'filters',
        type: 'collection',
        placeholder: 'Add Filter',
        default: {},
        displayOptions: {
            show: {
                source: ['managementApi'],
                resource: ['story'],
                operation: ['getAll'],
            },
        },
        options: [
            {
                displayName: 'Starts With',
                name: 'starts_with',
                type: 'string',
                default: '',
                description: 'Filter by slug',
            },
        ],
    },
    {
        displayName: 'Space Name or ID',
        name: 'space',
        type: 'options',
        typeOptions: {
            loadOptionsMethod: 'getSpaces',
        },
        default: '',
        required: true,
        displayOptions: {
            show: {
                source: ['managementApi'],
                resource: ['story'],
                operation: ['publish'],
            },
        },
        description: 'The name of the space. Choose from the list, or specify an ID using an <a href="https://docs.n8n.io/code-examples/expressions/">expression</a>.',
    },
    {
        displayName: 'Story ID',
        name: 'storyId',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                source: ['managementApi'],
                resource: ['story'],
                operation: ['publish'],
            },
        },
        description: 'Numeric ID of the story',
    },
    {
        displayName: 'Options',
        name: 'options',
        type: 'collection',
        placeholder: 'Add Option',
        displayOptions: {
            show: {
                source: ['managementApi'],
                resource: ['story'],
                operation: ['publish'],
            },
        },
        default: {},
        options: [
            {
                displayName: 'Release ID',
                name: 'releaseId',
                type: 'string',
                default: '',
                description: 'Numeric ID of release',
            },
            {
                displayName: 'Language',
                name: 'language',
                type: 'string',
                default: '',
                description: 'Language code to publish the story individually (must be enabled in the space settings)',
            },
        ],
    },
    {
        displayName: 'Space Name or ID',
        name: 'space',
        type: 'options',
        typeOptions: {
            loadOptionsMethod: 'getSpaces',
        },
        default: '',
        required: true,
        displayOptions: {
            show: {
                source: ['managementApi'],
                resource: ['story'],
                operation: ['unpublish'],
            },
        },
        description: 'The name of the space. Choose from the list, or specify an ID using an <a href="https://docs.n8n.io/code-examples/expressions/">expression</a>.',
    },
    {
        displayName: 'Story ID',
        name: 'storyId',
        type: 'string',
        default: '',
        required: true,
        displayOptions: {
            show: {
                source: ['managementApi'],
                resource: ['story'],
                operation: ['unpublish'],
            },
        },
        description: 'Numeric ID of the story',
    },
];
