// n8n-nodes-base/dist/nodes/Cockpit/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createDataFromParameters = exports.cockpitApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
async function cockpitApiRequest(method, resource, body = {}, uri, option = {}) {
    const credentials = await this.getCredentials('cockpitApi');
    let options = {
        headers: {
            Accept: 'application/json',
            'Content-Type': 'application/json',
        },
        method,
        qs: {
            token: credentials.accessToken,
        },
        body,
        uri: uri || `${credentials.url}/api${resource}`,
        json: true,
    };
    options = Object.assign({}, options, option);
    if (Object.keys(options.body).length === 0) {
        delete options.body;
    }
    try {
        return await this.helpers.request(options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.cockpitApiRequest = cockpitApiRequest;
function createDataFromParameters(itemIndex) {
    const dataFieldsAreJson = this.getNodeParameter('jsonDataFields', itemIndex);
    if (dataFieldsAreJson) {
        return (0, n8n_workflow_1.jsonParse)(this.getNodeParameter('dataFieldsJson', itemIndex, '{}'));
    }
    const uiDataFields = this.getNodeParameter('dataFieldsUi', itemIndex, {});
    const unpacked = {};
    if (uiDataFields.field === undefined) {
        return unpacked;
    }
    for (const field of uiDataFields.field) {
        unpacked[field.name] = field.value;
    }
    return unpacked;
}
exports.createDataFromParameters = createDataFromParameters;
