// n8n-nodes-base/dist/nodes/OpenAi/OpenAi.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenAi = void 0;
const ImageDescription_1 = require("./ImageDescription");
const TextDescription_1 = require("./TextDescription");
const ChatDescription_1 = require("./ChatDescription");
class OpenAi {
    constructor() {
        this.description = {
            displayName: 'OpenAI',
            name: 'openAi',
            icon: 'file:openAi.svg',
            group: ['transform'],
            version: 1,
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            description: 'Consume Open AI',
            defaults: {
                name: 'OpenAI',
            },
            inputs: ['main'],
            outputs: ['main'],
            credentials: [
                {
                    name: 'openAiApi',
                    required: true,
                },
            ],
            requestDefaults: {
                baseURL: 'https://api.openai.com',
            },
            properties: [
                {
                    displayName: 'Resource',
                    name: 'resource',
                    type: 'options',
                    noDataExpression: true,
                    options: [
                        {
                            name: 'Chat',
                            value: 'chat',
                        },
                        {
                            name: 'Image',
                            value: 'image',
                        },
                        {
                            name: 'Text',
                            value: 'text',
                        },
                    ],
                    default: 'text',
                },
                ...ChatDescription_1.chatOperations,
                ...ChatDescription_1.chatFields,
                ...ImageDescription_1.imageOperations,
                ...ImageDescription_1.imageFields,
                ...TextDescription_1.textOperations,
                ...TextDescription_1.textFields,
            ],
        };
    }
}
exports.OpenAi = OpenAi;
