// n8n/dist/push/sse.push.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SSEPush = void 0;
const sse_channel_1 = __importDefault(require("sse-channel"));
const abstract_push_1 = require("./abstract.push");
class SSEPush extends abstract_push_1.AbstractPush {
    constructor() {
        super();
        this.channel = new sse_channel_1.default();
        this.connections = {};
        this.channel.on('disconnect', (channel, { req }) => {
            var _a;
            this.remove((_a = req === null || req === void 0 ? void 0 : req.query) === null || _a === void 0 ? void 0 : _a.sessionId);
        });
    }
    add(sessionId, connection) {
        super.add(sessionId, connection);
        this.channel.addClient(connection.req, connection.res);
    }
    close({ res }) {
        res.end();
        this.channel.removeClient(res);
    }
    sendToOne(connection, data) {
        this.channel.send(data, [connection.res]);
    }
}
exports.SSEPush = SSEPush;
