// n8n-nodes-base/dist/nodes/Spotify/GenericFunctions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.spotifyApiRequestAllItems = exports.spotifyApiRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const lodash_get_1 = __importDefault(require("lodash.get"));
async function spotifyApiRequest(method, endpoint, body, query, uri) {
    const options = {
        method,
        headers: {
            'User-Agent': 'n8n',
            'Content-Type': 'text/plain',
            Accept: ' application/json',
        },
        qs: query,
        uri: uri || `https://api.spotify.com/v1${endpoint}`,
        json: true,
    };
    if (Object.keys(body).length > 0) {
        options.body = body;
    }
    try {
        return await this.helpers.requestOAuth2.call(this, 'spotifyOAuth2Api', options);
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
exports.spotifyApiRequest = spotifyApiRequest;
async function spotifyApiRequestAllItems(propertyName, method, endpoint, body, query) {
    const returnData = [];
    let responseData;
    let uri;
    do {
        responseData = await spotifyApiRequest.call(this, method, endpoint, body, query, uri);
        returnData.push.apply(returnData, (0, lodash_get_1.default)(responseData, propertyName));
        uri = responseData.next || responseData[propertyName.split('.')[0]].next;
        query = {};
        if ((uri === null || uri === void 0 ? void 0 : uri.includes('offset=1000')) && endpoint === '/search') {
            return returnData;
        }
    } while ((responseData.next !== null && responseData.next !== undefined) ||
        (responseData[propertyName.split('.')[0]].next !== null &&
            responseData[propertyName.split('.')[0]].next !== undefined));
    return returnData;
}
exports.spotifyApiRequestAllItems = spotifyApiRequestAllItems;
