// n8n-nodes-base/dist/nodes/HttpRequest/HttpRequest.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpRequest = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const HttpRequestV1_node_1 = require("./V1/HttpRequestV1.node");
const HttpRequestV2_node_1 = require("./V2/HttpRequestV2.node");
const HttpRequestV3_node_1 = require("./V3/HttpRequestV3.node");
class HttpRequest extends n8n_workflow_1.VersionedNodeType {
    constructor() {
        const baseDescription = {
            displayName: 'HTTP Request',
            name: 'httpRequest',
            icon: 'fa:at',
            group: ['output'],
            subtitle: '={{$parameter["requestMethod"] + ": " + $parameter["url"]}}',
            description: 'Makes an HTTP request and returns the response data',
            defaultVersion: 4.1,
        };
        const nodeVersions = {
            1: new HttpRequestV1_node_1.HttpRequestV1(baseDescription),
            2: new HttpRequestV2_node_1.HttpRequestV2(baseDescription),
            3: new HttpRequestV3_node_1.HttpRequestV3(baseDescription),
            4: new HttpRequestV3_node_1.HttpRequestV3(baseDescription),
            4.1: new HttpRequestV3_node_1.HttpRequestV3(baseDescription),
        };
        super(nodeVersions, baseDescription);
    }
}
exports.HttpRequest = HttpRequest;
