// n8n-nodes-base/dist/nodes/SendInBlue/GenericFunctions.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SendInBlueWebhookApi = exports.SendInBlueNode = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const mail_composer_1 = __importDefault(require("nodemailer/lib/mail-composer"));
var SendInBlueNode;
(function (SendInBlueNode) {
    let OVERRIDE_MAP_VALUES;
    (function (OVERRIDE_MAP_VALUES) {
        OVERRIDE_MAP_VALUES["CATEGORY"] = "category";
        OVERRIDE_MAP_VALUES["NORMAL"] = "boolean";
        OVERRIDE_MAP_VALUES["TRANSACTIONAL"] = "id";
    })(OVERRIDE_MAP_VALUES || (OVERRIDE_MAP_VALUES = {}));
    let OVERRIDE_MAP_TYPE;
    (function (OVERRIDE_MAP_TYPE) {
        OVERRIDE_MAP_TYPE["CATEGORY"] = "category";
        OVERRIDE_MAP_TYPE["NORMAL"] = "normal";
        OVERRIDE_MAP_TYPE["TRANSACTIONAL"] = "transactional";
    })(OVERRIDE_MAP_TYPE || (OVERRIDE_MAP_TYPE = {}));
    SendInBlueNode.INTERCEPTORS = new Map([
        [
            "category",
            (body) => {
                body.type = "category";
            },
        ],
        [
            "normal",
            (body) => {
                body.type = "boolean";
            },
        ],
        [
            "transactional",
            (body) => {
                body.type = "id";
            },
        ],
    ]);
    let Validators;
    (function (Validators) {
        function getFileName(itemIndex, mimeType, fileExt, fileName) {
            let ext = fileExt;
            if (fileExt === undefined) {
                ext = mimeType.split('/')[1];
            }
            let name = `${fileName}.${ext}`;
            if (fileName === undefined) {
                name = `file-${itemIndex}.${ext}`;
            }
            return name;
        }
        async function validateAndCompileAttachmentsData(requestOptions) {
            const dataPropertyList = this.getNodeParameter('additionalFields.emailAttachments.attachment');
            const { body } = requestOptions;
            const { attachment = [] } = body;
            try {
                const { binaryPropertyName } = dataPropertyList;
                const dataMappingList = binaryPropertyName.split(',');
                for (const attachmentDataName of dataMappingList) {
                    const binaryData = this.helpers.assertBinaryData(attachmentDataName);
                    const bufferFromIncomingData = await this.helpers.getBinaryDataBuffer(attachmentDataName);
                    const { data: content, mimeType, fileName, fileExtension, } = await this.helpers.prepareBinaryData(bufferFromIncomingData);
                    const itemIndex = this.getItemIndex();
                    const name = getFileName(itemIndex, mimeType, fileExtension, fileName !== null && fileName !== void 0 ? fileName : binaryData.fileName);
                    attachment.push({ content, name });
                }
                Object.assign(body, { attachment });
                return requestOptions;
            }
            catch (err) {
                throw new n8n_workflow_1.NodeOperationError(this.getNode(), err);
            }
        }
        Validators.validateAndCompileAttachmentsData = validateAndCompileAttachmentsData;
        async function validateAndCompileTags(requestOptions) {
            const { tag } = this.getNodeParameter('additionalFields.emailTags.tags');
            const tags = tag
                .split(',')
                .map((entry) => entry.trim())
                .filter((entry) => {
                return entry !== '';
            });
            const { body } = requestOptions;
            Object.assign(body, { tags });
            return requestOptions;
        }
        Validators.validateAndCompileTags = validateAndCompileTags;
        function formatToEmailName(data) {
            const { address: email, name } = data;
            const result = { email };
            if (name !== undefined && name !== '') {
                Object.assign(result, { name });
            }
            return { ...result };
        }
        function validateEmailStrings(input) {
            const composer = new mail_composer_1.default({ ...input });
            const addressFields = composer.compile().getAddresses();
            const fieldFetcher = new Map([
                [
                    'bcc',
                    () => {
                        var _a;
                        return (_a = addressFields.bcc) === null || _a === void 0 ? void 0 : _a.map(formatToEmailName);
                    },
                ],
                [
                    'cc',
                    () => {
                        var _a;
                        return (_a = addressFields.cc) === null || _a === void 0 ? void 0 : _a.map(formatToEmailName);
                    },
                ],
                [
                    'from',
                    () => {
                        var _a;
                        return (_a = addressFields.from) === null || _a === void 0 ? void 0 : _a.map(formatToEmailName);
                    },
                ],
                [
                    'reply-to',
                    () => {
                        var _a;
                        return (_a = addressFields['reply-to']) === null || _a === void 0 ? void 0 : _a.map(formatToEmailName);
                    },
                ],
                [
                    'sender',
                    () => {
                        var _a;
                        return (_a = addressFields.sender) === null || _a === void 0 ? void 0 : _a.map(formatToEmailName)[0];
                    },
                ],
                [
                    'to',
                    () => {
                        var _a;
                        return (_a = addressFields.to) === null || _a === void 0 ? void 0 : _a.map(formatToEmailName);
                    },
                ],
            ]);
            const result = {};
            Object.keys(input).reduce((obj, key) => {
                const getter = fieldFetcher.get(key);
                const value = getter();
                obj[key] = value;
                return obj;
            }, result);
            return result;
        }
        async function validateAndCompileCCEmails(requestOptions) {
            const ccData = this.getNodeParameter('additionalFields.receipientsCC.receipientCc');
            const { cc } = ccData;
            const { body } = requestOptions;
            const data = validateEmailStrings({ cc: cc });
            Object.assign(body, data);
            return requestOptions;
        }
        Validators.validateAndCompileCCEmails = validateAndCompileCCEmails;
        async function validateAndCompileBCCEmails(requestOptions) {
            const bccData = this.getNodeParameter('additionalFields.receipientsBCC.receipientBcc');
            const { bcc } = bccData;
            const { body } = requestOptions;
            const data = validateEmailStrings({ bcc: bcc });
            Object.assign(body, data);
            return requestOptions;
        }
        Validators.validateAndCompileBCCEmails = validateAndCompileBCCEmails;
        async function validateAndCompileReceipientEmails(requestOptions) {
            const to = this.getNodeParameter('receipients');
            const { body } = requestOptions;
            const data = validateEmailStrings({ to });
            Object.assign(body, data);
            return requestOptions;
        }
        Validators.validateAndCompileReceipientEmails = validateAndCompileReceipientEmails;
        async function validateAndCompileSenderEmail(requestOptions) {
            const sender = this.getNodeParameter('sender');
            const { body } = requestOptions;
            const data = validateEmailStrings({ sender });
            Object.assign(body, data);
            return requestOptions;
        }
        Validators.validateAndCompileSenderEmail = validateAndCompileSenderEmail;
        async function validateAndCompileTemplateParameters(requestOptions) {
            const parameterData = this.getNodeParameter('additionalFields.templateParameters.parameterValues');
            const { body } = requestOptions;
            const { parameters } = parameterData;
            const params = parameters
                .split(',')
                .filter((parameter) => {
                return parameter.split('=').length === 2;
            })
                .map((parameter) => {
                const [key, value] = parameter.split('=');
                return {
                    [key]: value,
                };
            })
                .reduce((obj, cObj) => {
                Object.assign(obj, cObj);
                return obj;
            }, {});
            Object.assign(body, { params });
            return requestOptions;
        }
        Validators.validateAndCompileTemplateParameters = validateAndCompileTemplateParameters;
    })(Validators = SendInBlueNode.Validators || (SendInBlueNode.Validators = {}));
})(SendInBlueNode = exports.SendInBlueNode || (exports.SendInBlueNode = {}));
var SendInBlueWebhookApi;
(function (SendInBlueWebhookApi) {
    const credentialsName = 'sendInBlueApi';
    const baseURL = 'https://api.sendinblue.com/v3';
    SendInBlueWebhookApi.supportedAuthMap = new Map([
        [
            'apiKey',
            async (ref) => {
                const credentials = await ref.getCredentials(credentialsName);
                return credentials.sharedSecret;
            },
        ],
    ]);
    SendInBlueWebhookApi.fetchWebhooks = async (ref, type) => {
        const endpoint = `${baseURL}/webhooks?type=${type}`;
        const options = {
            method: 'GET',
            headers: {
                Accept: 'application/json',
            },
            uri: endpoint,
        };
        const webhooks = (await ref.helpers.requestWithAuthentication.call(ref, credentialsName, options));
        return (0, n8n_workflow_1.jsonParse)(webhooks);
    };
    SendInBlueWebhookApi.createWebHook = async (ref, type, events, url) => {
        const endpoint = `${baseURL}/webhooks`;
        const options = {
            method: 'POST',
            headers: {
                Accept: 'application/json',
            },
            uri: endpoint,
            body: {
                events,
                type,
                url,
            },
        };
        const webhookId = await ref.helpers.requestWithAuthentication.call(ref, credentialsName, options);
        return (0, n8n_workflow_1.jsonParse)(webhookId);
    };
    SendInBlueWebhookApi.deleteWebhook = async (ref, webhookId) => {
        const endpoint = `${baseURL}/webhooks/${webhookId}`;
        const body = {};
        const options = {
            method: 'DELETE',
            headers: {
                Accept: 'application/json',
            },
            uri: endpoint,
            body,
        };
        return ref.helpers.requestWithAuthentication.call(ref, credentialsName, options);
    };
})(SendInBlueWebhookApi = exports.SendInBlueWebhookApi || (exports.SendInBlueWebhookApi = {}));
