// n8n-nodes-base/dist/nodes/SendInBlue/SendInBlue.node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SendInBlue = void 0;
const AttributeDescription_1 = require("./AttributeDescription");
const ContactDescription_1 = require("./ContactDescription");
const EmailDescription_1 = require("./EmailDescription");
const SenderDescrition_1 = require("./SenderDescrition");
class SendInBlue {
    constructor() {
        this.description = {
            displayName: 'SendInBlue',
            name: 'sendInBlue',
            icon: 'file:sendinblue.svg',
            group: ['transform'],
            version: 1,
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            description: 'Consume Sendinblue API',
            defaults: {
                name: 'SendInBlue',
            },
            inputs: ['main'],
            outputs: ['main'],
            credentials: [
                {
                    name: 'sendInBlueApi',
                    required: true,
                },
            ],
            requestDefaults: {
                baseURL: 'https://api.sendinblue.com',
            },
            properties: [
                {
                    displayName: 'Resource',
                    name: 'resource',
                    type: 'options',
                    noDataExpression: true,
                    options: [
                        {
                            name: 'Contact',
                            value: 'contact',
                        },
                        {
                            name: 'Contact Attribute',
                            value: 'attribute',
                        },
                        {
                            name: 'Email',
                            value: 'email',
                        },
                        {
                            name: 'Sender',
                            value: 'sender',
                        },
                    ],
                    default: 'email',
                },
                ...AttributeDescription_1.attributeOperations,
                ...AttributeDescription_1.attributeFields,
                ...SenderDescrition_1.senderOperations,
                ...SenderDescrition_1.senderFields,
                ...ContactDescription_1.contactOperations,
                ...ContactDescription_1.contactFields,
                ...EmailDescription_1.emailOperations,
                ...EmailDescription_1.emailFields,
            ],
        };
    }
}
exports.SendInBlue = SendInBlue;
