// n8n-nodes-base/dist/nodes/QuickBooks/QuickBooks.node.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QuickBooks = void 0;
const n8n_workflow_1 = require("n8n-workflow");
const descriptions_1 = require("./descriptions");
const GenericFunctions_1 = require("./GenericFunctions");
const change_case_1 = require("change-case");
const lodash_isempty_1 = __importDefault(require("lodash.isempty"));
class QuickBooks {
    constructor() {
        this.description = {
            displayName: 'QuickBooks Online',
            name: 'quickbooks',
            icon: 'file:quickbooks.svg',
            group: ['transform'],
            version: 1,
            subtitle: '={{$parameter["operation"] + ": " + $parameter["resource"]}}',
            description: 'Consume the QuickBooks Online API',
            defaults: {
                name: 'QuickBooks Online',
            },
            inputs: ['main'],
            outputs: ['main'],
            credentials: [
                {
                    name: 'quickBooksOAuth2Api',
                    required: true,
                },
            ],
            properties: [
                {
                    displayName: 'Resource',
                    name: 'resource',
                    type: 'options',
                    noDataExpression: true,
                    options: [
                        {
                            name: 'Bill',
                            value: 'bill',
                        },
                        {
                            name: 'Customer',
                            value: 'customer',
                        },
                        {
                            name: 'Employee',
                            value: 'employee',
                        },
                        {
                            name: 'Estimate',
                            value: 'estimate',
                        },
                        {
                            name: 'Invoice',
                            value: 'invoice',
                        },
                        {
                            name: 'Item',
                            value: 'item',
                        },
                        {
                            name: 'Payment',
                            value: 'payment',
                        },
                        {
                            name: 'Purchase',
                            value: 'purchase',
                        },
                        {
                            name: 'Transaction',
                            value: 'transaction',
                        },
                        {
                            name: 'Vendor',
                            value: 'vendor',
                        },
                    ],
                    default: 'customer',
                },
                ...descriptions_1.billOperations,
                ...descriptions_1.billFields,
                ...descriptions_1.customerOperations,
                ...descriptions_1.customerFields,
                ...descriptions_1.employeeOperations,
                ...descriptions_1.employeeFields,
                ...descriptions_1.estimateOperations,
                ...descriptions_1.estimateFields,
                ...descriptions_1.invoiceOperations,
                ...descriptions_1.invoiceFields,
                ...descriptions_1.itemOperations,
                ...descriptions_1.itemFields,
                ...descriptions_1.paymentOperations,
                ...descriptions_1.paymentFields,
                ...descriptions_1.purchaseOperations,
                ...descriptions_1.purchaseFields,
                ...descriptions_1.transactionOperations,
                ...descriptions_1.transactionFields,
                ...descriptions_1.vendorOperations,
                ...descriptions_1.vendorFields,
            ],
        };
        this.methods = {
            loadOptions: {
                async getCustomers() {
                    return GenericFunctions_1.loadResource.call(this, 'customer');
                },
                async getCustomFields() {
                    return GenericFunctions_1.loadResource.call(this, 'preferences');
                },
                async getDepartments() {
                    return GenericFunctions_1.loadResource.call(this, 'department');
                },
                async getItems() {
                    return GenericFunctions_1.loadResource.call(this, 'item');
                },
                async getMemos() {
                    return GenericFunctions_1.loadResource.call(this, 'CreditMemo');
                },
                async getPurchases() {
                    return GenericFunctions_1.loadResource.call(this, 'purchase');
                },
                async getTaxCodeRefs() {
                    return GenericFunctions_1.loadResource.call(this, 'TaxCode');
                },
                async getTerms() {
                    return GenericFunctions_1.loadResource.call(this, 'Term');
                },
                async getVendors() {
                    return GenericFunctions_1.loadResource.call(this, 'vendor');
                },
            },
        };
    }
    async execute() {
        const items = this.getInputData();
        const resource = this.getNodeParameter('resource', 0);
        const operation = this.getNodeParameter('operation', 0);
        let responseData;
        const returnData = [];
        const { oauthTokenData } = (await this.getCredentials('quickBooksOAuth2Api'));
        const companyId = oauthTokenData.callbackQueryString.realmId;
        for (let i = 0; i < items.length; i++) {
            try {
                if (resource === 'bill') {
                    if (operation === 'create') {
                        const lines = this.getNodeParameter('Line', i);
                        if (!lines.length) {
                            throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Please enter at least one line for the ${resource}.`, { itemIndex: i });
                        }
                        if (lines.some((line) => line.DetailType === undefined ||
                            line.Amount === undefined ||
                            line.Description === undefined)) {
                            throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Please enter detail type, amount and description for every line.', { itemIndex: i });
                        }
                        lines.forEach((line) => {
                            if (line.DetailType === 'AccountBasedExpenseLineDetail' &&
                                line.accountId === undefined) {
                                throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Please enter an account ID for the associated line.', { itemIndex: i });
                            }
                            else if (line.DetailType === 'ItemBasedExpenseLineDetail' &&
                                line.itemId === undefined) {
                                throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Please enter an item ID for the associated line.', { itemIndex: i });
                            }
                        });
                        let body = {
                            VendorRef: {
                                value: this.getNodeParameter('VendorRef', i),
                            },
                        };
                        body.Line = GenericFunctions_1.processLines.call(this, body, lines, resource);
                        const additionalFields = this.getNodeParameter('additionalFields', i);
                        body = GenericFunctions_1.populateFields.call(this, body, additionalFields, resource);
                        const endpoint = `/v3/company/${companyId}/${resource}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, {}, body);
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'delete') {
                        const qs = {
                            operation: 'delete',
                        };
                        const body = {
                            Id: this.getNodeParameter('billId', i),
                            SyncToken: await GenericFunctions_1.getSyncToken.call(this, i, companyId, resource),
                        };
                        const endpoint = `/v3/company/${companyId}/${resource}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, qs, body);
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'get') {
                        const billId = this.getNodeParameter('billId', i);
                        const endpoint = `/v3/company/${companyId}/${resource}/${billId}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'GET', endpoint, {}, {});
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'getAll') {
                        const endpoint = `/v3/company/${companyId}/query`;
                        responseData = await GenericFunctions_1.handleListing.call(this, i, endpoint, resource);
                    }
                    else if (operation === 'update') {
                        const { ref, syncToken } = await GenericFunctions_1.getRefAndSyncToken.call(this, i, companyId, resource, 'VendorRef');
                        let body = {
                            Id: this.getNodeParameter('billId', i),
                            SyncToken: syncToken,
                            sparse: true,
                            VendorRef: {
                                name: ref.name,
                                value: ref.value,
                            },
                        };
                        const updateFields = this.getNodeParameter('updateFields', i);
                        if ((0, lodash_isempty_1.default)(updateFields)) {
                            throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Please enter at least one field to update for the ${resource}.`, { itemIndex: i });
                        }
                        body = GenericFunctions_1.populateFields.call(this, body, updateFields, resource);
                        const endpoint = `/v3/company/${companyId}/${resource}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, {}, body);
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                }
                else if (resource === 'customer') {
                    if (operation === 'create') {
                        let body = {
                            DisplayName: this.getNodeParameter('displayName', i),
                        };
                        const additionalFields = this.getNodeParameter('additionalFields', i);
                        body = GenericFunctions_1.populateFields.call(this, body, additionalFields, resource);
                        const endpoint = `/v3/company/${companyId}/${resource}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, {}, body);
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'get') {
                        const customerId = this.getNodeParameter('customerId', i);
                        const endpoint = `/v3/company/${companyId}/${resource}/${customerId}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'GET', endpoint, {}, {});
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'getAll') {
                        const endpoint = `/v3/company/${companyId}/query`;
                        responseData = await GenericFunctions_1.handleListing.call(this, i, endpoint, resource);
                    }
                    else if (operation === 'update') {
                        let body = {
                            Id: this.getNodeParameter('customerId', i),
                            SyncToken: await GenericFunctions_1.getSyncToken.call(this, i, companyId, resource),
                            sparse: true,
                        };
                        const updateFields = this.getNodeParameter('updateFields', i);
                        if ((0, lodash_isempty_1.default)(updateFields)) {
                            throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Please enter at least one field to update for the ${resource}.`, { itemIndex: i });
                        }
                        body = GenericFunctions_1.populateFields.call(this, body, updateFields, resource);
                        const endpoint = `/v3/company/${companyId}/${resource}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, {}, body);
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                }
                else if (resource === 'employee') {
                    if (operation === 'create') {
                        let body = {
                            FamilyName: this.getNodeParameter('FamilyName', i),
                            GivenName: this.getNodeParameter('GivenName', i),
                        };
                        const additionalFields = this.getNodeParameter('additionalFields', i);
                        body = GenericFunctions_1.populateFields.call(this, body, additionalFields, resource);
                        const endpoint = `/v3/company/${companyId}/${resource}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, {}, body);
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'get') {
                        const employeeId = this.getNodeParameter('employeeId', i);
                        const endpoint = `/v3/company/${companyId}/${resource}/${employeeId}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'GET', endpoint, {}, {});
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'getAll') {
                        const endpoint = `/v3/company/${companyId}/query`;
                        responseData = await GenericFunctions_1.handleListing.call(this, i, endpoint, resource);
                    }
                    else if (operation === 'update') {
                        let body = {
                            Id: this.getNodeParameter('employeeId', i),
                            SyncToken: await GenericFunctions_1.getSyncToken.call(this, i, companyId, resource),
                            sparse: true,
                        };
                        const updateFields = this.getNodeParameter('updateFields', i);
                        if ((0, lodash_isempty_1.default)(updateFields)) {
                            throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Please enter at least one field to update for the ${resource}.`, { itemIndex: i });
                        }
                        body = GenericFunctions_1.populateFields.call(this, body, updateFields, resource);
                        const endpoint = `/v3/company/${companyId}/${resource}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, {}, body);
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                }
                else if (resource === 'estimate') {
                    if (operation === 'create') {
                        const lines = this.getNodeParameter('Line', i);
                        if (!lines.length) {
                            throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Please enter at least one line for the ${resource}.`, { itemIndex: i });
                        }
                        if (lines.some((line) => line.DetailType === undefined ||
                            line.Amount === undefined ||
                            line.Description === undefined)) {
                            throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Please enter detail type, amount and description for every line.', { itemIndex: i });
                        }
                        lines.forEach((line) => {
                            if (line.DetailType === 'SalesItemLineDetail' && line.itemId === undefined) {
                                throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Please enter an item ID for the associated line.', { itemIndex: i });
                            }
                        });
                        let body = {
                            CustomerRef: {
                                value: this.getNodeParameter('CustomerRef', i),
                            },
                        };
                        body.Line = GenericFunctions_1.processLines.call(this, body, lines, resource);
                        const additionalFields = this.getNodeParameter('additionalFields', i);
                        body = GenericFunctions_1.populateFields.call(this, body, additionalFields, resource);
                        const endpoint = `/v3/company/${companyId}/${resource}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, {}, body);
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'delete') {
                        const qs = {
                            operation: 'delete',
                        };
                        const body = {
                            Id: this.getNodeParameter('estimateId', i),
                            SyncToken: await GenericFunctions_1.getSyncToken.call(this, i, companyId, resource),
                        };
                        const endpoint = `/v3/company/${companyId}/${resource}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, qs, body);
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'get') {
                        const estimateId = this.getNodeParameter('estimateId', i);
                        const download = this.getNodeParameter('download', i);
                        if (download) {
                            responseData = await GenericFunctions_1.handleBinaryData.call(this, items, i, companyId, resource, estimateId);
                        }
                        else {
                            const endpoint = `/v3/company/${companyId}/${resource}/${estimateId}`;
                            responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'GET', endpoint, {}, {});
                            responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                        }
                    }
                    else if (operation === 'getAll') {
                        const endpoint = `/v3/company/${companyId}/query`;
                        responseData = await GenericFunctions_1.handleListing.call(this, i, endpoint, resource);
                    }
                    else if (operation === 'send') {
                        const estimateId = this.getNodeParameter('estimateId', i);
                        const qs = {
                            sendTo: this.getNodeParameter('email', i),
                        };
                        const endpoint = `/v3/company/${companyId}/${resource}/${estimateId}/send`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, qs, {});
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'update') {
                        const { ref, syncToken } = await GenericFunctions_1.getRefAndSyncToken.call(this, i, companyId, resource, 'CustomerRef');
                        let body = {
                            Id: this.getNodeParameter('estimateId', i),
                            SyncToken: syncToken,
                            sparse: true,
                            CustomerRef: {
                                name: ref.name,
                                value: ref.value,
                            },
                        };
                        const updateFields = this.getNodeParameter('updateFields', i);
                        if ((0, lodash_isempty_1.default)(updateFields)) {
                            throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Please enter at least one field to update for the ${resource}.`, { itemIndex: i });
                        }
                        body = GenericFunctions_1.populateFields.call(this, body, updateFields, resource);
                        const endpoint = `/v3/company/${companyId}/${resource}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, {}, body);
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                }
                else if (resource === 'invoice') {
                    if (operation === 'create') {
                        const lines = this.getNodeParameter('Line', i);
                        if (!lines.length) {
                            throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Please enter at least one line for the ${resource}.`, { itemIndex: i });
                        }
                        if (lines.some((line) => line.DetailType === undefined ||
                            line.Amount === undefined ||
                            line.Description === undefined)) {
                            throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Please enter detail type, amount and description for every line.', { itemIndex: i });
                        }
                        lines.forEach((line) => {
                            if (line.DetailType === 'SalesItemLineDetail' && line.itemId === undefined) {
                                throw new n8n_workflow_1.NodeOperationError(this.getNode(), 'Please enter an item ID for the associated line.', { itemIndex: i });
                            }
                        });
                        let body = {
                            CustomerRef: {
                                value: this.getNodeParameter('CustomerRef', i),
                            },
                        };
                        body.Line = GenericFunctions_1.processLines.call(this, body, lines, resource);
                        const additionalFields = this.getNodeParameter('additionalFields', i);
                        body = GenericFunctions_1.populateFields.call(this, body, additionalFields, resource);
                        const endpoint = `/v3/company/${companyId}/${resource}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, {}, body);
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'delete') {
                        const qs = {
                            operation: 'delete',
                        };
                        const body = {
                            Id: this.getNodeParameter('invoiceId', i),
                            SyncToken: await GenericFunctions_1.getSyncToken.call(this, i, companyId, resource),
                        };
                        const endpoint = `/v3/company/${companyId}/${resource}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, qs, body);
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'get') {
                        const invoiceId = this.getNodeParameter('invoiceId', i);
                        const download = this.getNodeParameter('download', i);
                        if (download) {
                            responseData = await GenericFunctions_1.handleBinaryData.call(this, items, i, companyId, resource, invoiceId);
                        }
                        else {
                            const endpoint = `/v3/company/${companyId}/${resource}/${invoiceId}`;
                            responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'GET', endpoint, {}, {});
                            responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                        }
                    }
                    else if (operation === 'getAll') {
                        const endpoint = `/v3/company/${companyId}/query`;
                        responseData = await GenericFunctions_1.handleListing.call(this, i, endpoint, resource);
                    }
                    else if (operation === 'send') {
                        const invoiceId = this.getNodeParameter('invoiceId', i);
                        const qs = {
                            sendTo: this.getNodeParameter('email', i),
                        };
                        const endpoint = `/v3/company/${companyId}/${resource}/${invoiceId}/send`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, qs, {});
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'update') {
                        const { ref, syncToken } = await GenericFunctions_1.getRefAndSyncToken.call(this, i, companyId, resource, 'CustomerRef');
                        let body = {
                            Id: this.getNodeParameter('invoiceId', i),
                            SyncToken: syncToken,
                            sparse: true,
                            CustomerRef: {
                                name: ref.name,
                                value: ref.value,
                            },
                        };
                        const updateFields = this.getNodeParameter('updateFields', i);
                        if ((0, lodash_isempty_1.default)(updateFields)) {
                            throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Please enter at least one field to update for the ${resource}.`, { itemIndex: i });
                        }
                        body = GenericFunctions_1.populateFields.call(this, body, updateFields, resource);
                        const endpoint = `/v3/company/${companyId}/${resource}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, {}, body);
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'void') {
                        const qs = {
                            Id: this.getNodeParameter('invoiceId', i),
                            SyncToken: await GenericFunctions_1.getSyncToken.call(this, i, companyId, resource),
                            operation: 'void',
                        };
                        const endpoint = `/v3/company/${companyId}/${resource}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, qs, {});
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                }
                else if (resource === 'item') {
                    if (operation === 'get') {
                        const item = this.getNodeParameter('itemId', i);
                        const endpoint = `/v3/company/${companyId}/${resource}/${item}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'GET', endpoint, {}, {});
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'getAll') {
                        const endpoint = `/v3/company/${companyId}/query`;
                        responseData = await GenericFunctions_1.handleListing.call(this, i, endpoint, resource);
                    }
                }
                else if (resource === 'payment') {
                    if (operation === 'create') {
                        let body = {
                            CustomerRef: {
                                value: this.getNodeParameter('CustomerRef', i),
                            },
                            TotalAmt: this.getNodeParameter('TotalAmt', i),
                        };
                        const additionalFields = this.getNodeParameter('additionalFields', i);
                        body = GenericFunctions_1.populateFields.call(this, body, additionalFields, resource);
                        const endpoint = `/v3/company/${companyId}/${resource}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, {}, body);
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'delete') {
                        const qs = {
                            operation: 'delete',
                        };
                        const body = {
                            Id: this.getNodeParameter('paymentId', i),
                            SyncToken: await GenericFunctions_1.getSyncToken.call(this, i, companyId, resource),
                        };
                        const endpoint = `/v3/company/${companyId}/${resource}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, qs, body);
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'get') {
                        const paymentId = this.getNodeParameter('paymentId', i);
                        const download = this.getNodeParameter('download', i);
                        if (download) {
                            responseData = await GenericFunctions_1.handleBinaryData.call(this, items, i, companyId, resource, paymentId);
                        }
                        else {
                            const endpoint = `/v3/company/${companyId}/${resource}/${paymentId}`;
                            responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'GET', endpoint, {}, {});
                            responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                        }
                    }
                    else if (operation === 'getAll') {
                        const endpoint = `/v3/company/${companyId}/query`;
                        responseData = await GenericFunctions_1.handleListing.call(this, i, endpoint, resource);
                    }
                    else if (operation === 'send') {
                        const paymentId = this.getNodeParameter('paymentId', i);
                        const qs = {
                            sendTo: this.getNodeParameter('email', i),
                        };
                        const endpoint = `/v3/company/${companyId}/${resource}/${paymentId}/send`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, qs, {});
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'update') {
                        const { ref, syncToken } = await GenericFunctions_1.getRefAndSyncToken.call(this, i, companyId, resource, 'CustomerRef');
                        let body = {
                            Id: this.getNodeParameter('paymentId', i),
                            SyncToken: syncToken,
                            sparse: true,
                            CustomerRef: {
                                name: ref.name,
                                value: ref.value,
                            },
                        };
                        const updateFields = this.getNodeParameter('updateFields', i);
                        if ((0, lodash_isempty_1.default)(updateFields)) {
                            throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Please enter at least one field to update for the ${resource}.`, { itemIndex: i });
                        }
                        body = GenericFunctions_1.populateFields.call(this, body, updateFields, resource);
                        const endpoint = `/v3/company/${companyId}/${resource}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, {}, body);
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'void') {
                        const qs = {
                            Id: this.getNodeParameter('paymentId', i),
                            SyncToken: await GenericFunctions_1.getSyncToken.call(this, i, companyId, resource),
                            operation: 'void',
                        };
                        const endpoint = `/v3/company/${companyId}/${resource}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, qs, {});
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                }
                else if (resource === 'purchase') {
                    if (operation === 'get') {
                        const purchaseId = this.getNodeParameter('purchaseId', i);
                        const endpoint = `/v3/company/${companyId}/${resource}/${purchaseId}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'GET', endpoint, {}, {});
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'getAll') {
                        const endpoint = `/v3/company/${companyId}/query`;
                        responseData = await GenericFunctions_1.handleListing.call(this, i, endpoint, resource);
                    }
                }
                else if (resource === 'transaction') {
                    if (operation === 'getReport') {
                        const { columns, memo, term, customer, vendor, ...rest } = this.getNodeParameter('filters', i);
                        let qs = { ...rest };
                        if (columns === null || columns === void 0 ? void 0 : columns.length) {
                            qs.columns = columns.join(',');
                        }
                        if (memo === null || memo === void 0 ? void 0 : memo.length) {
                            qs.memo = memo.join(',');
                        }
                        if (term === null || term === void 0 ? void 0 : term.length) {
                            qs.term = term.join(',');
                        }
                        if (customer === null || customer === void 0 ? void 0 : customer.length) {
                            qs.customer = customer.join(',');
                        }
                        if (vendor === null || vendor === void 0 ? void 0 : vendor.length) {
                            qs.vendor = vendor.join(',');
                        }
                        qs = (0, GenericFunctions_1.adjustTransactionDates)(qs);
                        const endpoint = `/v3/company/${companyId}/reports/TransactionList`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'GET', endpoint, qs, {});
                        const simplifyResponse = this.getNodeParameter('simple', i, true);
                        if (!Object.keys(responseData === null || responseData === void 0 ? void 0 : responseData.Rows).length) {
                            responseData = [];
                        }
                        if (simplifyResponse && !Array.isArray(responseData)) {
                            responseData = (0, GenericFunctions_1.simplifyTransactionReport)(responseData);
                        }
                    }
                }
                else if (resource === 'vendor') {
                    if (operation === 'create') {
                        let body = {
                            DisplayName: this.getNodeParameter('displayName', i),
                        };
                        const additionalFields = this.getNodeParameter('additionalFields', i);
                        body = GenericFunctions_1.populateFields.call(this, body, additionalFields, resource);
                        const endpoint = `/v3/company/${companyId}/${resource}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, {}, body);
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'get') {
                        const vendorId = this.getNodeParameter('vendorId', i);
                        const endpoint = `/v3/company/${companyId}/${resource}/${vendorId}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'GET', endpoint, {}, {});
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                    else if (operation === 'getAll') {
                        const endpoint = `/v3/company/${companyId}/query`;
                        responseData = await GenericFunctions_1.handleListing.call(this, i, endpoint, resource);
                    }
                    else if (operation === 'update') {
                        let body = {
                            Id: this.getNodeParameter('vendorId', i),
                            SyncToken: await GenericFunctions_1.getSyncToken.call(this, i, companyId, resource),
                            sparse: true,
                        };
                        const updateFields = this.getNodeParameter('updateFields', i);
                        if ((0, lodash_isempty_1.default)(updateFields)) {
                            throw new n8n_workflow_1.NodeOperationError(this.getNode(), `Please enter at least one field to update for the ${resource}.`, { itemIndex: i });
                        }
                        body = GenericFunctions_1.populateFields.call(this, body, updateFields, resource);
                        const endpoint = `/v3/company/${companyId}/${resource}`;
                        responseData = await GenericFunctions_1.quickBooksApiRequest.call(this, 'POST', endpoint, {}, body);
                        responseData = responseData[(0, change_case_1.capitalCase)(resource)];
                    }
                }
            }
            catch (error) {
                if (this.continueOnFail()) {
                    const download = this.getNodeParameter('download', 0, false);
                    if (['invoice', 'estimate', 'payment'].includes(resource) &&
                        ['get'].includes(operation) &&
                        download) {
                        if (!responseData) {
                            items[i].json = { error: error.message };
                            responseData = items;
                        }
                        else {
                            responseData[i].json = { error: error.message };
                        }
                    }
                    else {
                        const executionErrorData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray({ error: error.message }), { itemData: { item: i } });
                        returnData.push(...executionErrorData);
                    }
                    continue;
                }
                throw error;
            }
            const executionData = this.helpers.constructExecutionMetaData(this.helpers.returnJsonArray(responseData), { itemData: { item: i } });
            returnData.push(...executionData);
        }
        const download = this.getNodeParameter('download', 0, false);
        if (['invoice', 'estimate', 'payment'].includes(resource) &&
            ['get'].includes(operation) &&
            download) {
            return this.prepareOutputData(responseData);
        }
        else {
            return this.prepareOutputData(returnData);
        }
    }
}
exports.QuickBooks = QuickBooks;
