// n8n-nodes-base/dist/nodes/ERPNext/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.toSQL = exports.processNames = void 0;
const lodash_flow_1 = __importDefault(require("lodash.flow"));
const lodash_sortby_1 = __importDefault(require("lodash.sortby"));
const lodash_uniqby_1 = __importDefault(require("lodash.uniqby"));
const ensureName = (docFields) => docFields.filter((o) => o.name);
const sortByName = (docFields) => (0, lodash_sortby_1.default)(docFields, ['name']);
const uniqueByName = (docFields) => (0, lodash_uniqby_1.default)(docFields, (o) => o.name);
exports.processNames = (0, lodash_flow_1.default)(ensureName, sortByName, uniqueByName);
const toSQL = (operator) => {
    const operators = {
        is: '=',
        isNot: '!=',
        greater: '>',
        less: '<',
        equalsGreater: '>=',
        equalsLess: '<=',
    };
    return operators[operator];
};
exports.toSQL = toSQL;
